{
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python",
      "version": "3.10.12",
      "mimetype": "text/x-python",
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "pygments_lexer": "ipython3",
      "nbconvert_exporter": "python",
      "file_extension": ".py"
    },
    "kaggle": {
      "accelerator": "gpu",
      "dataSources": [
        {
          "sourceId": 87793,
          "databundleVersionId": 11553390,
          "sourceType": "competition"
        },
        {
          "sourceId": 11068753,
          "sourceType": "datasetVersion",
          "datasetId": 6896780
        }
      ],
      "dockerImageVersionId": 30919,
      "isInternetEnabled": false,
      "language": "python",
      "sourceType": "notebook",
      "isGpuEnabled": true
    },
    "colab": {
      "name": "protenix inference",
      "provenance": [],
      "gpuType": "T4",
      "include_colab_link": true
    },
    "accelerator": "GPU",
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "17f295861ecc49eca10afdb246a91595": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "VBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "VBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "VBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_dc2fed3ae46346f0a6808548c9e83282"
            ],
            "layout": "IPY_MODEL_04d95293ed16479e8c354ac43c8870ef"
          }
        },
        "5ffba5e38dbb4278ad9aa9373284da16": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_0e5921a5e47c49d58806ced70b4004e6",
            "placeholder": "​",
            "style": "IPY_MODEL_5ffa303441cc4f3eaa3a5bbe44388fae",
            "value": "<center> <img\nsrc=https://www.kaggle.com/static/images/site-logo.png\nalt='Kaggle'> <br> Create an API token from <a\nhref=\"https://www.kaggle.com/settings/account\" target=\"_blank\">your Kaggle\nsettings page</a> and paste it below along with your Kaggle username. <br> </center>"
          }
        },
        "85447710686e4e5bb3504336d6b0eb39": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "TextModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "TextModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "TextView",
            "continuous_update": true,
            "description": "Username:",
            "description_tooltip": null,
            "disabled": false,
            "layout": "IPY_MODEL_667a151dab3c43eaa90fd5e36aea1af9",
            "placeholder": "​",
            "style": "IPY_MODEL_85c43a90f52c40d5b98078eeb318bc6e",
            "value": "georgemc"
          }
        },
        "8ae645655fda4e4198d3397a50c754db": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "PasswordModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "PasswordModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "PasswordView",
            "continuous_update": true,
            "description": "Token:",
            "description_tooltip": null,
            "disabled": false,
            "layout": "IPY_MODEL_3d8495e6bec747758fb0bc19ab75579f",
            "placeholder": "​",
            "style": "IPY_MODEL_29af83119cf14029a3c36699d39ac1a6",
            "value": ""
          }
        },
        "6745d94c306a4c989eb5a4c94ab575c6": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ButtonModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ButtonModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ButtonView",
            "button_style": "",
            "description": "Login",
            "disabled": false,
            "icon": "",
            "layout": "IPY_MODEL_91ad3f35b03b4b41b21f0a267f85bd1f",
            "style": "IPY_MODEL_60a0136887c3425b91d3d573ce310b5b",
            "tooltip": ""
          }
        },
        "c2393b7bff104069b4cd09e6a8114c3c": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_d091ba00b2744d4480f0a27206360405",
            "placeholder": "​",
            "style": "IPY_MODEL_c46aad3702824bdaa9869a09f4b698ee",
            "value": "\n<b>Thank You</b></center>"
          }
        },
        "04d95293ed16479e8c354ac43c8870ef": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": "center",
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": "flex",
            "flex": null,
            "flex_flow": "column",
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": "50%"
          }
        },
        "0e5921a5e47c49d58806ced70b4004e6": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "5ffa303441cc4f3eaa3a5bbe44388fae": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "667a151dab3c43eaa90fd5e36aea1af9": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "85c43a90f52c40d5b98078eeb318bc6e": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "3d8495e6bec747758fb0bc19ab75579f": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "29af83119cf14029a3c36699d39ac1a6": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "91ad3f35b03b4b41b21f0a267f85bd1f": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "60a0136887c3425b91d3d573ce310b5b": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ButtonStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ButtonStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "button_color": null,
            "font_weight": ""
          }
        },
        "d091ba00b2744d4480f0a27206360405": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "c46aad3702824bdaa9869a09f4b698ee": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "270bca2aae3c4b69b07e53b6cd0d0e05": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "LabelModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "LabelModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "LabelView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_0d82a6b3f62240dab2649b711cd18e29",
            "placeholder": "​",
            "style": "IPY_MODEL_669c829743e048149b16fff382cdf973",
            "value": "Connecting..."
          }
        },
        "0d82a6b3f62240dab2649b711cd18e29": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "669c829743e048149b16fff382cdf973": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "dc2fed3ae46346f0a6808548c9e83282": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "LabelModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "LabelModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "LabelView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_cb598bf9d29b4969a466b0b8fb07c4ad",
            "placeholder": "​",
            "style": "IPY_MODEL_c62963cde5c245cca0f93654b24f39a5",
            "value": "Kaggle credentials successfully validated."
          }
        },
        "cb598bf9d29b4969a466b0b8fb07c4ad": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "c62963cde5c245cca0f93654b24f39a5": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        }
      }
    }
  },
  "nbformat_minor": 0,
  "nbformat": 4,
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/gtg975n/BME558/blob/main/protenix_inference.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "source": [
        "# IMPORTANT: SOME KAGGLE DATA SOURCES ARE PRIVATE\n",
        "# RUN THIS CELL IN ORDER TO IMPORT YOUR KAGGLE DATA SOURCES.\n",
        "import kagglehub\n",
        "kagglehub.login()\n"
      ],
      "metadata": {
        "id": "W3ERNiktZceo",
        "outputId": "5bc5a13e-64b5-47f5-e4de-37457785f519",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 83,
          "referenced_widgets": [
            "17f295861ecc49eca10afdb246a91595",
            "5ffba5e38dbb4278ad9aa9373284da16",
            "85447710686e4e5bb3504336d6b0eb39",
            "8ae645655fda4e4198d3397a50c754db",
            "6745d94c306a4c989eb5a4c94ab575c6",
            "c2393b7bff104069b4cd09e6a8114c3c",
            "04d95293ed16479e8c354ac43c8870ef",
            "0e5921a5e47c49d58806ced70b4004e6",
            "5ffa303441cc4f3eaa3a5bbe44388fae",
            "667a151dab3c43eaa90fd5e36aea1af9",
            "85c43a90f52c40d5b98078eeb318bc6e",
            "3d8495e6bec747758fb0bc19ab75579f",
            "29af83119cf14029a3c36699d39ac1a6",
            "91ad3f35b03b4b41b21f0a267f85bd1f",
            "60a0136887c3425b91d3d573ce310b5b",
            "d091ba00b2744d4480f0a27206360405",
            "c46aad3702824bdaa9869a09f4b698ee",
            "270bca2aae3c4b69b07e53b6cd0d0e05",
            "0d82a6b3f62240dab2649b711cd18e29",
            "669c829743e048149b16fff382cdf973",
            "dc2fed3ae46346f0a6808548c9e83282",
            "cb598bf9d29b4969a466b0b8fb07c4ad",
            "c62963cde5c245cca0f93654b24f39a5"
          ]
        }
      },
      "cell_type": "code",
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "VBox(children=(HTML(value='<center> <img\\nsrc=https://www.kaggle.com/static/images/site-logo.png\\nalt=\\'Kaggle…"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "17f295861ecc49eca10afdb246a91595"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Kaggle credentials set.\n",
            "Kaggle credentials successfully validated.\n"
          ]
        }
      ],
      "execution_count": 11
    },
    {
      "source": [
        "# IMPORTANT: RUN THIS CELL IN ORDER TO IMPORT YOUR KAGGLE DATA SOURCES,\n",
        "# THEN FEEL FREE TO DELETE THIS CELL.\n",
        "# NOTE: THIS NOTEBOOK ENVIRONMENT DIFFERS FROM KAGGLE'S PYTHON\n",
        "# ENVIRONMENT SO THERE MAY BE MISSING LIBRARIES USED BY YOUR\n",
        "# NOTEBOOK.\n",
        "\n",
        "stanford_rna_3d_folding_path = kagglehub.competition_download('stanford-rna-3d-folding')\n",
        "doheon114_required_presets_path = kagglehub.dataset_download('doheon114/required-presets')\n",
        "\n",
        "print('Data source import complete.')\n"
      ],
      "metadata": {
        "id": "ZIASBsNCZcep",
        "outputId": "4b5f82bf-4e99-42a9-b57a-5c801943bdcd",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "cell_type": "code",
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Downloading from https://www.kaggle.com/api/v1/competitions/data/download-all/stanford-rna-3d-folding...\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 120M/120M [00:07<00:00, 16.3MB/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Extracting files...\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Downloading from https://www.kaggle.com/api/v1/datasets/download/doheon114/required-presets?dataset_version_number=6...\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 2.97G/2.97G [02:34<00:00, 20.7MB/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Extracting files...\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Data source import complete.\n"
          ]
        }
      ],
      "execution_count": 6
    },
    {
      "cell_type": "code",
      "source": [
        "!mkdir -p ~/.kaggle\n",
        "!mv kaggle-2.json ~/.kaggle/kaggle.json\n",
        "!chmod 600 ~/.kaggle/kaggle.json\n",
        "\n",
        "!kaggle datasets download -d doheon114/required-presets\n"
      ],
      "metadata": {
        "id": "vf3KSMuklad6",
        "outputId": "1b1db380-25fa-43a6-d22e-9fe84fdbc806",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Dataset URL: https://www.kaggle.com/datasets/doheon114/required-presets\n",
            "License(s): MIT\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!unzip -o required-presets.zip -d required_presets"
      ],
      "metadata": {
        "id": "K3A6a-dEnbG0",
        "outputId": "1de7b74f-b188-4fb8-feaf-21f7e56b496e",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Archive:  required-presets.zip\n",
            "  inflating: required_presets/Protenix/.flake8  \n",
            "  inflating: required_presets/Protenix/.git/HEAD  \n",
            "  inflating: required_presets/Protenix/.git/config  \n",
            "  inflating: required_presets/Protenix/.git/description  \n",
            "  inflating: required_presets/Protenix/.git/hooks/applypatch-msg.sample  \n",
            "  inflating: required_presets/Protenix/.git/hooks/commit-msg.sample  \n",
            "  inflating: required_presets/Protenix/.git/hooks/fsmonitor-watchman.sample  \n",
            "  inflating: required_presets/Protenix/.git/hooks/post-update.sample  \n",
            "  inflating: required_presets/Protenix/.git/hooks/pre-applypatch.sample  \n",
            "  inflating: required_presets/Protenix/.git/hooks/pre-commit.sample  \n",
            "  inflating: required_presets/Protenix/.git/hooks/pre-merge-commit.sample  \n",
            "  inflating: required_presets/Protenix/.git/hooks/pre-push.sample  \n",
            "  inflating: required_presets/Protenix/.git/hooks/pre-rebase.sample  \n",
            "  inflating: required_presets/Protenix/.git/hooks/pre-receive.sample  \n",
            "  inflating: required_presets/Protenix/.git/hooks/prepare-commit-msg.sample  \n",
            "  inflating: required_presets/Protenix/.git/hooks/push-to-checkout.sample  \n",
            "  inflating: required_presets/Protenix/.git/hooks/update.sample  \n",
            "  inflating: required_presets/Protenix/.git/index  \n",
            "  inflating: required_presets/Protenix/.git/info/exclude  \n",
            "  inflating: required_presets/Protenix/.git/logs/HEAD  \n",
            "  inflating: required_presets/Protenix/.git/logs/refs/heads/main  \n",
            "  inflating: required_presets/Protenix/.git/logs/refs/remotes/origin/HEAD  \n",
            "  inflating: required_presets/Protenix/.git/objects/pack/pack-2702e54af38ccc2b34268b329610c99b0530f617.idx  \n",
            "  inflating: required_presets/Protenix/.git/objects/pack/pack-2702e54af38ccc2b34268b329610c99b0530f617.pack  \n",
            "  inflating: required_presets/Protenix/.git/packed-refs  \n",
            "  inflating: required_presets/Protenix/.git/refs/heads/main  \n",
            "  inflating: required_presets/Protenix/.git/refs/remotes/origin/HEAD  \n",
            "  inflating: required_presets/Protenix/.github/workflows/ci.yml  \n",
            "  inflating: required_presets/Protenix/.github/workflows/scorecard.yml  \n",
            "  inflating: required_presets/Protenix/.gitignore  \n",
            "  inflating: required_presets/Protenix/.pre-commit-config.yaml  \n",
            "  inflating: required_presets/Protenix/CODE_OF_CONDUCT.md  \n",
            "  inflating: required_presets/Protenix/CONTRIBUTING.md  \n",
            "  inflating: required_presets/Protenix/Dockerfile  \n",
            "  inflating: required_presets/Protenix/LICENSE  \n",
            "  inflating: required_presets/Protenix/Protenix_Technical_Report.pdf  \n",
            "  inflating: required_presets/Protenix/README.md  \n",
            "  inflating: required_presets/Protenix/assets/constraint_metrics.png  \n",
            "  inflating: required_presets/Protenix/assets/license_header.txt  \n",
            "  inflating: required_presets/Protenix/assets/overall_metrics.png  \n",
            "  inflating: required_presets/Protenix/assets/protenix_predictions.gif  \n",
            "  inflating: required_presets/Protenix/build/lib/configs/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/configs/configs_base.py  \n",
            "  inflating: required_presets/Protenix/build/lib/configs/configs_data.py  \n",
            "  inflating: required_presets/Protenix/build/lib/configs/configs_inference.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/config/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/config/config.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/config/extend_types.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/ccd.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/constants.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/data_pipeline.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/dataloader.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/dataset.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/featurizer.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/filter.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/infer_data_pipeline.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/json_maker.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/json_parser.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/json_to_feature.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/msa_featurizer.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/msa_utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/parser.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/substructure_perms.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/tokenizer.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/data/utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/metrics/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/metrics/clash.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/metrics/lddt_metrics.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/metrics/rmsd.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/generator.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/layer_norm/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/layer_norm/kernel/compat.h  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/layer_norm/kernel/layer_norm_cuda.cpp  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/layer_norm/kernel/layer_norm_cuda_kernel.cu  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/layer_norm/kernel/type_shim.h  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/layer_norm/layer_norm.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/layer_norm/torch_ext_compile.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/loss.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/modules/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/modules/confidence.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/modules/diffusion.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/modules/embedders.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/modules/frames.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/modules/head.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/modules/pairformer.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/modules/primitives.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/modules/transformer.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/protenix.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/sample_confidence.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/model/utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/README.md  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/data/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/data/data_transforms.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/data/errors.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/data/mmcif_parsing.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/data/msa_identifiers.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/data/msa_pairing.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/data/parsers.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/data/templates.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/data/tools/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/data/tools/jackhmmer.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/data/tools/utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/model/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/model/dropout.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/model/outer_product_mean.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/model/primitives.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/model/triangular_attention.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/model/triangular_multiplicative_update.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/np/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/np/residue_constants.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/all_atom_multimer.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/checkpointing.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/chunk_utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/feats.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/geometry/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/geometry/quat_rigid.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/geometry/rigid_matrix_vector.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/geometry/rotation_matrix.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/geometry/test_utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/geometry/utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/geometry/vector.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/kernel/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/kernel/attention_core.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/kernel/csrc/compat.h  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/kernel/csrc/softmax_cuda.cpp  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/kernel/csrc/softmax_cuda_kernel.cu  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/kernel/csrc/softmax_cuda_stub.cpp  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/precision_utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/rigid_utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/openfold_local/utils/tensor_utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/cropping.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/distributed.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/file_io.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/geometry.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/logger.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/lr_scheduler.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/metrics.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/permutation/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/permutation/atom_permutation.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/permutation/chain_permutation/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/permutation/chain_permutation/heuristic.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/permutation/chain_permutation/pocket_based_permutation.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/permutation/chain_permutation/utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/permutation/permutation.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/permutation/utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/scatter_utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/seed.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/torch_utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/utils/training.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/web_service/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/web_service/colab_request_parser.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/web_service/colab_request_utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/web_service/dependency_url.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/web_service/prediction_visualization.py  \n",
            "  inflating: required_presets/Protenix/build/lib/protenix/web_service/viewer.py  \n",
            "  inflating: required_presets/Protenix/build/lib/runner/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/runner/batch_inference.py  \n",
            "  inflating: required_presets/Protenix/build/lib/runner/dumper.py  \n",
            "  inflating: required_presets/Protenix/build/lib/runner/ema.py  \n",
            "  inflating: required_presets/Protenix/build/lib/runner/inference.py  \n",
            "  inflating: required_presets/Protenix/build/lib/runner/msa_search.py  \n",
            "  inflating: required_presets/Protenix/build/lib/runner/train.py  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/colabfold_msa.py  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/gen_ccd_cache.py  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif/102m.cif  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif/1k1a.cif  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif/5zyh.cif  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif/7zzx.cif  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif/8f9h.cif  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa/0/mmseqs_other_hits.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa/0/uniref100_hits.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa/1/mmseqs_other_hits.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa/1/uniref100_hits.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa/2/mmseqs_other_hits.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa/2/uniref100_hits.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa/3/mmseqs_other_hits.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa/3/uniref100_hits.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa_initial/0.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa_initial/1.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa_initial/2.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa_initial/3.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa_initial/pdb70_220313_db.m8  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa_initial/uniref_tax.m8  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa_with_taxid/0.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa_with_taxid/1.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa_with_taxid/2.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa_with_taxid/3.a3m  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/pdb_seqs/pdb_index_to_seq.json  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/pdb_seqs/pdb_seq.csv  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/pdb_seqs/pdb_seq.fasta  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/pdb_seqs/seq_to_pdb_id_entity_id.json  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/data/pdb_seqs/seq_to_pdb_index.json  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/step1-get_prot_seq.py  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/step2-get_msa.ipynb  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/step3-uniref_add_taxid.py  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/step4-split_msa_to_uniref_and_others.py  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/msa/utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/scripts/prepare_training_data.py  \n",
            "  inflating: required_presets/Protenix/build/lib/tests/__init__.py  \n",
            "  inflating: required_presets/Protenix/build/lib/tests/test_attention_pair_bias.py  \n",
            "  inflating: required_presets/Protenix/build/lib/tests/test_condition_transition.py  \n",
            "  inflating: required_presets/Protenix/build/lib/tests/test_diffusion_transformer.py  \n",
            "  inflating: required_presets/Protenix/build/lib/tests/test_frame.py  \n",
            "  inflating: required_presets/Protenix/build/lib/tests/test_local_rearrange.py  \n",
            "  inflating: required_presets/Protenix/build/lib/tests/test_lr_schedule.py  \n",
            "  inflating: required_presets/Protenix/build/lib/tests/test_utils.py  \n",
            "  inflating: required_presets/Protenix/build/lib/tests/test_weighted_rigid_align.py  \n",
            "  inflating: required_presets/Protenix/configs/__init__.py  \n",
            "  inflating: required_presets/Protenix/configs/configs_base.py  \n",
            "  inflating: required_presets/Protenix/configs/configs_data.py  \n",
            "  inflating: required_presets/Protenix/configs/configs_inference.py  \n",
            "  inflating: required_presets/Protenix/docs/colabfold_compatiable_msa.md  \n",
            "  inflating: required_presets/Protenix/docs/docker_installation.md  \n",
            "  inflating: required_presets/Protenix/docs/infer_json_format.md  \n",
            "  inflating: required_presets/Protenix/docs/kernels.md  \n",
            "  inflating: required_presets/Protenix/docs/model_train_inference_cost.md  \n",
            "  inflating: required_presets/Protenix/docs/msa_pipeline.md  \n",
            "  inflating: required_presets/Protenix/docs/prepare_training_data.md  \n",
            "  inflating: required_presets/Protenix/docs/training.md  \n",
            "  inflating: required_presets/Protenix/examples/7pzb.pdb  \n",
            "  inflating: required_presets/Protenix/examples/7pzb/msa/1/non_pairing.a3m  \n",
            "  inflating: required_presets/Protenix/examples/7pzb/msa/1/pairing.a3m  \n",
            "  inflating: required_presets/Protenix/examples/7r6r/msa/1/non_pairing.a3m  \n",
            "  inflating: required_presets/Protenix/examples/7r6r/msa/1/pairing.a3m  \n",
            "  inflating: required_presets/Protenix/examples/7wux/msa/1/non_pairing.a3m  \n",
            "  inflating: required_presets/Protenix/examples/7wux/msa/1/pairing.a3m  \n",
            "  inflating: required_presets/Protenix/examples/7wux/msa/2/non_pairing.a3m  \n",
            "  inflating: required_presets/Protenix/examples/7wux/msa/2/pairing.a3m  \n",
            "  inflating: required_presets/Protenix/examples/dimer.fasta  \n",
            "  inflating: required_presets/Protenix/examples/example.json  \n",
            "  inflating: required_presets/Protenix/examples/example_without_msa.json  \n",
            "  inflating: required_presets/Protenix/examples/finetune_subset.txt  \n",
            "  inflating: required_presets/Protenix/examples/ligands/7wux_smiles.smi  \n",
            "  inflating: required_presets/Protenix/examples/ligands/compounds-3d-R.sdf  \n",
            "  inflating: required_presets/Protenix/examples/ligands/compounds-3d-RS.sdf  \n",
            "  inflating: required_presets/Protenix/examples/prot.fasta  \n",
            "  inflating: required_presets/Protenix/finetune_demo.sh  \n",
            "  inflating: required_presets/Protenix/inference_demo.sh  \n",
            "  inflating: required_presets/Protenix/notebooks/protenix_inference.ipynb  \n",
            "  inflating: required_presets/Protenix/protenix.egg-info/PKG-INFO  \n",
            "  inflating: required_presets/Protenix/protenix.egg-info/SOURCES.txt  \n",
            "  inflating: required_presets/Protenix/protenix.egg-info/dependency_links.txt  \n",
            "  inflating: required_presets/Protenix/protenix.egg-info/entry_points.txt  \n",
            "  inflating: required_presets/Protenix/protenix.egg-info/requires.txt  \n",
            "  inflating: required_presets/Protenix/protenix.egg-info/top_level.txt  \n",
            "  inflating: required_presets/Protenix/protenix/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/config/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/config/config.py  \n",
            "  inflating: required_presets/Protenix/protenix/config/extend_types.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/ccd.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/constants.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/data_pipeline.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/dataloader.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/dataset.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/featurizer.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/filter.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/infer_data_pipeline.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/json_maker.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/json_parser.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/json_to_feature.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/msa_featurizer.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/msa_utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/parser.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/substructure_perms.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/tokenizer.py  \n",
            "  inflating: required_presets/Protenix/protenix/data/utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/metrics/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/metrics/clash.py  \n",
            "  inflating: required_presets/Protenix/protenix/metrics/lddt_metrics.py  \n",
            "  inflating: required_presets/Protenix/protenix/metrics/rmsd.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/generator.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/layer_norm/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/layer_norm/kernel/compat.h  \n",
            "  inflating: required_presets/Protenix/protenix/model/layer_norm/kernel/layer_norm_cuda.cpp  \n",
            "  inflating: required_presets/Protenix/protenix/model/layer_norm/kernel/layer_norm_cuda_kernel.cu  \n",
            "  inflating: required_presets/Protenix/protenix/model/layer_norm/kernel/type_shim.h  \n",
            "  inflating: required_presets/Protenix/protenix/model/layer_norm/layer_norm.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/layer_norm/torch_ext_compile.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/loss.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/modules/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/modules/confidence.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/modules/diffusion.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/modules/embedders.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/modules/frames.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/modules/head.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/modules/pairformer.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/modules/primitives.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/modules/transformer.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/protenix.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/sample_confidence.py  \n",
            "  inflating: required_presets/Protenix/protenix/model/utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/README.md  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/data/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/data/data_transforms.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/data/errors.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/data/mmcif_parsing.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/data/msa_identifiers.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/data/msa_pairing.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/data/parsers.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/data/templates.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/data/tools/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/data/tools/jackhmmer.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/data/tools/utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/model/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/model/dropout.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/model/outer_product_mean.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/model/primitives.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/model/triangular_attention.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/model/triangular_multiplicative_update.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/np/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/np/residue_constants.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/all_atom_multimer.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/checkpointing.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/chunk_utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/feats.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/geometry/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/geometry/quat_rigid.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/geometry/rigid_matrix_vector.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/geometry/rotation_matrix.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/geometry/test_utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/geometry/utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/geometry/vector.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/kernel/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/kernel/attention_core.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/kernel/csrc/compat.h  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/kernel/csrc/softmax_cuda.cpp  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/kernel/csrc/softmax_cuda_kernel.cu  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/kernel/csrc/softmax_cuda_stub.cpp  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/precision_utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/rigid_utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/openfold_local/utils/tensor_utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/cropping.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/distributed.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/file_io.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/geometry.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/logger.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/lr_scheduler.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/metrics.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/permutation/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/permutation/atom_permutation.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/permutation/chain_permutation/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/permutation/chain_permutation/heuristic.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/permutation/chain_permutation/pocket_based_permutation.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/permutation/chain_permutation/utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/permutation/permutation.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/permutation/utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/scatter_utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/seed.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/torch_utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/utils/training.py  \n",
            "  inflating: required_presets/Protenix/protenix/web_service/__init__.py  \n",
            "  inflating: required_presets/Protenix/protenix/web_service/colab_request_parser.py  \n",
            "  inflating: required_presets/Protenix/protenix/web_service/colab_request_utils.py  \n",
            "  inflating: required_presets/Protenix/protenix/web_service/dependency_url.py  \n",
            "  inflating: required_presets/Protenix/protenix/web_service/prediction_visualization.py  \n",
            "  inflating: required_presets/Protenix/protenix/web_service/viewer.py  \n",
            "  inflating: required_presets/Protenix/requirements.txt  \n",
            "  inflating: required_presets/Protenix/runner/__init__.py  \n",
            "  inflating: required_presets/Protenix/runner/batch_inference.py  \n",
            "  inflating: required_presets/Protenix/runner/dumper.py  \n",
            "  inflating: required_presets/Protenix/runner/ema.py  \n",
            "  inflating: required_presets/Protenix/runner/inference.py  \n",
            "  inflating: required_presets/Protenix/runner/msa_search.py  \n",
            "  inflating: required_presets/Protenix/runner/train.py  \n",
            "  inflating: required_presets/Protenix/scripts/__init__.py  \n",
            "  inflating: required_presets/Protenix/scripts/colabfold_msa.py  \n",
            "  inflating: required_presets/Protenix/scripts/gen_ccd_cache.py  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif/102m.cif  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif/1k1a.cif  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif/5zyh.cif  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif/7zzx.cif  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif/8f9h.cif  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa/0/mmseqs_other_hits.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa/0/uniref100_hits.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa/1/mmseqs_other_hits.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa/1/uniref100_hits.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa/2/mmseqs_other_hits.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa/2/uniref100_hits.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa/3/mmseqs_other_hits.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa/3/uniref100_hits.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa_initial/0.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa_initial/1.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa_initial/2.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa_initial/3.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa_initial/pdb70_220313_db.m8  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa_initial/uniref_tax.m8  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa_with_taxid/0.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa_with_taxid/1.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa_with_taxid/2.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/mmcif_msa_with_taxid/3.a3m  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/pdb_seqs/pdb_index_to_seq.json  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/pdb_seqs/pdb_seq.csv  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/pdb_seqs/pdb_seq.fasta  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/pdb_seqs/seq_to_pdb_id_entity_id.json  \n",
            "  inflating: required_presets/Protenix/scripts/msa/data/pdb_seqs/seq_to_pdb_index.json  \n",
            "  inflating: required_presets/Protenix/scripts/msa/step1-get_prot_seq.py  \n",
            "  inflating: required_presets/Protenix/scripts/msa/step2-get_msa.ipynb  \n",
            "  inflating: required_presets/Protenix/scripts/msa/step3-uniref_add_taxid.py  \n",
            "  inflating: required_presets/Protenix/scripts/msa/step4-split_msa_to_uniref_and_others.py  \n",
            "  inflating: required_presets/Protenix/scripts/msa/utils.py  \n",
            "  inflating: required_presets/Protenix/scripts/prepare_training_data.py  \n",
            "  inflating: required_presets/Protenix/setup.py  \n",
            "  inflating: required_presets/Protenix/tests/__init__.py  \n",
            "  inflating: required_presets/Protenix/tests/test_attention_pair_bias.py  \n",
            "  inflating: required_presets/Protenix/tests/test_condition_transition.py  \n",
            "  inflating: required_presets/Protenix/tests/test_diffusion_transformer.py  \n",
            "  inflating: required_presets/Protenix/tests/test_frame.py  \n",
            "  inflating: required_presets/Protenix/tests/test_local_rearrange.py  \n",
            "  inflating: required_presets/Protenix/tests/test_lr_schedule.py  \n",
            "  inflating: required_presets/Protenix/tests/test_utils.py  \n",
            "  inflating: required_presets/Protenix/tests/test_weighted_rigid_align.py  \n",
            "  inflating: required_presets/Protenix/train_demo.sh  \n",
            "  inflating: required_presets/wheels/GitPython-3.1.44-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/MarkupSafe-3.0.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/PyYAML-6.0.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/absl_py-2.2.2-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/annotated_types-0.7.0-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/asttokens-3.0.0-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/biopython-1.83-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/biotite-1.0.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/biotraj-1.2.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/certifi-2025.1.31-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/charset_normalizer-3.4.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/click-8.1.8-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/colorama-0.4.6-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/comm-0.2.2-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/contourpy-1.3.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/cycler-0.12.1-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/decorator-5.2.1-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/deepspeed-0.16.5-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/docker_pycreds-0.4.0-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/einops-0.8.1-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/exceptiongroup-1.2.2-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/executing-2.2.0-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/filelock-3.18.0-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/fonttools-4.57.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/fsspec-2025.3.2-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/gitdb-4.0.12-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/hjson-3.1.0-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/icecream-2.1.4-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/idna-3.10-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/ihm-2.4-cp310-cp310-linux_x86_64.whl  \n",
            "  inflating: required_presets/wheels/ipdb-0.13.13-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/ipython-8.35.0-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/ipywidgets-8.1.6-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/jedi-0.19.2-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/jinja2-3.1.6-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/joblib-1.4.2-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/jupyterlab_widgets-3.0.14-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/kiwisolver-1.4.8-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl  \n",
            "  inflating: required_presets/wheels/matplotlib-3.9.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/matplotlib_inline-0.1.7-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/ml_collections-1.0.0-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/modelcif-0.7-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/mpmath-1.3.0-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/msgpack-1.1.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/networkx-3.4.2-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/ninja-1.11.1.4-py3-none-manylinux_2_12_x86_64.manylinux2010_x86_64.whl  \n",
            "  inflating: required_presets/wheels/numpy-1.26.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl  \n",
            "  inflating: required_presets/wheels/nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl  \n",
            "  inflating: required_presets/wheels/nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl  \n",
            "  inflating: required_presets/wheels/nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl  \n",
            "  inflating: required_presets/wheels/nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl  \n",
            "  inflating: required_presets/wheels/nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl  \n",
            "  inflating: required_presets/wheels/nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl  \n",
            "  inflating: required_presets/wheels/nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl  \n",
            "  inflating: required_presets/wheels/nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl  \n",
            "  inflating: required_presets/wheels/nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/nvidia_nvjitlink_cu12-12.8.93-py3-none-manylinux2010_x86_64.manylinux_2_12_x86_64.whl  \n",
            "  inflating: required_presets/wheels/nvidia_nvtx_cu12-12.1.105-py3-none-manylinux1_x86_64.whl  \n",
            "  inflating: required_presets/wheels/optree-0.15.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/packaging-24.2-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/pandas-2.2.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/parso-0.8.4-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/pexpect-4.9.0-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/pillow-11.2.1-cp310-cp310-manylinux_2_28_x86_64.whl  \n",
            "  inflating: required_presets/wheels/platformdirs-4.3.7-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/prompt_toolkit-3.0.50-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/protenix-0.4.6-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/protobuf-3.20.2-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl  \n",
            "  inflating: required_presets/wheels/psutil-7.0.0-cp36-abi3-manylinux_2_12_x86_64.manylinux2010_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/ptyprocess-0.7.0-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/pure_eval-0.2.3-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/py3Dmol-2.4.2-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/py_cpuinfo-9.0.0-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/pydantic-2.11.3-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/pydantic_core-2.33.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/pygments-2.19.1-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/pyparsing-3.2.3-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/python_dateutil-2.9.0.post0-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/pytz-2025.2-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/rdkit-2024.9.6-cp310-cp310-manylinux_2_28_x86_64.whl  \n",
            "  inflating: required_presets/wheels/requests-2.32.3-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/scikit_learn-1.6.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/scikit_learn_extra-0.3.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/scipy-1.15.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/sentry_sdk-2.26.0-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/setproctitle-1.3.5-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/setuptools-78.1.0-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/six-1.17.0-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/smmap-5.0.2-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/stack_data-0.6.3-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/sympy-1.13.3-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/threadpoolctl-3.6.0-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/tomli-2.2.1-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/torch-2.3.1-cp310-cp310-manylinux1_x86_64.whl  \n",
            "  inflating: required_presets/wheels/tqdm-4.67.1-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/traitlets-5.14.3-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/triton-2.3.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/typing_extensions-4.13.2-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/typing_inspection-0.4.0-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/tzdata-2025.2-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/urllib3-2.4.0-py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/wandb-0.19.9-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl  \n",
            "  inflating: required_presets/wheels/wcwidth-0.2.13-py2.py3-none-any.whl  \n",
            "  inflating: required_presets/wheels/widgetsnbextension-4.0.14-py3-none-any.whl  \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!kaggle competitions download -c stanford-rna-3d-folding -p /content\n",
        "!unzip -o /content/stanford-rna-3d-folding.zip -d /content/stanford_rna_3d_folding\n"
      ],
      "metadata": {
        "id": "nDK5tcs7ocbB",
        "outputId": "457a3ddd-c8d8-4f8f-9008-3f7c40935f41",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Archive:  /content/stanford-rna-3d-folding.zip\n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/17RA_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1A1T_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1A4T_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1A51_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1A60_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1A9N_Q.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1AFX_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ANR_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1AQO_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ATO_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ATV_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ATW_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1AUD_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1B36_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1BAU_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1BGZ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1BIV_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1BVJ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1BZ2_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1CQ5_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1D0U_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1D6K_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1E4P_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1E7K_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1E95_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1EBQ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1EBR_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1EIY_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1EKZ_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ESH_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ESY_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ETF_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1EUQ_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1EXY_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1F5U_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1F6X_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1F6Z_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1F84_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1F85_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1F9L_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1FFK_9.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1FHK_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1FJE_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1FOQ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1FQZ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1FYO_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1G70_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1HLX_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1HMH_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1HS1_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1HS2_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1HS3_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1HS4_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1HS8_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1HVU_I.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1HWQ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1I3X_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1I46_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1I4C_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1IBM_Y.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1IDV_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1IK1_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1IKD_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1JO7_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1JOX_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1JTJ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1JTW_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1JUR_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1JWC_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1JZC_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1K2G_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1K4A_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1K4B_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1K5I_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1K6G_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1K6H_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1K9W_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1KAJ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1KIS_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1KIS_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1KKA_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1KKS_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1KP7_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1KPD_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1L1C_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1L1W_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1LC6_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1LDZ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1LS2_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1M5L_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1M82_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1MFJ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1MFK_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1MFY_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1MME_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1MNX_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1MT4_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1N34_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1N66_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1N8X_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1NA2_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1NC0_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1NYB_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1OQ0_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1OSW_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1OW9_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1P5M_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1P5N_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1P5P_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1P6V_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1P6V_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1PJY_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1Q75_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1QC8_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1QFQ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1QWA_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1QWB_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1QZA_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1QZB_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1QZC_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1QZC_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1R2P_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1R2W_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1R4H_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1R7W_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1R7Z_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1RFR_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1RHT_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1RNG_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1RNK_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ROQ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1RY1_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1S34_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1S9S_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1SCL_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1SLO_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1SZY_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1T28_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1T4L_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1TBK_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1TJZ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1TLR_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1TXS_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1U63_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ULL_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1UUU_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1VOP_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1WKS_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1WZ2_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1X18_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1X18_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1X18_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1XHP_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1XSG_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1XSH_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1XWP_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1XWU_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1Y1Y_P.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1YLG_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1YMO_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1YNE_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1YSH_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1YSH_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1YSH_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1YSV_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1Z2J_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1Z30_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1Z31_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZBH_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZC5_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZC8_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZC8_G.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZC8_H.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZC8_I.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZC8_J.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZC8_Z.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZDI_S.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZDK_S.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZIF_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZIG_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZIH_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/1ZN1_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/28SP_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2A64_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2A9L_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2ADT_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2AGN_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2AGN_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2AHT_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2AKE_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2B2E_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2B2E_S.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2B6G_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2BJ2_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2BJ2_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2BQ5_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2BQ5_S.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2BS0_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2BS0_S.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2D19_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2D1B_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2DER_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2DET_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2DR2_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2DU4_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2DU5_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2DU6_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2ESE_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2EUY_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2EVY_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2F4X_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2F87_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2F88_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2FDT_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2FEY_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2FY1_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2G1W_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2GIO_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2GIP_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2GO5_9.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2GO5_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2GRW_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2GV3_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2GV4_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2GVO_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2HEM_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2HGH_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2HNS_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2HUA_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2IHX_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2IL9_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2IL9_M.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2IXY_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2IXZ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2IY3_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2IZ8_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2IZ8_S.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2IZM_S.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2IZN_S.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2J28_8.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2J28_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2J37_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2JPP_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2JR4_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2JSE_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2JTP_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2JWV_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2JXV_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2JYM_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2K4C_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2K5Z_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2K66_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2K95_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KD8_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KE6_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KHY_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KMJ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KOC_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KPC_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KPD_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KPV_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KRL_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KRP_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KUR_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KUU_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KUV_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KUW_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KVN_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2KZL_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2L1F_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2L1F_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2L2J_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2L3C_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2L3E_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2L3J_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2L5Z_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2L6I_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LA5_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LAC_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LBJ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LBK_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LBL_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LBS_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LC8_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LDL_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LDT_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LHP_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LI4_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LJJ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LK3_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LKR_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LP9_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LPA_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LQZ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LU0_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LUB_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LUN_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2LV0_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2M12_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2M21_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2M22_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2M23_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2M4W_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2M57_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2M5U_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2M8K_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MEQ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MF0_G.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MFC_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MFD_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MFE_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MFF_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MFG_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MHI_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MI0_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MI0_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MNC_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MQT_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MS1_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MTJ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MXJ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2MXL_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N0R_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N1Q_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N2O_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N2P_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N3O_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N3Q_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N4L_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N6S_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N6T_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N6W_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N6X_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N7M_X.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N7X_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2N8V_X.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2NBX_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2NBY_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2NBZ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2NC0_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2NC1_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2NCI_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2NOQ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2NOQ_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2NOQ_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2NR0_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2NR0_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2NR0_G.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2NR0_H.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2O33_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2OB7_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2OJ8_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2OM3_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2OM7_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2OM7_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2OM7_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2OM7_G.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2OM7_H.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2OM7_I.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2OM7_J.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2PCW_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2PN9_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2QH2_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2QH3_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2QH4_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2R1G_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2R1G_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2R1G_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2R1G_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2R1G_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2R1G_X.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2R93_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2RLU_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2RN1_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2RO2_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2RPK_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2RPT_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2RVO_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2TPK_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2U2A_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2W2H_S.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2X7N_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2XXA_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2Y95_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2ZJQ_X.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/2ZJQ_Y.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3A2K_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3A3A_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3AKZ_H.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3AMU_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3CW1_v.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3DEG_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3DEG_G.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3DEG_I.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3DEG_J.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3DEG_K.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3DKN_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3DKN_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3DKN_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3EP2_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3EP2_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3EP2_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3EQ3_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3EQ3_Y.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3EQ4_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3HAY_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3IYR_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3IZ4_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3IZD_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3IZZ_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3J3V_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3J3V_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3J3W_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3J6B_e.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3J8G_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3J8G_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3JB9_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3JD5_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3JQ4_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3LWQ_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3NDB_M.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3PGW_N.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3PHP_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3PIP_Y.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3TUP_T.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3UZS_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3W1K_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3WC1_Q.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3WC2_P.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3WC2_Q.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3WFQ_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3WFR_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3WFR_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/3WQZ_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/484D_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4A4S_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4A4T_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4A4U_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4BY9_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4C4Q_N.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4D5N_X.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4DR5_W.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4ILL_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4ILM_M.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4KJI_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4KR6_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4OQ9_3.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4P8Z_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4QIL_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4TUA_XY.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4TUC_QY.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4TZP_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4TZV_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V4G_CA.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V4N_A1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V4N_A3.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V4N_B1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V4N_B2.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V4V_AA.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V4V_B0.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V4W_B9.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5F_CA.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5G_BA.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5G_BB.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5Z_AF.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5Z_AH.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5Z_BA.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5Z_BC.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5Z_BH.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5Z_BK.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5Z_BL.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5Z_BM.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5Z_BP.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5Z_BQ.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V5Z_BU.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V65_A1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V65_BB.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V6D_AV.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V6U_A0.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V6U_B1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V6W_A5.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V6W_A7.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V6W_A8.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V6W_B2.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V6W_BC.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V6X_A5.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V6X_A7.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V6X_A8.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V6X_B2.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V6X_BC.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V7E_Aa.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V7E_Ab.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V7E_Ac.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V7E_Ad.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V7E_Ae.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V8J_AV.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V8J_CW.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V99_FX.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4V9B_CD.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4WC3_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4WJ3_Q.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4WSA_V.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4WZJ_VV.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4X0A_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4X0B_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4X4P_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4X4S_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4YVI_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4YVJ_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4YVK_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4Z7L_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/4ZT9_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5A18_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5AMQ_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5ELS_I.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5EPI_H.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5FJ1_H.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5FMZ_V.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5GAP_U.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5GAP_V.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5IEM_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5KH8_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5KK5_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5KMZ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5KQE_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5LSN_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5MC6_BR.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5MC6_BS.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5MC6_m.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5MC6_n.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5MS0_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5MSF_S.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5N5C_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5O7H_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5OA3_1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5UF3_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5UZT_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5V16_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5V17_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5V6X_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5V6X_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5V93_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5V93_a.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5WQ1_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5XBL_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5XJ2_G.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5Y36_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5Z3G_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5ZAL_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5ZAM_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/5ZEY_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6AAX_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6AEB_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6AGB_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6AH3_T.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6AHU_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6AHU_T.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6AWC_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6AWD_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6BZ7_QW.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6BZ7_XY.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6BZ8_XY.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6C4H_x.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6C66_J.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6CYT_N.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6DTI_X.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6DU5_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6EVJ_V.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6FHI_V.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6FLQ_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6FT6_2.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6GBM_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6HYK_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6HYU_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6IFO_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6IV6_G.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6K0A_X.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6K0B_U.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6K3Z_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6K4S_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6KUT_V.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6KUV_V.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6LTP_H.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6MCE_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6MJ0_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6MXQ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6N7R_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6NDK_XY.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6NM9_G.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6NOA_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6O0Y_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6O0Z_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6O3M_XV.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6OF6_XW.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6OJ2_QW.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6OJ2_QY.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6OPE_XY.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6ORD_QY.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6OSI_RB.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6PK9_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6POM_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6QW6_5.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6QWL_W.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6QX3_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6QX9_1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6RFL_U.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6RR7_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6S0X_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6S0X_a.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6S0Z_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6SDW_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6SY6_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6U79_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6UES_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6UFM_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6V5B_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6VAR_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6VZC_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6W3M_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6W5C_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6W62_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6W64_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WB0_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WB1_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WD2_1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WD9_2.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WLK_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WLM_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WLN_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WLO_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WLS_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WLT_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WPI_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WW6_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WW6_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6WW6_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6XRZ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6XWJ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6XWW_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6XZD_IN1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6XZR_IN1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6Y0C_IN1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6Z6B_UUU.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6ZDU_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6ZVK_E1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6ZVK_K3.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6ZVK_d2.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6ZVK_e2.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/6ZVK_h2.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7A01_K3.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7A01_e2.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7B9V_2.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7B9V_5.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7D1A_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7D8C_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7DD4_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7DLZ_Y.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7ELE_G.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7EXY_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7JRS_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7JRT_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7K4L_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7KUB_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7KUC_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7KUD_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7LHD_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7LJY_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7LMA_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7LMB_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7LVA_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7LYG_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7LYS_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7LYT_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7M2T_ss.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7M57_cc.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7M57_ii.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7M5O_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7MSF_S.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7NK4_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7NQ4_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7OBQ_1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7ORI_H.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7ORJ_H.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7ORM_H.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7OZS_3.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7PTL_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7PTQ_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7PTS_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7QDU_Q.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7QGG_v.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7RGU_H.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7RQ5_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7RYG_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7S36_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7S38_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7S3H_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7S4U_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7S4V_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7SAM_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7SHX_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7SLP_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7U2A_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7UGA_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7UMC_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7UMD_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7UME_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7UO0_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7UO1_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7UR5_y.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7URM_v.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7UVT_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7V06_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7V59_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7VTN_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7VW3_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7WB1_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7X34_3.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7XSN_N.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7XW2_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7YOJ_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7YR7_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7ZAP_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7ZPQ_6.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7ZRS_6.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/7ZRZ_ZN1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8AW3_1.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8BD5_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8BTZ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8BU8_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8BWT_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8C8X_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8C98_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8C9C_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8CI5_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8CLR_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8CQ1_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8CTI_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8D49_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8D4A_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8DC2_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8DK7_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8DVS_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8E29_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8E79_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8EVQ_EC.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8EVR_EC.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8EWG_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8EXY_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8FCS_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8FZT_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8G1I_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8G1S_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8GAP_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8H69_5.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8H6E_4A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8H6E_5A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8HUD_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8I0P_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8IDF_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8IEW_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8J72_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8JHP_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8JTJ_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8JTR_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8K8B_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8KAG_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8KAI_I.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8KAM_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8KDA_O.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8KI7_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8OKD_X.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8OMR_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8OPP_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8OPS_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8OPT_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8OST_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8P0B_V.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8P0G_V.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8PM4_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8QO2_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8QO3_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8QO4_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8QO5_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8SA6_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8SCF_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8SCH_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8SFH_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8SFI_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8SFJ_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8SFL_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8SFN_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8SFO_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8T29_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8T2A_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8T2B_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8T2O_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8T2P_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8T2P_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8T2Y_tA.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8T2Y_tP.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8T3E_EC.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8T3F_EC.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8TOC_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8TVZ_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8U3M_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8UAU_R.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8UBE_I.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8UO6_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8UPT_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8URY_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8URY_d.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8UYJ_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8UYL_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8UYP_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8UYS_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8VCI_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8WT6_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8WT6_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8WT7_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8WT7_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8WT8_E.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8WT8_F.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8WUS_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8XCA_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8XCC_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8XPP_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8YE6_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8YGJ_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8Z1F_T.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8Z1G_T.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8Z85_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8Z8N_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/8Z90_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/9CF1_W.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/9CF3_W.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/9DTT_B.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/9ENC_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/9ENE_D.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/9ENF_C.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/9EOW_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/9FO8_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/9FO9_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/9G7C_A.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/9G8O_S2.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/R1107.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/R1108.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/R1116.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/R1117v2.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/R1126.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/R1128.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/R1136.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/R1138.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/R1149.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/R1156.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/R1189.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/MSA/R1190.MSA.fasta  \n",
            "  inflating: /content/stanford_rna_3d_folding/sample_submission.csv  \n",
            "  inflating: /content/stanford_rna_3d_folding/test_sequences.csv  \n",
            "  inflating: /content/stanford_rna_3d_folding/train_labels.csv  \n",
            "  inflating: /content/stanford_rna_3d_folding/train_sequences.csv  \n",
            "  inflating: /content/stanford_rna_3d_folding/validation_labels.csv  \n",
            "  inflating: /content/stanford_rna_3d_folding/validation_sequences.csv  \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "print(os.listdir('/content/'))\n"
      ],
      "metadata": {
        "id": "nGEND6Q8o9fK",
        "outputId": "2b301552-5c35-44a7-ecaa-b5cd67f0dcdf",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['.config', 'required_presets', 'required-presets.zip', 'stanford_rna_3d_folding', 'kaggle.json', 'stanford-rna-3d-folding.zip', 'sample_data']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pip install rdkit\n"
      ],
      "metadata": {
        "id": "ypj0MmZltuKz",
        "outputId": "efb7b4b3-efb6-44ad-b4fa-942e3516ef95",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting rdkit\n",
            "  Downloading rdkit-2024.9.6-cp311-cp311-manylinux_2_28_x86_64.whl.metadata (4.0 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from rdkit) (2.0.2)\n",
            "Requirement already satisfied: Pillow in /usr/local/lib/python3.11/dist-packages (from rdkit) (11.1.0)\n",
            "Downloading rdkit-2024.9.6-cp311-cp311-manylinux_2_28_x86_64.whl (34.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m34.3/34.3 MB\u001b[0m \u001b[31m17.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: rdkit\n",
            "Successfully installed rdkit-2024.9.6\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pip install torch==2.0.0\n"
      ],
      "metadata": {
        "id": "Td5IHXBluATO",
        "outputId": "fa1fdf74-979e-47b0-b767-785c64367953",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting torch==2.0.0\n",
            "  Downloading torch-2.0.0-cp311-cp311-manylinux1_x86_64.whl.metadata (24 kB)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.11/dist-packages (from torch==2.0.0) (3.18.0)\n",
            "Requirement already satisfied: typing-extensions in /usr/local/lib/python3.11/dist-packages (from torch==2.0.0) (4.13.2)\n",
            "Requirement already satisfied: sympy in /usr/local/lib/python3.11/dist-packages (from torch==2.0.0) (1.13.1)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.11/dist-packages (from torch==2.0.0) (3.4.2)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.11/dist-packages (from torch==2.0.0) (3.1.6)\n",
            "Collecting nvidia-cuda-nvrtc-cu11==11.7.99 (from torch==2.0.0)\n",
            "  Downloading nvidia_cuda_nvrtc_cu11-11.7.99-2-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)\n",
            "Collecting nvidia-cuda-runtime-cu11==11.7.99 (from torch==2.0.0)\n",
            "  Downloading nvidia_cuda_runtime_cu11-11.7.99-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)\n",
            "Collecting nvidia-cuda-cupti-cu11==11.7.101 (from torch==2.0.0)\n",
            "  Downloading nvidia_cuda_cupti_cu11-11.7.101-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)\n",
            "Collecting nvidia-cudnn-cu11==8.5.0.96 (from torch==2.0.0)\n",
            "  Downloading nvidia_cudnn_cu11-8.5.0.96-2-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)\n",
            "Collecting nvidia-cublas-cu11==11.10.3.66 (from torch==2.0.0)\n",
            "  Downloading nvidia_cublas_cu11-11.10.3.66-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)\n",
            "Collecting nvidia-cufft-cu11==10.9.0.58 (from torch==2.0.0)\n",
            "  Downloading nvidia_cufft_cu11-10.9.0.58-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)\n",
            "Collecting nvidia-curand-cu11==10.2.10.91 (from torch==2.0.0)\n",
            "  Downloading nvidia_curand_cu11-10.2.10.91-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)\n",
            "Collecting nvidia-cusolver-cu11==11.4.0.1 (from torch==2.0.0)\n",
            "  Downloading nvidia_cusolver_cu11-11.4.0.1-2-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)\n",
            "Collecting nvidia-cusparse-cu11==11.7.4.91 (from torch==2.0.0)\n",
            "  Downloading nvidia_cusparse_cu11-11.7.4.91-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)\n",
            "Collecting nvidia-nccl-cu11==2.14.3 (from torch==2.0.0)\n",
            "  Downloading nvidia_nccl_cu11-2.14.3-py3-none-manylinux1_x86_64.whl.metadata (1.8 kB)\n",
            "Collecting nvidia-nvtx-cu11==11.7.91 (from torch==2.0.0)\n",
            "  Downloading nvidia_nvtx_cu11-11.7.91-py3-none-manylinux1_x86_64.whl.metadata (1.7 kB)\n",
            "Collecting triton==2.0.0 (from torch==2.0.0)\n",
            "  Downloading triton-2.0.0-1-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (1.0 kB)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.11/dist-packages (from nvidia-cublas-cu11==11.10.3.66->torch==2.0.0) (75.2.0)\n",
            "Requirement already satisfied: wheel in /usr/local/lib/python3.11/dist-packages (from nvidia-cublas-cu11==11.10.3.66->torch==2.0.0) (0.45.1)\n",
            "Requirement already satisfied: cmake in /usr/local/lib/python3.11/dist-packages (from triton==2.0.0->torch==2.0.0) (3.31.6)\n",
            "Collecting lit (from triton==2.0.0->torch==2.0.0)\n",
            "  Downloading lit-18.1.8-py3-none-any.whl.metadata (2.5 kB)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2->torch==2.0.0) (3.0.2)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy->torch==2.0.0) (1.3.0)\n",
            "Downloading torch-2.0.0-cp311-cp311-manylinux1_x86_64.whl (619.9 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m619.9/619.9 MB\u001b[0m \u001b[31m3.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cublas_cu11-11.10.3.66-py3-none-manylinux1_x86_64.whl (317.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m317.1/317.1 MB\u001b[0m \u001b[31m4.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cuda_cupti_cu11-11.7.101-py3-none-manylinux1_x86_64.whl (11.8 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m11.8/11.8 MB\u001b[0m \u001b[31m48.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cuda_nvrtc_cu11-11.7.99-2-py3-none-manylinux1_x86_64.whl (21.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m21.0/21.0 MB\u001b[0m \u001b[31m43.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cuda_runtime_cu11-11.7.99-py3-none-manylinux1_x86_64.whl (849 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m849.3/849.3 kB\u001b[0m \u001b[31m44.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cudnn_cu11-8.5.0.96-2-py3-none-manylinux1_x86_64.whl (557.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m557.1/557.1 MB\u001b[0m \u001b[31m1.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cufft_cu11-10.9.0.58-py3-none-manylinux2014_x86_64.whl (168.4 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m168.4/168.4 MB\u001b[0m \u001b[31m6.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_curand_cu11-10.2.10.91-py3-none-manylinux1_x86_64.whl (54.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m54.6/54.6 MB\u001b[0m \u001b[31m12.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cusolver_cu11-11.4.0.1-2-py3-none-manylinux1_x86_64.whl (102.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m102.6/102.6 MB\u001b[0m \u001b[31m8.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cusparse_cu11-11.7.4.91-py3-none-manylinux1_x86_64.whl (173.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m173.2/173.2 MB\u001b[0m \u001b[31m6.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_nccl_cu11-2.14.3-py3-none-manylinux1_x86_64.whl (177.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m177.1/177.1 MB\u001b[0m \u001b[31m6.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_nvtx_cu11-11.7.91-py3-none-manylinux1_x86_64.whl (98 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m98.6/98.6 kB\u001b[0m \u001b[31m9.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading triton-2.0.0-1-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (63.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m63.3/63.3 MB\u001b[0m \u001b[31m12.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading lit-18.1.8-py3-none-any.whl (96 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m96.4/96.4 kB\u001b[0m \u001b[31m9.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: lit, nvidia-nvtx-cu11, nvidia-nccl-cu11, nvidia-cusparse-cu11, nvidia-curand-cu11, nvidia-cufft-cu11, nvidia-cuda-runtime-cu11, nvidia-cuda-nvrtc-cu11, nvidia-cuda-cupti-cu11, nvidia-cublas-cu11, nvidia-cusolver-cu11, nvidia-cudnn-cu11, triton, torch\n",
            "  Attempting uninstall: triton\n",
            "    Found existing installation: triton 2.1.0\n",
            "    Uninstalling triton-2.1.0:\n",
            "      Successfully uninstalled triton-2.1.0\n",
            "  Attempting uninstall: torch\n",
            "    Found existing installation: torch 2.1.1+cu118\n",
            "    Uninstalling torch-2.1.1+cu118:\n",
            "      Successfully uninstalled torch-2.1.1+cu118\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "torchaudio 2.1.1+cu118 requires torch==2.1.1, but you have torch 2.0.0 which is incompatible.\n",
            "torchvision 0.16.1+cu118 requires torch==2.1.1, but you have torch 2.0.0 which is incompatible.\u001b[0m\u001b[31m\n",
            "\u001b[0mSuccessfully installed lit-18.1.8 nvidia-cublas-cu11-11.10.3.66 nvidia-cuda-cupti-cu11-11.7.101 nvidia-cuda-nvrtc-cu11-11.7.99 nvidia-cuda-runtime-cu11-11.7.99 nvidia-cudnn-cu11-8.5.0.96 nvidia-cufft-cu11-10.9.0.58 nvidia-curand-cu11-10.2.10.91 nvidia-cusolver-cu11-11.4.0.1 nvidia-cusparse-cu11-11.7.4.91 nvidia-nccl-cu11-2.14.3 nvidia-nvtx-cu11-11.7.91 torch-2.0.0 triton-2.0.0\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "torch",
                  "torchgen"
                ]
              },
              "id": "3ab727736824435f87b15fd2cc2a868d"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython==1.83\n"
      ],
      "metadata": {
        "id": "SmzOliTUuuvr",
        "outputId": "406dea61-1ac5-4742-f394-59ec21e88878",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython==1.83\n",
            "  Downloading biopython-1.83-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython==1.83) (2.0.2)\n",
            "Downloading biopython-1.83-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
            "\u001b[?25l   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/3.1 MB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K   \u001b[91m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m\u001b[91m╸\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m177.9 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m86.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.83\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install biotite\n"
      ],
      "metadata": {
        "id": "NUl38jCju3YL",
        "outputId": "bb8cb541-5a3e-4d27-82ec-699d60110d60",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biotite\n",
            "  Downloading biotite-1.2.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (7.3 kB)\n",
            "Collecting biotraj<2.0,>=1.0 (from biotite)\n",
            "  Downloading biotraj-1.2.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (32 kB)\n",
            "Requirement already satisfied: msgpack>=0.5.6 in /usr/local/lib/python3.11/dist-packages (from biotite) (1.1.0)\n",
            "Requirement already satisfied: networkx>=2.0 in /usr/local/lib/python3.11/dist-packages (from biotite) (3.4.2)\n",
            "Requirement already satisfied: numpy>=1.25 in /usr/local/lib/python3.11/dist-packages (from biotite) (2.0.2)\n",
            "Requirement already satisfied: packaging>=24.0 in /usr/local/lib/python3.11/dist-packages (from biotite) (24.2)\n",
            "Requirement already satisfied: requests>=2.12 in /usr/local/lib/python3.11/dist-packages (from biotite) (2.32.3)\n",
            "Requirement already satisfied: scipy>=1.13 in /usr/local/lib/python3.11/dist-packages (from biotraj<2.0,>=1.0->biotite) (1.14.1)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite) (2.3.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite) (2025.1.31)\n",
            "Downloading biotite-1.2.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (53.8 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m53.8/53.8 MB\u001b[0m \u001b[31m11.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading biotraj-1.2.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.2/2.2 MB\u001b[0m \u001b[31m40.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biotraj, biotite\n",
            "Successfully installed biotite-1.2.0 biotraj-1.2.2\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!git clone https://github.com/biotite-dev/biotite.git\n"
      ],
      "metadata": {
        "id": "ykhVN8SLvC83",
        "outputId": "55bdb216-f373-4498-fd4d-116002fe39cc",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'biotite'...\n",
            "remote: Enumerating objects: 35914, done.\u001b[K\n",
            "remote: Counting objects: 100% (66/66), done.\u001b[K\n",
            "remote: Compressing objects: 100% (34/34), done.\u001b[K\n",
            "remote: Total 35914 (delta 39), reused 34 (delta 30), pack-reused 35848 (from 2)\u001b[K\n",
            "Receiving objects: 100% (35914/35914), 183.85 MiB | 14.90 MiB/s, done.\n",
            "Resolving deltas: 100% (23965/23965), done.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install .\n"
      ],
      "metadata": {
        "id": "f6ZXIkePvRCT",
        "outputId": "a3ff3c6e-4b54-499e-c6f5-56ea9dc0b7bb",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[31mERROR: Directory '.' is not installable. Neither 'setup.py' nor 'pyproject.toml' found.\u001b[0m\u001b[31m\n",
            "\u001b[0m"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install scikit-learn-extra\n"
      ],
      "metadata": {
        "id": "EJFdw6AVxb4a",
        "outputId": "b9650d55-f0af-428c-f6ba-05175b4c75f6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting scikit-learn-extra\n",
            "  Downloading scikit_learn_extra-0.3.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.6 kB)\n",
            "Requirement already satisfied: numpy>=1.13.3 in /usr/local/lib/python3.11/dist-packages (from scikit-learn-extra) (2.0.2)\n",
            "Requirement already satisfied: scipy>=0.19.1 in /usr/local/lib/python3.11/dist-packages (from scikit-learn-extra) (1.14.1)\n",
            "Requirement already satisfied: scikit-learn>=0.23.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn-extra) (1.6.1)\n",
            "Requirement already satisfied: joblib>=1.2.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn>=0.23.0->scikit-learn-extra) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn>=0.23.0->scikit-learn-extra) (3.6.0)\n",
            "Downloading scikit_learn_extra-0.3.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.1 MB)\n",
            "\u001b[?25l   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/2.1 MB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K   \u001b[91m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m\u001b[91m╸\u001b[0m\u001b[90m━━━━━━━━━━\u001b[0m \u001b[32m1.6/2.1 MB\u001b[0m \u001b[31m65.5 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.1/2.1 MB\u001b[0m \u001b[31m32.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: scikit-learn-extra\n",
            "Successfully installed scikit-learn-extra-0.3.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install protobuf==3.20.2\n"
      ],
      "metadata": {
        "id": "Y5w6dQyXxkpz",
        "outputId": "6bffe400-80ba-4de4-f5e3-92892601b874",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 388
        }
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting protobuf==3.20.2\n",
            "  Downloading protobuf-3.20.2-py2.py3-none-any.whl.metadata (720 bytes)\n",
            "Downloading protobuf-3.20.2-py2.py3-none-any.whl (162 kB)\n",
            "\u001b[?25l   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/162.1 kB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m162.1/162.1 kB\u001b[0m \u001b[31m11.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: protobuf\n",
            "  Attempting uninstall: protobuf\n",
            "    Found existing installation: protobuf 5.29.4\n",
            "    Uninstalling protobuf-5.29.4:\n",
            "      Successfully uninstalled protobuf-5.29.4\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "tensorflow-metadata 1.17.1 requires protobuf<6.0.0,>=4.25.2; python_version >= \"3.11\", but you have protobuf 3.20.2 which is incompatible.\n",
            "tensorflow 2.18.0 requires protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<6.0.0dev,>=3.20.3, but you have protobuf 3.20.2 which is incompatible.\n",
            "grpcio-status 1.71.0 requires protobuf<6.0dev,>=5.26.1, but you have protobuf 3.20.2 which is incompatible.\n",
            "ydf 0.11.0 requires protobuf<6.0.0,>=5.29.1, but you have protobuf 3.20.2 which is incompatible.\u001b[0m\u001b[31m\n",
            "\u001b[0mSuccessfully installed protobuf-3.20.2\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "google"
                ]
              },
              "id": "64437e18dabc4e4db1e2c9f5389832a2"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install --use-deprecated=legacy-resolver protenix\n"
      ],
      "metadata": {
        "id": "dJayeguryKmr",
        "outputId": "55970525-2368-4aa2-8bfb-28bc571e48d1",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting protenix\n",
            "  Downloading protenix-0.4.6-py3-none-any.whl (441 kB)\n",
            "\u001b[?25l     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/441.1 kB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m441.1/441.1 kB\u001b[0m \u001b[31m28.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: PyYaml in /usr/local/lib/python3.11/dist-packages (from protenix) (6.0.2)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.11/dist-packages (from protenix) (1.14.1)\n",
            "Collecting ml_collections (from protenix)\n",
            "  Downloading ml_collections-1.1.0-py3-none-any.whl (76 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m76.7/76.7 kB\u001b[0m \u001b[31m7.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: tqdm in /usr/local/lib/python3.11/dist-packages (from protenix) (4.67.1)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.11/dist-packages (from protenix) (2.2.2)\n",
            "Requirement already satisfied: optree in /usr/local/lib/python3.11/dist-packages (from protenix) (0.15.0)\n",
            "Requirement already satisfied: rdkit in /usr/local/lib/python3.11/dist-packages (from protenix) (2024.9.6)\n",
            "Requirement already satisfied: biopython==1.83 in /usr/local/lib/python3.11/dist-packages (from protenix) (1.83)\n",
            "Collecting modelcif==0.7 (from protenix)\n",
            "  Downloading modelcif-0.7.tar.gz (59 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m59.0/59.0 kB\u001b[0m \u001b[31m5.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting biotite==1.0.1 (from protenix)\n",
            "  Downloading biotite-1.0.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (34.9 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m34.9/34.9 MB\u001b[0m \u001b[31m20.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: scikit-learn in /usr/local/lib/python3.11/dist-packages (from protenix) (1.6.1)\n",
            "Requirement already satisfied: scikit-learn-extra in /usr/local/lib/python3.11/dist-packages (from protenix) (0.3.0)\n",
            "Collecting deepspeed>=0.15.1 (from protenix)\n",
            "  Downloading deepspeed-0.16.7.tar.gz (1.5 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.5/1.5 MB\u001b[0m \u001b[31m62.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: protobuf==3.20.2 in /usr/local/lib/python3.11/dist-packages (from protenix) (3.20.2)\n",
            "Collecting icecream (from protenix)\n",
            "  Downloading icecream-2.1.4-py3-none-any.whl (14 kB)\n",
            "Collecting ipdb (from protenix)\n",
            "  Downloading ipdb-0.13.13-py3-none-any.whl (12 kB)\n",
            "Requirement already satisfied: wandb in /usr/local/lib/python3.11/dist-packages (from protenix) (0.19.9)\n",
            "Collecting numpy==1.26.3 (from protenix)\n",
            "  Downloading numpy-1.26.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.3 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m18.3/18.3 MB\u001b[0m \u001b[31m111.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting matplotlib==3.9.2 (from protenix)\n",
            "  Downloading matplotlib-3.9.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (8.3 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.3/8.3 MB\u001b[0m \u001b[31m108.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: ipywidgets in /usr/local/lib/python3.11/dist-packages (from protenix) (7.7.1)\n",
            "Collecting py3Dmol (from protenix)\n",
            "  Downloading py3Dmol-2.4.2-py2.py3-none-any.whl (7.0 kB)\n",
            "Requirement already satisfied: nvidia-cublas-cu12 in /usr/local/lib/python3.11/dist-packages (from protenix) (12.5.3.2)\n",
            "Collecting torch==2.3.1 (from protenix)\n",
            "  Downloading torch-2.3.1-cp311-cp311-manylinux1_x86_64.whl (779.2 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m779.2/779.2 MB\u001b[0m \u001b[31m782.1 kB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: absl-py in /usr/local/lib/python3.11/dist-packages (from ml_collections->protenix) (1.4.0)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas->protenix) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas->protenix) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas->protenix) (2025.2)\n",
            "Requirement already satisfied: typing-extensions>=4.5.0 in /usr/local/lib/python3.11/dist-packages (from optree->protenix) (4.13.2)\n",
            "Requirement already satisfied: Pillow in /usr/local/lib/python3.11/dist-packages (from rdkit->protenix) (11.1.0)\n",
            "Collecting ihm>=0.34 (from modelcif==0.7->protenix)\n",
            "  Downloading ihm-2.4.tar.gz (386 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m386.3/386.3 kB\u001b[0m \u001b[31m33.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: biotraj<2.0,>=1.0 in /usr/local/lib/python3.11/dist-packages (from biotite==1.0.1->protenix) (1.2.2)\n",
            "Requirement already satisfied: msgpack>=0.5.6 in /usr/local/lib/python3.11/dist-packages (from biotite==1.0.1->protenix) (1.1.0)\n",
            "Requirement already satisfied: networkx>=2.0 in /usr/local/lib/python3.11/dist-packages (from biotite==1.0.1->protenix) (3.4.2)\n",
            "Requirement already satisfied: requests>=2.12 in /usr/local/lib/python3.11/dist-packages (from biotite==1.0.1->protenix) (2.32.3)\n",
            "Requirement already satisfied: joblib>=1.2.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn->protenix) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn->protenix) (3.6.0)\n",
            "Requirement already satisfied: einops in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix) (0.8.1)\n",
            "Collecting hjson (from deepspeed>=0.15.1->protenix)\n",
            "  Downloading hjson-3.1.0-py3-none-any.whl (54 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m54.0/54.0 kB\u001b[0m \u001b[31m5.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting ninja (from deepspeed>=0.15.1->protenix)\n",
            "  Downloading ninja-1.11.1.4-py3-none-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (422 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m422.8/422.8 kB\u001b[0m \u001b[31m37.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix) (24.2)\n",
            "Requirement already satisfied: psutil in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix) (5.9.5)\n",
            "Requirement already satisfied: py-cpuinfo in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix) (9.0.0)\n",
            "Requirement already satisfied: pydantic>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix) (2.11.3)\n",
            "Requirement already satisfied: nvidia-ml-py in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix) (12.570.86)\n",
            "Collecting colorama>=0.3.9 (from icecream->protenix)\n",
            "  Downloading colorama-0.4.6-py2.py3-none-any.whl (25 kB)\n",
            "Requirement already satisfied: pygments>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from icecream->protenix) (2.18.0)\n",
            "Collecting executing>=2.1.0 (from icecream->protenix)\n",
            "  Downloading executing-2.2.0-py2.py3-none-any.whl (26 kB)\n",
            "Collecting asttokens>=2.0.1 (from icecream->protenix)\n",
            "  Downloading asttokens-3.0.0-py3-none-any.whl (26 kB)\n",
            "Requirement already satisfied: ipython>=7.31.1; python_version >= \"3.11\" in /usr/local/lib/python3.11/dist-packages (from ipdb->protenix) (7.34.0)\n",
            "Requirement already satisfied: decorator; python_version >= \"3.11\" in /usr/local/lib/python3.11/dist-packages (from ipdb->protenix) (4.4.2)\n",
            "Requirement already satisfied: click!=8.0.0,>=7.1 in /usr/local/lib/python3.11/dist-packages (from wandb->protenix) (8.1.8)\n",
            "Requirement already satisfied: docker-pycreds>=0.4.0 in /usr/local/lib/python3.11/dist-packages (from wandb->protenix) (0.4.0)\n",
            "Requirement already satisfied: gitpython!=3.1.29,>=1.0.0 in /usr/local/lib/python3.11/dist-packages (from wandb->protenix) (3.1.44)\n",
            "Requirement already satisfied: platformdirs in /usr/local/lib/python3.11/dist-packages (from wandb->protenix) (4.3.7)\n",
            "Requirement already satisfied: sentry-sdk>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from wandb->protenix) (2.26.1)\n",
            "Requirement already satisfied: setproctitle in /usr/local/lib/python3.11/dist-packages (from wandb->protenix) (1.3.5)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.11/dist-packages (from wandb->protenix) (75.2.0)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix) (1.3.2)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix) (4.57.0)\n",
            "Requirement already satisfied: kiwisolver>=1.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix) (1.4.8)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix) (3.2.3)\n",
            "Requirement already satisfied: ipykernel>=4.5.1 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix) (6.17.1)\n",
            "Requirement already satisfied: ipython-genutils~=0.2.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix) (0.2.0)\n",
            "Requirement already satisfied: traitlets>=4.3.1 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix) (5.7.1)\n",
            "Requirement already satisfied: widgetsnbextension~=3.6.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix) (3.6.10)\n",
            "Requirement already satisfied: jupyterlab-widgets>=1.0.0; python_version >= \"3.6\" in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix) (3.0.14)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix) (3.18.0)\n",
            "Requirement already satisfied: sympy in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix) (1.13.1)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix) (3.1.6)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix) (2025.3.2)\n",
            "Collecting nvidia-cuda-nvrtc-cu12==12.1.105; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch==2.3.1->protenix)\n",
            "  Downloading nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m23.7/23.7 MB\u001b[0m \u001b[31m96.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting nvidia-cuda-runtime-cu12==12.1.105; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch==2.3.1->protenix)\n",
            "  Downloading nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m823.6/823.6 kB\u001b[0m \u001b[31m56.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting nvidia-cuda-cupti-cu12==12.1.105; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch==2.3.1->protenix)\n",
            "  Downloading nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m14.1/14.1 MB\u001b[0m \u001b[31m111.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting nvidia-cudnn-cu12==8.9.2.26; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch==2.3.1->protenix)\n",
            "  Downloading nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m731.7/731.7 MB\u001b[0m \u001b[31m2.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting nvidia-cufft-cu12==11.0.2.54; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch==2.3.1->protenix)\n",
            "  Downloading nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m121.6/121.6 MB\u001b[0m \u001b[31m7.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting nvidia-curand-cu12==10.3.2.106; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch==2.3.1->protenix)\n",
            "  Downloading nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.5/56.5 MB\u001b[0m \u001b[31m13.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting nvidia-cusolver-cu12==11.4.5.107; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch==2.3.1->protenix)\n",
            "  Downloading nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m124.2/124.2 MB\u001b[0m \u001b[31m7.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting nvidia-cusparse-cu12==12.1.0.106; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch==2.3.1->protenix)\n",
            "  Downloading nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m196.0/196.0 MB\u001b[0m \u001b[31m5.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting nvidia-nccl-cu12==2.20.5; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch==2.3.1->protenix)\n",
            "  Downloading nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl (176.2 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m176.2/176.2 MB\u001b[0m \u001b[31m5.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting nvidia-nvtx-cu12==12.1.105; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch==2.3.1->protenix)\n",
            "  Downloading nvidia_nvtx_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (99 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m99.1/99.1 kB\u001b[0m \u001b[31m11.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting triton==2.3.1; platform_system == \"Linux\" and platform_machine == \"x86_64\" and python_version < \"3.12\" (from torch==2.3.1->protenix)\n",
            "  Downloading triton-2.3.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (168.1 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m168.1/168.1 MB\u001b[0m \u001b[31m6.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas->protenix) (1.17.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite==1.0.1->protenix) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite==1.0.1->protenix) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite==1.0.1->protenix) (2.3.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite==1.0.1->protenix) (2025.1.31)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/lib/python3.11/dist-packages (from pydantic>=2.0.0->deepspeed>=0.15.1->protenix) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.33.1 in /usr/local/lib/python3.11/dist-packages (from pydantic>=2.0.0->deepspeed>=0.15.1->protenix) (2.33.1)\n",
            "Requirement already satisfied: typing-inspection>=0.4.0 in /usr/local/lib/python3.11/dist-packages (from pydantic>=2.0.0->deepspeed>=0.15.1->protenix) (0.4.0)\n",
            "Collecting jedi>=0.16 (from ipython>=7.31.1; python_version >= \"3.11\"->ipdb->protenix)\n",
            "  Downloading jedi-0.19.2-py2.py3-none-any.whl (1.6 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m82.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: pickleshare in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1; python_version >= \"3.11\"->ipdb->protenix) (0.7.5)\n",
            "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1; python_version >= \"3.11\"->ipdb->protenix) (3.0.51)\n",
            "Requirement already satisfied: backcall in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1; python_version >= \"3.11\"->ipdb->protenix) (0.2.0)\n",
            "Requirement already satisfied: matplotlib-inline in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1; python_version >= \"3.11\"->ipdb->protenix) (0.1.7)\n",
            "Requirement already satisfied: pexpect>4.3; sys_platform != \"win32\" in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1; python_version >= \"3.11\"->ipdb->protenix) (4.9.0)\n",
            "Requirement already satisfied: gitdb<5,>=4.0.1 in /usr/local/lib/python3.11/dist-packages (from gitpython!=3.1.29,>=1.0.0->wandb->protenix) (4.0.12)\n",
            "Requirement already satisfied: debugpy>=1.0 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix) (1.8.0)\n",
            "Requirement already satisfied: jupyter-client>=6.1.12 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix) (6.1.12)\n",
            "Requirement already satisfied: nest-asyncio in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix) (1.6.0)\n",
            "Requirement already satisfied: pyzmq>=17 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix) (24.0.1)\n",
            "Requirement already satisfied: tornado>=6.1 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix) (6.4.2)\n",
            "Requirement already satisfied: notebook>=4.4.1 in /usr/local/lib/python3.11/dist-packages (from widgetsnbextension~=3.6.0->ipywidgets->protenix) (6.5.7)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy->torch==2.3.1->protenix) (1.3.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2->torch==2.3.1->protenix) (3.0.2)\n",
            "Requirement already satisfied: nvidia-nvjitlink-cu12 in /usr/local/lib/python3.11/dist-packages (from nvidia-cusolver-cu12==11.4.5.107; platform_system == \"Linux\" and platform_machine == \"x86_64\"->torch==2.3.1->protenix) (12.5.82)\n",
            "Requirement already satisfied: parso<0.9.0,>=0.8.4 in /usr/local/lib/python3.11/dist-packages (from jedi>=0.16->ipython>=7.31.1; python_version >= \"3.11\"->ipdb->protenix) (0.8.4)\n",
            "Requirement already satisfied: wcwidth in /usr/local/lib/python3.11/dist-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython>=7.31.1; python_version >= \"3.11\"->ipdb->protenix) (0.2.13)\n",
            "Requirement already satisfied: ptyprocess>=0.5 in /usr/local/lib/python3.11/dist-packages (from pexpect>4.3; sys_platform != \"win32\"->ipython>=7.31.1; python_version >= \"3.11\"->ipdb->protenix) (0.7.0)\n",
            "Requirement already satisfied: smmap<6,>=3.0.1 in /usr/local/lib/python3.11/dist-packages (from gitdb<5,>=4.0.1->gitpython!=3.1.29,>=1.0.0->wandb->protenix) (5.0.2)\n",
            "Requirement already satisfied: jupyter-core>=4.6.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets->protenix) (5.7.2)\n",
            "Requirement already satisfied: argon2-cffi in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (23.1.0)\n",
            "Requirement already satisfied: nbformat in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (5.10.4)\n",
            "Requirement already satisfied: nbconvert>=5 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (7.16.6)\n",
            "Requirement already satisfied: Send2Trash>=1.8.0 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (1.8.3)\n",
            "Requirement already satisfied: terminado>=0.8.3 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (0.18.1)\n",
            "Requirement already satisfied: prometheus-client in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (0.21.1)\n",
            "Requirement already satisfied: nbclassic>=0.4.7 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (1.2.0)\n",
            "Requirement already satisfied: argon2-cffi-bindings in /usr/local/lib/python3.11/dist-packages (from argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (21.2.0)\n",
            "Requirement already satisfied: fastjsonschema>=2.15 in /usr/local/lib/python3.11/dist-packages (from nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (2.21.1)\n",
            "Requirement already satisfied: jsonschema>=2.6 in /usr/local/lib/python3.11/dist-packages (from nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (4.23.0)\n",
            "Requirement already satisfied: beautifulsoup4 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (4.13.4)\n",
            "Requirement already satisfied: bleach[css]!=5.0.0 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (6.2.0)\n",
            "Requirement already satisfied: defusedxml in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (0.7.1)\n",
            "Requirement already satisfied: jupyterlab-pygments in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (0.3.0)\n",
            "Requirement already satisfied: mistune<4,>=2.0.3 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (3.1.3)\n",
            "Requirement already satisfied: nbclient>=0.5.0 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (0.10.2)\n",
            "Requirement already satisfied: pandocfilters>=1.4.1 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (1.5.1)\n",
            "Requirement already satisfied: notebook-shim>=0.2.3 in /usr/local/lib/python3.11/dist-packages (from nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (0.2.4)\n",
            "Requirement already satisfied: cffi>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from argon2-cffi-bindings->argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (1.17.1)\n",
            "Requirement already satisfied: attrs>=22.2.0 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (25.3.0)\n",
            "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (2024.10.1)\n",
            "Requirement already satisfied: referencing>=0.28.4 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (0.36.2)\n",
            "Requirement already satisfied: rpds-py>=0.7.1 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (0.24.0)\n",
            "Requirement already satisfied: soupsieve>1.2 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (2.6)\n",
            "Requirement already satisfied: webencodings in /usr/local/lib/python3.11/dist-packages (from bleach[css]!=5.0.0->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (0.5.1)\n",
            "Requirement already satisfied: tinycss2<1.5,>=1.1.0; extra == \"css\" in /usr/local/lib/python3.11/dist-packages (from bleach[css]!=5.0.0->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (1.4.0)\n",
            "Requirement already satisfied: jupyter-server<3,>=1.8 in /usr/local/lib/python3.11/dist-packages (from notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (1.16.0)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.11/dist-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (2.22)\n",
            "Requirement already satisfied: anyio>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (4.9.0)\n",
            "Requirement already satisfied: websocket-client in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (1.8.0)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.11/dist-packages (from anyio>=3.1.0->jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix) (1.3.1)\n",
            "Building wheels for collected packages: modelcif, deepspeed, ihm\n",
            "  Building wheel for modelcif (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for modelcif: filename=modelcif-0.7-py3-none-any.whl size=36633 sha256=4fe3ffdbb274cfbeb380b5040a8d1b59649c70bdf5c146fd59d3f7519a081d90\n",
            "  Stored in directory: /root/.cache/pip/wheels/4b/ef/49/a555e3e4e72da8f450d3bc240d3e5016c59c31228a148ffad0\n",
            "  Building wheel for deepspeed (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for deepspeed: filename=deepspeed-0.16.7-py3-none-any.whl size=1642789 sha256=90cf797cf1f70da0f5902e3d8a156930ef150fdc2fe124db6ce338aa55b80ce6\n",
            "  Stored in directory: /root/.cache/pip/wheels/42/e7/1a/2106f7197cc13e09c68f1b4f55f7e5117a985e726378968970\n",
            "  Building wheel for ihm (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for ihm: filename=ihm-2.4-cp311-cp311-linux_x86_64.whl size=380897 sha256=3f11d61b178460ba333fbe0e277eef6000d1fb3a51424f416e20b28e9b4107b5\n",
            "  Stored in directory: /root/.cache/pip/wheels/fc/ac/88/0df437ad2118b5b62695d991aeb3a373224fd7fbc215f01efe\n",
            "Successfully built modelcif deepspeed ihm\n",
            "Installing collected packages: ml_collections, ihm, modelcif, numpy, biotite, hjson, ninja, nvidia-cuda-nvrtc-cu12, nvidia-cuda-runtime-cu12, nvidia-cuda-cupti-cu12, nvidia-cudnn-cu12, nvidia-cufft-cu12, nvidia-curand-cu12, nvidia-cusparse-cu12, nvidia-cusolver-cu12, nvidia-nccl-cu12, nvidia-nvtx-cu12, triton, torch, deepspeed, colorama, executing, asttokens, icecream, ipdb, matplotlib, py3Dmol, protenix, jedi\n",
            "  Attempting uninstall: numpy\n",
            "    Found existing installation: numpy 2.0.2\n",
            "    Uninstalling numpy-2.0.2:\n",
            "      Successfully uninstalled numpy-2.0.2\n",
            "  Attempting uninstall: biotite\n",
            "    Found existing installation: biotite 1.2.0\n",
            "    Uninstalling biotite-1.2.0:\n",
            "      Successfully uninstalled biotite-1.2.0\n",
            "  Attempting uninstall: nvidia-cuda-nvrtc-cu12\n",
            "    Found existing installation: nvidia-cuda-nvrtc-cu12 12.5.82\n",
            "    Uninstalling nvidia-cuda-nvrtc-cu12-12.5.82:\n",
            "      Successfully uninstalled nvidia-cuda-nvrtc-cu12-12.5.82\n",
            "  Attempting uninstall: nvidia-cuda-runtime-cu12\n",
            "    Found existing installation: nvidia-cuda-runtime-cu12 12.5.82\n",
            "    Uninstalling nvidia-cuda-runtime-cu12-12.5.82:\n",
            "      Successfully uninstalled nvidia-cuda-runtime-cu12-12.5.82\n",
            "  Attempting uninstall: nvidia-cuda-cupti-cu12\n",
            "    Found existing installation: nvidia-cuda-cupti-cu12 12.5.82\n",
            "    Uninstalling nvidia-cuda-cupti-cu12-12.5.82:\n",
            "      Successfully uninstalled nvidia-cuda-cupti-cu12-12.5.82\n",
            "  Attempting uninstall: nvidia-cudnn-cu12\n",
            "    Found existing installation: nvidia-cudnn-cu12 9.3.0.75\n",
            "    Uninstalling nvidia-cudnn-cu12-9.3.0.75:\n",
            "      Successfully uninstalled nvidia-cudnn-cu12-9.3.0.75\n",
            "  Attempting uninstall: nvidia-cufft-cu12\n",
            "    Found existing installation: nvidia-cufft-cu12 11.2.3.61\n",
            "    Uninstalling nvidia-cufft-cu12-11.2.3.61:\n",
            "      Successfully uninstalled nvidia-cufft-cu12-11.2.3.61\n",
            "  Attempting uninstall: nvidia-curand-cu12\n",
            "    Found existing installation: nvidia-curand-cu12 10.3.6.82\n",
            "    Uninstalling nvidia-curand-cu12-10.3.6.82:\n",
            "      Successfully uninstalled nvidia-curand-cu12-10.3.6.82\n",
            "  Attempting uninstall: nvidia-cusparse-cu12\n",
            "    Found existing installation: nvidia-cusparse-cu12 12.5.1.3\n",
            "    Uninstalling nvidia-cusparse-cu12-12.5.1.3:\n",
            "      Successfully uninstalled nvidia-cusparse-cu12-12.5.1.3\n",
            "  Attempting uninstall: nvidia-cusolver-cu12\n",
            "    Found existing installation: nvidia-cusolver-cu12 11.6.3.83\n",
            "    Uninstalling nvidia-cusolver-cu12-11.6.3.83:\n",
            "      Successfully uninstalled nvidia-cusolver-cu12-11.6.3.83\n",
            "  Attempting uninstall: nvidia-nccl-cu12\n",
            "    Found existing installation: nvidia-nccl-cu12 2.21.5\n",
            "    Uninstalling nvidia-nccl-cu12-2.21.5:\n",
            "      Successfully uninstalled nvidia-nccl-cu12-2.21.5\n",
            "  Attempting uninstall: nvidia-nvtx-cu12\n",
            "    Found existing installation: nvidia-nvtx-cu12 12.4.127\n",
            "    Uninstalling nvidia-nvtx-cu12-12.4.127:\n",
            "      Successfully uninstalled nvidia-nvtx-cu12-12.4.127\n",
            "  Attempting uninstall: triton\n",
            "    Found existing installation: triton 2.0.0\n",
            "    Uninstalling triton-2.0.0:\n",
            "      Successfully uninstalled triton-2.0.0\n",
            "  Attempting uninstall: torch\n",
            "    Found existing installation: torch 2.0.0\n",
            "    Uninstalling torch-2.0.0:\n",
            "      Successfully uninstalled torch-2.0.0\n",
            "  Attempting uninstall: matplotlib\n",
            "    Found existing installation: matplotlib 3.10.0\n",
            "    Uninstalling matplotlib-3.10.0:\n",
            "      Successfully uninstalled matplotlib-3.10.0\n",
            "\u001b[31mERROR: pip's legacy dependency resolver does not consider dependency conflicts when selecting packages. This behaviour is the source of the following dependency conflicts.\n",
            "torchaudio 2.1.1+cu118 requires torch==2.1.1, but you'll have torch 2.3.1 which is incompatible.\n",
            "torchvision 0.16.1+cu118 requires torch==2.1.1, but you'll have torch 2.3.1 which is incompatible.\n",
            "torch 2.3.1 requires nvidia-cublas-cu12==12.1.3.1; platform_system == \"Linux\" and platform_machine == \"x86_64\", but you'll have nvidia-cublas-cu12 12.5.3.2 which is incompatible.\n",
            "tensorflow 2.18.0 requires protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<6.0.0dev,>=3.20.3, but you'll have protobuf 3.20.2 which is incompatible.\n",
            "ydf 0.11.0 requires protobuf<6.0.0,>=5.29.1, but you'll have protobuf 3.20.2 which is incompatible.\n",
            "thinc 8.3.6 requires numpy<3.0.0,>=2.0.0, but you'll have numpy 1.26.3 which is incompatible.\u001b[0m\u001b[31m\n",
            "\u001b[0mSuccessfully installed asttokens-3.0.0 biotite-1.0.1 colorama-0.4.6 deepspeed-0.16.7 executing-2.2.0 hjson-3.1.0 icecream-2.1.4 ihm-2.4 ipdb-0.13.13 jedi-0.19.2 matplotlib-3.9.2 ml_collections-1.1.0 modelcif-0.7 ninja-1.11.1.4 numpy-1.26.3 nvidia-cuda-cupti-cu12-12.1.105 nvidia-cuda-nvrtc-cu12-12.1.105 nvidia-cuda-runtime-cu12-12.1.105 nvidia-cudnn-cu12-8.9.2.26 nvidia-cufft-cu12-11.0.2.54 nvidia-curand-cu12-10.3.2.106 nvidia-cusolver-cu12-11.4.5.107 nvidia-cusparse-cu12-12.1.0.106 nvidia-nccl-cu12-2.20.5 nvidia-nvtx-cu12-12.1.105 protenix-0.4.6 py3Dmol-2.4.2 torch-2.3.1 triton-2.3.1\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip uninstall protenix\n",
        "!pip install \"protenix<0.4.6\"\n",
        "\n",
        "!pip install torch==2.1.1 torchvision==0.16.1 torchaudio==2.1.1 --index-url https://download.pytorch.org/whl/cpu\n"
      ],
      "metadata": {
        "id": "RWJ_te88zajr",
        "outputId": "15ebfa11-225b-4d62-bfb8-5d5e8a55d3f1",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Found existing installation: protenix 0.4.6\n",
            "Uninstalling protenix-0.4.6:\n",
            "  Would remove:\n",
            "    /usr/local/bin/protenix\n",
            "    /usr/local/lib/python3.11/dist-packages/configs/*\n",
            "    /usr/local/lib/python3.11/dist-packages/protenix-0.4.6.dist-info/*\n",
            "    /usr/local/lib/python3.11/dist-packages/protenix/*\n",
            "    /usr/local/lib/python3.11/dist-packages/runner/*\n",
            "    /usr/local/lib/python3.11/dist-packages/scripts/*\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/*\n",
            "  Would not remove (might be manually added):\n",
            "    /usr/local/lib/python3.11/dist-packages/scripts/readme-gen/readme_gen.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/check_ic_formatting.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/conversion_test.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/database_test.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/graph_entities_test.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/graph_server_test.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/install_test_import.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/magics_test.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/node_expansion_test.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/sample_notebook_test.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/schema_manager_test.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/server_test.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/test_basics.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/test_icecream.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/test_install.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/test_notebook.json\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/tox-plugin/setup.py\n",
            "    /usr/local/lib/python3.11/dist-packages/tests/tox-plugin/tox.ini\n",
            "Proceed (Y/n)? Y\n",
            "  Successfully uninstalled protenix-0.4.6\n",
            "Collecting protenix<0.4.6\n",
            "  Downloading protenix-0.4.5-py3-none-any.whl.metadata (1.1 kB)\n",
            "Requirement already satisfied: PyYaml in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (6.0.2)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (1.14.1)\n",
            "Requirement already satisfied: ml_collections in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (1.1.0)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (4.67.1)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (2.2.2)\n",
            "Requirement already satisfied: optree in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (0.15.0)\n",
            "Requirement already satisfied: rdkit in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (2024.9.6)\n",
            "Requirement already satisfied: biopython==1.83 in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (1.83)\n",
            "Requirement already satisfied: modelcif==0.7 in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (0.7)\n",
            "Requirement already satisfied: biotite==1.0.1 in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (1.0.1)\n",
            "Requirement already satisfied: scikit-learn in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (1.6.1)\n",
            "Requirement already satisfied: scikit-learn-extra in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (0.3.0)\n",
            "Requirement already satisfied: deepspeed>=0.15.1 in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (0.16.7)\n",
            "Requirement already satisfied: protobuf==3.20.2 in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (3.20.2)\n",
            "Requirement already satisfied: icecream in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (2.1.4)\n",
            "Requirement already satisfied: ipdb in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (0.13.13)\n",
            "Requirement already satisfied: wandb in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (0.19.9)\n",
            "Requirement already satisfied: numpy==1.26.3 in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (1.26.3)\n",
            "Requirement already satisfied: matplotlib==3.9.2 in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (3.9.2)\n",
            "Requirement already satisfied: ipywidgets in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (7.7.1)\n",
            "Requirement already satisfied: py3Dmol in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (2.4.2)\n",
            "Requirement already satisfied: nvidia-cublas-cu12 in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (12.5.3.2)\n",
            "Requirement already satisfied: torch==2.3.1 in /usr/local/lib/python3.11/dist-packages (from protenix<0.4.6) (2.3.1)\n",
            "Requirement already satisfied: biotraj<2.0,>=1.0 in /usr/local/lib/python3.11/dist-packages (from biotite==1.0.1->protenix<0.4.6) (1.2.2)\n",
            "Requirement already satisfied: msgpack>=0.5.6 in /usr/local/lib/python3.11/dist-packages (from biotite==1.0.1->protenix<0.4.6) (1.1.0)\n",
            "Requirement already satisfied: networkx>=2.0 in /usr/local/lib/python3.11/dist-packages (from biotite==1.0.1->protenix<0.4.6) (3.4.2)\n",
            "Requirement already satisfied: requests>=2.12 in /usr/local/lib/python3.11/dist-packages (from biotite==1.0.1->protenix<0.4.6) (2.32.3)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix<0.4.6) (1.3.2)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix<0.4.6) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix<0.4.6) (4.57.0)\n",
            "Requirement already satisfied: kiwisolver>=1.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix<0.4.6) (1.4.8)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix<0.4.6) (24.2)\n",
            "Requirement already satisfied: pillow>=8 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix<0.4.6) (11.1.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix<0.4.6) (3.2.3)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix<0.4.6) (2.8.2)\n",
            "Requirement already satisfied: ihm>=0.34 in /usr/local/lib/python3.11/dist-packages (from modelcif==0.7->protenix<0.4.6) (2.4)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (3.18.0)\n",
            "Requirement already satisfied: typing-extensions>=4.8.0 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (4.13.2)\n",
            "Requirement already satisfied: sympy in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (1.13.1)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (3.1.6)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (2025.3.2)\n",
            "Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.1.105 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (12.1.105)\n",
            "Requirement already satisfied: nvidia-cuda-runtime-cu12==12.1.105 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (12.1.105)\n",
            "Requirement already satisfied: nvidia-cuda-cupti-cu12==12.1.105 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (12.1.105)\n",
            "Requirement already satisfied: nvidia-cudnn-cu12==8.9.2.26 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (8.9.2.26)\n",
            "Collecting nvidia-cublas-cu12 (from protenix<0.4.6)\n",
            "  Downloading nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)\n",
            "Requirement already satisfied: nvidia-cufft-cu12==11.0.2.54 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (11.0.2.54)\n",
            "Requirement already satisfied: nvidia-curand-cu12==10.3.2.106 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (10.3.2.106)\n",
            "Requirement already satisfied: nvidia-cusolver-cu12==11.4.5.107 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (11.4.5.107)\n",
            "Requirement already satisfied: nvidia-cusparse-cu12==12.1.0.106 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (12.1.0.106)\n",
            "Requirement already satisfied: nvidia-nccl-cu12==2.20.5 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (2.20.5)\n",
            "Requirement already satisfied: nvidia-nvtx-cu12==12.1.105 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (12.1.105)\n",
            "Requirement already satisfied: triton==2.3.1 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix<0.4.6) (2.3.1)\n",
            "Requirement already satisfied: nvidia-nvjitlink-cu12 in /usr/local/lib/python3.11/dist-packages (from nvidia-cusolver-cu12==11.4.5.107->torch==2.3.1->protenix<0.4.6) (12.5.82)\n",
            "Requirement already satisfied: einops in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix<0.4.6) (0.8.1)\n",
            "Requirement already satisfied: hjson in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix<0.4.6) (3.1.0)\n",
            "Requirement already satisfied: ninja in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix<0.4.6) (1.11.1.4)\n",
            "Requirement already satisfied: psutil in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix<0.4.6) (5.9.5)\n",
            "Requirement already satisfied: py-cpuinfo in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix<0.4.6) (9.0.0)\n",
            "Requirement already satisfied: pydantic>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix<0.4.6) (2.11.3)\n",
            "Requirement already satisfied: nvidia-ml-py in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix<0.4.6) (12.570.86)\n",
            "Requirement already satisfied: colorama>=0.3.9 in /usr/local/lib/python3.11/dist-packages (from icecream->protenix<0.4.6) (0.4.6)\n",
            "Requirement already satisfied: pygments>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from icecream->protenix<0.4.6) (2.18.0)\n",
            "Requirement already satisfied: executing>=2.1.0 in /usr/local/lib/python3.11/dist-packages (from icecream->protenix<0.4.6) (2.2.0)\n",
            "Requirement already satisfied: asttokens>=2.0.1 in /usr/local/lib/python3.11/dist-packages (from icecream->protenix<0.4.6) (3.0.0)\n",
            "Requirement already satisfied: ipython>=7.31.1 in /usr/local/lib/python3.11/dist-packages (from ipdb->protenix<0.4.6) (7.34.0)\n",
            "Requirement already satisfied: decorator in /usr/local/lib/python3.11/dist-packages (from ipdb->protenix<0.4.6) (4.4.2)\n",
            "Requirement already satisfied: ipykernel>=4.5.1 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix<0.4.6) (6.17.1)\n",
            "Requirement already satisfied: ipython-genutils~=0.2.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix<0.4.6) (0.2.0)\n",
            "Requirement already satisfied: traitlets>=4.3.1 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix<0.4.6) (5.7.1)\n",
            "Requirement already satisfied: widgetsnbextension~=3.6.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix<0.4.6) (3.6.10)\n",
            "Requirement already satisfied: jupyterlab-widgets>=1.0.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix<0.4.6) (3.0.14)\n",
            "Requirement already satisfied: absl-py in /usr/local/lib/python3.11/dist-packages (from ml_collections->protenix<0.4.6) (1.4.0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas->protenix<0.4.6) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas->protenix<0.4.6) (2025.2)\n",
            "Requirement already satisfied: joblib>=1.2.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn->protenix<0.4.6) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn->protenix<0.4.6) (3.6.0)\n",
            "Requirement already satisfied: click!=8.0.0,>=7.1 in /usr/local/lib/python3.11/dist-packages (from wandb->protenix<0.4.6) (8.1.8)\n",
            "Requirement already satisfied: docker-pycreds>=0.4.0 in /usr/local/lib/python3.11/dist-packages (from wandb->protenix<0.4.6) (0.4.0)\n",
            "Requirement already satisfied: gitpython!=3.1.29,>=1.0.0 in /usr/local/lib/python3.11/dist-packages (from wandb->protenix<0.4.6) (3.1.44)\n",
            "Requirement already satisfied: platformdirs in /usr/local/lib/python3.11/dist-packages (from wandb->protenix<0.4.6) (4.3.7)\n",
            "Requirement already satisfied: sentry-sdk>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from wandb->protenix<0.4.6) (2.26.1)\n",
            "Requirement already satisfied: setproctitle in /usr/local/lib/python3.11/dist-packages (from wandb->protenix<0.4.6) (1.3.5)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.11/dist-packages (from wandb->protenix<0.4.6) (75.2.0)\n",
            "Requirement already satisfied: six>=1.4.0 in /usr/local/lib/python3.11/dist-packages (from docker-pycreds>=0.4.0->wandb->protenix<0.4.6) (1.17.0)\n",
            "Requirement already satisfied: gitdb<5,>=4.0.1 in /usr/local/lib/python3.11/dist-packages (from gitpython!=3.1.29,>=1.0.0->wandb->protenix<0.4.6) (4.0.12)\n",
            "Requirement already satisfied: debugpy>=1.0 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix<0.4.6) (1.8.0)\n",
            "Requirement already satisfied: jupyter-client>=6.1.12 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix<0.4.6) (6.1.12)\n",
            "Requirement already satisfied: matplotlib-inline>=0.1 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix<0.4.6) (0.1.7)\n",
            "Requirement already satisfied: nest-asyncio in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix<0.4.6) (1.6.0)\n",
            "Requirement already satisfied: pyzmq>=17 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix<0.4.6) (24.0.1)\n",
            "Requirement already satisfied: tornado>=6.1 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix<0.4.6) (6.4.2)\n",
            "Requirement already satisfied: jedi>=0.16 in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1->ipdb->protenix<0.4.6) (0.19.2)\n",
            "Requirement already satisfied: pickleshare in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1->ipdb->protenix<0.4.6) (0.7.5)\n",
            "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1->ipdb->protenix<0.4.6) (3.0.51)\n",
            "Requirement already satisfied: backcall in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1->ipdb->protenix<0.4.6) (0.2.0)\n",
            "Requirement already satisfied: pexpect>4.3 in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1->ipdb->protenix<0.4.6) (4.9.0)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/lib/python3.11/dist-packages (from pydantic>=2.0.0->deepspeed>=0.15.1->protenix<0.4.6) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.33.1 in /usr/local/lib/python3.11/dist-packages (from pydantic>=2.0.0->deepspeed>=0.15.1->protenix<0.4.6) (2.33.1)\n",
            "Requirement already satisfied: typing-inspection>=0.4.0 in /usr/local/lib/python3.11/dist-packages (from pydantic>=2.0.0->deepspeed>=0.15.1->protenix<0.4.6) (0.4.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite==1.0.1->protenix<0.4.6) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite==1.0.1->protenix<0.4.6) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite==1.0.1->protenix<0.4.6) (2.3.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite==1.0.1->protenix<0.4.6) (2025.1.31)\n",
            "Requirement already satisfied: notebook>=4.4.1 in /usr/local/lib/python3.11/dist-packages (from widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (6.5.7)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2->torch==2.3.1->protenix<0.4.6) (3.0.2)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy->torch==2.3.1->protenix<0.4.6) (1.3.0)\n",
            "Requirement already satisfied: smmap<6,>=3.0.1 in /usr/local/lib/python3.11/dist-packages (from gitdb<5,>=4.0.1->gitpython!=3.1.29,>=1.0.0->wandb->protenix<0.4.6) (5.0.2)\n",
            "Requirement already satisfied: parso<0.9.0,>=0.8.4 in /usr/local/lib/python3.11/dist-packages (from jedi>=0.16->ipython>=7.31.1->ipdb->protenix<0.4.6) (0.8.4)\n",
            "Requirement already satisfied: jupyter-core>=4.6.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets->protenix<0.4.6) (5.7.2)\n",
            "Requirement already satisfied: argon2-cffi in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (23.1.0)\n",
            "Requirement already satisfied: nbformat in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (5.10.4)\n",
            "Requirement already satisfied: nbconvert>=5 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (7.16.6)\n",
            "Requirement already satisfied: Send2Trash>=1.8.0 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (1.8.3)\n",
            "Requirement already satisfied: terminado>=0.8.3 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (0.18.1)\n",
            "Requirement already satisfied: prometheus-client in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (0.21.1)\n",
            "Requirement already satisfied: nbclassic>=0.4.7 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (1.2.0)\n",
            "Requirement already satisfied: ptyprocess>=0.5 in /usr/local/lib/python3.11/dist-packages (from pexpect>4.3->ipython>=7.31.1->ipdb->protenix<0.4.6) (0.7.0)\n",
            "Requirement already satisfied: wcwidth in /usr/local/lib/python3.11/dist-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython>=7.31.1->ipdb->protenix<0.4.6) (0.2.13)\n",
            "Requirement already satisfied: notebook-shim>=0.2.3 in /usr/local/lib/python3.11/dist-packages (from nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (0.2.4)\n",
            "Requirement already satisfied: beautifulsoup4 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (4.13.4)\n",
            "Requirement already satisfied: bleach!=5.0.0 in /usr/local/lib/python3.11/dist-packages (from bleach[css]!=5.0.0->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (6.2.0)\n",
            "Requirement already satisfied: defusedxml in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (0.7.1)\n",
            "Requirement already satisfied: jupyterlab-pygments in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (0.3.0)\n",
            "Requirement already satisfied: mistune<4,>=2.0.3 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (3.1.3)\n",
            "Requirement already satisfied: nbclient>=0.5.0 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (0.10.2)\n",
            "Requirement already satisfied: pandocfilters>=1.4.1 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (1.5.1)\n",
            "Requirement already satisfied: fastjsonschema>=2.15 in /usr/local/lib/python3.11/dist-packages (from nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (2.21.1)\n",
            "Requirement already satisfied: jsonschema>=2.6 in /usr/local/lib/python3.11/dist-packages (from nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (4.23.0)\n",
            "Requirement already satisfied: argon2-cffi-bindings in /usr/local/lib/python3.11/dist-packages (from argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (21.2.0)\n",
            "Requirement already satisfied: webencodings in /usr/local/lib/python3.11/dist-packages (from bleach!=5.0.0->bleach[css]!=5.0.0->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (0.5.1)\n",
            "Requirement already satisfied: tinycss2<1.5,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from bleach[css]!=5.0.0->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (1.4.0)\n",
            "Requirement already satisfied: attrs>=22.2.0 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (25.3.0)\n",
            "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (2024.10.1)\n",
            "Requirement already satisfied: referencing>=0.28.4 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (0.36.2)\n",
            "Requirement already satisfied: rpds-py>=0.7.1 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (0.24.0)\n",
            "Requirement already satisfied: jupyter-server<3,>=1.8 in /usr/local/lib/python3.11/dist-packages (from notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (1.16.0)\n",
            "Requirement already satisfied: cffi>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from argon2-cffi-bindings->argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (1.17.1)\n",
            "Requirement already satisfied: soupsieve>1.2 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (2.6)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.11/dist-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (2.22)\n",
            "Requirement already satisfied: anyio>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (4.9.0)\n",
            "Requirement already satisfied: websocket-client in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (1.8.0)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.11/dist-packages (from anyio>=3.1.0->jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix<0.4.6) (1.3.1)\n",
            "Downloading protenix-0.4.5-py3-none-any.whl (420 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m420.7/420.7 kB\u001b[0m \u001b[31m23.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m410.6/410.6 MB\u001b[0m \u001b[31m3.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: nvidia-cublas-cu12, protenix\n",
            "  Attempting uninstall: nvidia-cublas-cu12\n",
            "    Found existing installation: nvidia-cublas-cu12 12.5.3.2\n",
            "    Uninstalling nvidia-cublas-cu12-12.5.3.2:\n",
            "      Successfully uninstalled nvidia-cublas-cu12-12.5.3.2\n",
            "Successfully installed nvidia-cublas-cu12-12.1.3.1 protenix-0.4.5\n",
            "Looking in indexes: https://download.pytorch.org/whl/cpu\n",
            "Collecting torch==2.1.1\n",
            "  Downloading https://download.pytorch.org/whl/cpu/torch-2.1.1%2Bcpu-cp311-cp311-linux_x86_64.whl (184.9 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m184.9/184.9 MB\u001b[0m \u001b[31m6.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: torchvision==0.16.1 in /usr/local/lib/python3.11/dist-packages (0.16.1+cu118)\n",
            "Requirement already satisfied: torchaudio==2.1.1 in /usr/local/lib/python3.11/dist-packages (2.1.1+cu118)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.11/dist-packages (from torch==2.1.1) (3.18.0)\n",
            "Requirement already satisfied: typing-extensions in /usr/local/lib/python3.11/dist-packages (from torch==2.1.1) (4.13.2)\n",
            "Requirement already satisfied: sympy in /usr/local/lib/python3.11/dist-packages (from torch==2.1.1) (1.13.1)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.11/dist-packages (from torch==2.1.1) (3.4.2)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.11/dist-packages (from torch==2.1.1) (3.1.6)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.11/dist-packages (from torch==2.1.1) (2025.3.2)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from torchvision==0.16.1) (1.26.3)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.11/dist-packages (from torchvision==0.16.1) (2.32.3)\n",
            "Requirement already satisfied: pillow!=8.3.*,>=5.3.0 in /usr/local/lib/python3.11/dist-packages (from torchvision==0.16.1) (11.1.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2->torch==2.1.1) (3.0.2)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests->torchvision==0.16.1) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests->torchvision==0.16.1) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests->torchvision==0.16.1) (2.3.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests->torchvision==0.16.1) (2025.1.31)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy->torch==2.1.1) (1.3.0)\n",
            "Installing collected packages: torch\n",
            "  Attempting uninstall: torch\n",
            "    Found existing installation: torch 2.3.1\n",
            "    Uninstalling torch-2.3.1:\n",
            "      Successfully uninstalled torch-2.3.1\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "protenix 0.4.5 requires torch==2.3.1, but you have torch 2.1.1+cpu which is incompatible.\u001b[0m\u001b[31m\n",
            "\u001b[0mSuccessfully installed torch-2.1.1+cpu\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 1. Uninstall anything conflicting\n",
        "!pip uninstall torch torchvision torchaudio -y\n",
        "\n",
        "# 2. Reinstall torch==2.3.1 explicitly (CPU-only)\n",
        "!pip install torch==2.3.1+cpu torchvision==0.18.1+cpu torchaudio==2.3.1+cpu \\\n",
        "  -f https://download.pytorch.org/whl/cpu/torch_stable.html\n",
        "\n",
        "# 3. Install protenix\n",
        "!pip install protenix==0.4.6\n"
      ],
      "metadata": {
        "id": "lxMiNRm-9-2B",
        "outputId": "cbb6a6f1-f5ba-4e54-a8d3-799310b22777",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Found existing installation: torch 2.1.1+cpu\n",
            "Uninstalling torch-2.1.1+cpu:\n",
            "  Successfully uninstalled torch-2.1.1+cpu\n",
            "Found existing installation: torchvision 0.16.1+cu118\n",
            "Uninstalling torchvision-0.16.1+cu118:\n",
            "  Successfully uninstalled torchvision-0.16.1+cu118\n",
            "Found existing installation: torchaudio 2.1.1+cu118\n",
            "Uninstalling torchaudio-2.1.1+cu118:\n",
            "  Successfully uninstalled torchaudio-2.1.1+cu118\n",
            "Looking in links: https://download.pytorch.org/whl/cpu/torch_stable.html\n",
            "Collecting torch==2.3.1+cpu\n",
            "  Downloading https://download.pytorch.org/whl/cpu/torch-2.3.1%2Bcpu-cp311-cp311-linux_x86_64.whl (190.4 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m190.4/190.4 MB\u001b[0m \u001b[31m6.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting torchvision==0.18.1+cpu\n",
            "  Downloading https://download.pytorch.org/whl/cpu/torchvision-0.18.1%2Bcpu-cp311-cp311-linux_x86_64.whl (1.6 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m73.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting torchaudio==2.3.1+cpu\n",
            "  Downloading https://download.pytorch.org/whl/cpu/torchaudio-2.3.1%2Bcpu-cp311-cp311-linux_x86_64.whl (1.7 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.7/1.7 MB\u001b[0m \u001b[31m70.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: filelock in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1+cpu) (3.18.0)\n",
            "Requirement already satisfied: typing-extensions>=4.8.0 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1+cpu) (4.13.2)\n",
            "Requirement already satisfied: sympy in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1+cpu) (1.13.1)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1+cpu) (3.4.2)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1+cpu) (3.1.6)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1+cpu) (2025.3.2)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from torchvision==0.18.1+cpu) (1.26.3)\n",
            "Requirement already satisfied: pillow!=8.3.*,>=5.3.0 in /usr/local/lib/python3.11/dist-packages (from torchvision==0.18.1+cpu) (11.1.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2->torch==2.3.1+cpu) (3.0.2)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy->torch==2.3.1+cpu) (1.3.0)\n",
            "Installing collected packages: torch, torchvision, torchaudio\n",
            "Successfully installed torch-2.3.1+cpu torchaudio-2.3.1+cpu torchvision-0.18.1+cpu\n",
            "Collecting protenix==0.4.6\n",
            "  Downloading protenix-0.4.6-py3-none-any.whl.metadata (1.1 kB)\n",
            "Requirement already satisfied: PyYaml in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (6.0.2)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (1.14.1)\n",
            "Requirement already satisfied: ml_collections in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (1.1.0)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (4.67.1)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (2.2.2)\n",
            "Requirement already satisfied: optree in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (0.15.0)\n",
            "Requirement already satisfied: rdkit in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (2024.9.6)\n",
            "Requirement already satisfied: biopython==1.83 in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (1.83)\n",
            "Requirement already satisfied: modelcif==0.7 in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (0.7)\n",
            "Requirement already satisfied: biotite==1.0.1 in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (1.0.1)\n",
            "Requirement already satisfied: scikit-learn in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (1.6.1)\n",
            "Requirement already satisfied: scikit-learn-extra in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (0.3.0)\n",
            "Requirement already satisfied: deepspeed>=0.15.1 in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (0.16.7)\n",
            "Requirement already satisfied: protobuf==3.20.2 in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (3.20.2)\n",
            "Requirement already satisfied: icecream in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (2.1.4)\n",
            "Requirement already satisfied: ipdb in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (0.13.13)\n",
            "Requirement already satisfied: wandb in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (0.19.9)\n",
            "Requirement already satisfied: numpy==1.26.3 in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (1.26.3)\n",
            "Requirement already satisfied: matplotlib==3.9.2 in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (3.9.2)\n",
            "Requirement already satisfied: ipywidgets in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (7.7.1)\n",
            "Requirement already satisfied: py3Dmol in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (2.4.2)\n",
            "Requirement already satisfied: nvidia-cublas-cu12 in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (12.1.3.1)\n",
            "Requirement already satisfied: torch==2.3.1 in /usr/local/lib/python3.11/dist-packages (from protenix==0.4.6) (2.3.1+cpu)\n",
            "Requirement already satisfied: biotraj<2.0,>=1.0 in /usr/local/lib/python3.11/dist-packages (from biotite==1.0.1->protenix==0.4.6) (1.2.2)\n",
            "Requirement already satisfied: msgpack>=0.5.6 in /usr/local/lib/python3.11/dist-packages (from biotite==1.0.1->protenix==0.4.6) (1.1.0)\n",
            "Requirement already satisfied: networkx>=2.0 in /usr/local/lib/python3.11/dist-packages (from biotite==1.0.1->protenix==0.4.6) (3.4.2)\n",
            "Requirement already satisfied: requests>=2.12 in /usr/local/lib/python3.11/dist-packages (from biotite==1.0.1->protenix==0.4.6) (2.32.3)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix==0.4.6) (1.3.2)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix==0.4.6) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix==0.4.6) (4.57.0)\n",
            "Requirement already satisfied: kiwisolver>=1.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix==0.4.6) (1.4.8)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix==0.4.6) (24.2)\n",
            "Requirement already satisfied: pillow>=8 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix==0.4.6) (11.1.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix==0.4.6) (3.2.3)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.11/dist-packages (from matplotlib==3.9.2->protenix==0.4.6) (2.8.2)\n",
            "Requirement already satisfied: ihm>=0.34 in /usr/local/lib/python3.11/dist-packages (from modelcif==0.7->protenix==0.4.6) (2.4)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix==0.4.6) (3.18.0)\n",
            "Requirement already satisfied: typing-extensions>=4.8.0 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix==0.4.6) (4.13.2)\n",
            "Requirement already satisfied: sympy in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix==0.4.6) (1.13.1)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix==0.4.6) (3.1.6)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.11/dist-packages (from torch==2.3.1->protenix==0.4.6) (2025.3.2)\n",
            "Requirement already satisfied: einops in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix==0.4.6) (0.8.1)\n",
            "Requirement already satisfied: hjson in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix==0.4.6) (3.1.0)\n",
            "Requirement already satisfied: ninja in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix==0.4.6) (1.11.1.4)\n",
            "Requirement already satisfied: psutil in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix==0.4.6) (5.9.5)\n",
            "Requirement already satisfied: py-cpuinfo in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix==0.4.6) (9.0.0)\n",
            "Requirement already satisfied: pydantic>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix==0.4.6) (2.11.3)\n",
            "Requirement already satisfied: nvidia-ml-py in /usr/local/lib/python3.11/dist-packages (from deepspeed>=0.15.1->protenix==0.4.6) (12.570.86)\n",
            "Requirement already satisfied: colorama>=0.3.9 in /usr/local/lib/python3.11/dist-packages (from icecream->protenix==0.4.6) (0.4.6)\n",
            "Requirement already satisfied: pygments>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from icecream->protenix==0.4.6) (2.18.0)\n",
            "Requirement already satisfied: executing>=2.1.0 in /usr/local/lib/python3.11/dist-packages (from icecream->protenix==0.4.6) (2.2.0)\n",
            "Requirement already satisfied: asttokens>=2.0.1 in /usr/local/lib/python3.11/dist-packages (from icecream->protenix==0.4.6) (3.0.0)\n",
            "Requirement already satisfied: ipython>=7.31.1 in /usr/local/lib/python3.11/dist-packages (from ipdb->protenix==0.4.6) (7.34.0)\n",
            "Requirement already satisfied: decorator in /usr/local/lib/python3.11/dist-packages (from ipdb->protenix==0.4.6) (4.4.2)\n",
            "Requirement already satisfied: ipykernel>=4.5.1 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix==0.4.6) (6.17.1)\n",
            "Requirement already satisfied: ipython-genutils~=0.2.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix==0.4.6) (0.2.0)\n",
            "Requirement already satisfied: traitlets>=4.3.1 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix==0.4.6) (5.7.1)\n",
            "Requirement already satisfied: widgetsnbextension~=3.6.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix==0.4.6) (3.6.10)\n",
            "Requirement already satisfied: jupyterlab-widgets>=1.0.0 in /usr/local/lib/python3.11/dist-packages (from ipywidgets->protenix==0.4.6) (3.0.14)\n",
            "Requirement already satisfied: absl-py in /usr/local/lib/python3.11/dist-packages (from ml_collections->protenix==0.4.6) (1.4.0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas->protenix==0.4.6) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas->protenix==0.4.6) (2025.2)\n",
            "Requirement already satisfied: joblib>=1.2.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn->protenix==0.4.6) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn->protenix==0.4.6) (3.6.0)\n",
            "Requirement already satisfied: click!=8.0.0,>=7.1 in /usr/local/lib/python3.11/dist-packages (from wandb->protenix==0.4.6) (8.1.8)\n",
            "Requirement already satisfied: docker-pycreds>=0.4.0 in /usr/local/lib/python3.11/dist-packages (from wandb->protenix==0.4.6) (0.4.0)\n",
            "Requirement already satisfied: gitpython!=3.1.29,>=1.0.0 in /usr/local/lib/python3.11/dist-packages (from wandb->protenix==0.4.6) (3.1.44)\n",
            "Requirement already satisfied: platformdirs in /usr/local/lib/python3.11/dist-packages (from wandb->protenix==0.4.6) (4.3.7)\n",
            "Requirement already satisfied: sentry-sdk>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from wandb->protenix==0.4.6) (2.26.1)\n",
            "Requirement already satisfied: setproctitle in /usr/local/lib/python3.11/dist-packages (from wandb->protenix==0.4.6) (1.3.5)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.11/dist-packages (from wandb->protenix==0.4.6) (75.2.0)\n",
            "Requirement already satisfied: six>=1.4.0 in /usr/local/lib/python3.11/dist-packages (from docker-pycreds>=0.4.0->wandb->protenix==0.4.6) (1.17.0)\n",
            "Requirement already satisfied: gitdb<5,>=4.0.1 in /usr/local/lib/python3.11/dist-packages (from gitpython!=3.1.29,>=1.0.0->wandb->protenix==0.4.6) (4.0.12)\n",
            "Requirement already satisfied: debugpy>=1.0 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix==0.4.6) (1.8.0)\n",
            "Requirement already satisfied: jupyter-client>=6.1.12 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix==0.4.6) (6.1.12)\n",
            "Requirement already satisfied: matplotlib-inline>=0.1 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix==0.4.6) (0.1.7)\n",
            "Requirement already satisfied: nest-asyncio in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix==0.4.6) (1.6.0)\n",
            "Requirement already satisfied: pyzmq>=17 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix==0.4.6) (24.0.1)\n",
            "Requirement already satisfied: tornado>=6.1 in /usr/local/lib/python3.11/dist-packages (from ipykernel>=4.5.1->ipywidgets->protenix==0.4.6) (6.4.2)\n",
            "Requirement already satisfied: jedi>=0.16 in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1->ipdb->protenix==0.4.6) (0.19.2)\n",
            "Requirement already satisfied: pickleshare in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1->ipdb->protenix==0.4.6) (0.7.5)\n",
            "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1->ipdb->protenix==0.4.6) (3.0.51)\n",
            "Requirement already satisfied: backcall in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1->ipdb->protenix==0.4.6) (0.2.0)\n",
            "Requirement already satisfied: pexpect>4.3 in /usr/local/lib/python3.11/dist-packages (from ipython>=7.31.1->ipdb->protenix==0.4.6) (4.9.0)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/lib/python3.11/dist-packages (from pydantic>=2.0.0->deepspeed>=0.15.1->protenix==0.4.6) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.33.1 in /usr/local/lib/python3.11/dist-packages (from pydantic>=2.0.0->deepspeed>=0.15.1->protenix==0.4.6) (2.33.1)\n",
            "Requirement already satisfied: typing-inspection>=0.4.0 in /usr/local/lib/python3.11/dist-packages (from pydantic>=2.0.0->deepspeed>=0.15.1->protenix==0.4.6) (0.4.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite==1.0.1->protenix==0.4.6) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite==1.0.1->protenix==0.4.6) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite==1.0.1->protenix==0.4.6) (2.3.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests>=2.12->biotite==1.0.1->protenix==0.4.6) (2025.1.31)\n",
            "Requirement already satisfied: notebook>=4.4.1 in /usr/local/lib/python3.11/dist-packages (from widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (6.5.7)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2->torch==2.3.1->protenix==0.4.6) (3.0.2)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy->torch==2.3.1->protenix==0.4.6) (1.3.0)\n",
            "Requirement already satisfied: smmap<6,>=3.0.1 in /usr/local/lib/python3.11/dist-packages (from gitdb<5,>=4.0.1->gitpython!=3.1.29,>=1.0.0->wandb->protenix==0.4.6) (5.0.2)\n",
            "Requirement already satisfied: parso<0.9.0,>=0.8.4 in /usr/local/lib/python3.11/dist-packages (from jedi>=0.16->ipython>=7.31.1->ipdb->protenix==0.4.6) (0.8.4)\n",
            "Requirement already satisfied: jupyter-core>=4.6.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets->protenix==0.4.6) (5.7.2)\n",
            "Requirement already satisfied: argon2-cffi in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (23.1.0)\n",
            "Requirement already satisfied: nbformat in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (5.10.4)\n",
            "Requirement already satisfied: nbconvert>=5 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (7.16.6)\n",
            "Requirement already satisfied: Send2Trash>=1.8.0 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (1.8.3)\n",
            "Requirement already satisfied: terminado>=0.8.3 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (0.18.1)\n",
            "Requirement already satisfied: prometheus-client in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (0.21.1)\n",
            "Requirement already satisfied: nbclassic>=0.4.7 in /usr/local/lib/python3.11/dist-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (1.2.0)\n",
            "Requirement already satisfied: ptyprocess>=0.5 in /usr/local/lib/python3.11/dist-packages (from pexpect>4.3->ipython>=7.31.1->ipdb->protenix==0.4.6) (0.7.0)\n",
            "Requirement already satisfied: wcwidth in /usr/local/lib/python3.11/dist-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython>=7.31.1->ipdb->protenix==0.4.6) (0.2.13)\n",
            "Requirement already satisfied: notebook-shim>=0.2.3 in /usr/local/lib/python3.11/dist-packages (from nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (0.2.4)\n",
            "Requirement already satisfied: beautifulsoup4 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (4.13.4)\n",
            "Requirement already satisfied: bleach!=5.0.0 in /usr/local/lib/python3.11/dist-packages (from bleach[css]!=5.0.0->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (6.2.0)\n",
            "Requirement already satisfied: defusedxml in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (0.7.1)\n",
            "Requirement already satisfied: jupyterlab-pygments in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (0.3.0)\n",
            "Requirement already satisfied: mistune<4,>=2.0.3 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (3.1.3)\n",
            "Requirement already satisfied: nbclient>=0.5.0 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (0.10.2)\n",
            "Requirement already satisfied: pandocfilters>=1.4.1 in /usr/local/lib/python3.11/dist-packages (from nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (1.5.1)\n",
            "Requirement already satisfied: fastjsonschema>=2.15 in /usr/local/lib/python3.11/dist-packages (from nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (2.21.1)\n",
            "Requirement already satisfied: jsonschema>=2.6 in /usr/local/lib/python3.11/dist-packages (from nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (4.23.0)\n",
            "Requirement already satisfied: argon2-cffi-bindings in /usr/local/lib/python3.11/dist-packages (from argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (21.2.0)\n",
            "Requirement already satisfied: webencodings in /usr/local/lib/python3.11/dist-packages (from bleach!=5.0.0->bleach[css]!=5.0.0->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (0.5.1)\n",
            "Requirement already satisfied: tinycss2<1.5,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from bleach[css]!=5.0.0->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (1.4.0)\n",
            "Requirement already satisfied: attrs>=22.2.0 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (25.3.0)\n",
            "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (2024.10.1)\n",
            "Requirement already satisfied: referencing>=0.28.4 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (0.36.2)\n",
            "Requirement already satisfied: rpds-py>=0.7.1 in /usr/local/lib/python3.11/dist-packages (from jsonschema>=2.6->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (0.24.0)\n",
            "Requirement already satisfied: jupyter-server<3,>=1.8 in /usr/local/lib/python3.11/dist-packages (from notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (1.16.0)\n",
            "Requirement already satisfied: cffi>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from argon2-cffi-bindings->argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (1.17.1)\n",
            "Requirement already satisfied: soupsieve>1.2 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4->nbconvert>=5->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (2.6)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.11/dist-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (2.22)\n",
            "Requirement already satisfied: anyio>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (4.9.0)\n",
            "Requirement already satisfied: websocket-client in /usr/local/lib/python3.11/dist-packages (from jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (1.8.0)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.11/dist-packages (from anyio>=3.1.0->jupyter-server<3,>=1.8->notebook-shim>=0.2.3->nbclassic>=0.4.7->notebook>=4.4.1->widgetsnbextension~=3.6.0->ipywidgets->protenix==0.4.6) (1.3.1)\n",
            "Using cached protenix-0.4.6-py3-none-any.whl (441 kB)\n",
            "Installing collected packages: protenix\n",
            "  Attempting uninstall: protenix\n",
            "    Found existing installation: protenix 0.4.5\n",
            "    Uninstalling protenix-0.4.5:\n",
            "      Successfully uninstalled protenix-0.4.5\n",
            "Successfully installed protenix-0.4.6\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "stanford_rna_3d_folding_path = kagglehub.competition_download('stanford-rna-3d-folding')\n",
        "\n",
        "import sys\n",
        "import os\n",
        "from pathlib import Path\n",
        "\n",
        "# Update to correct local path in Colab\n",
        "DATASET_PATH = '/content/required_presets'\n",
        "\n",
        "# Install dependencies from wheels\n",
        "wheel_path = os.path.join(DATASET_PATH, 'wheels')\n",
        "!pip install -qqq --no-index --find-links {wheel_path} torch numpy pandas scipy rdkit protenix\n",
        "\n",
        "# Add Protenix to Python path\n",
        "protenix_path = os.path.join(DATASET_PATH, 'Protenix')\n",
        "sys.path.append(protenix_path)\n"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T13:33:28.721573Z",
          "iopub.execute_input": "2025-04-04T13:33:28.721887Z",
          "iopub.status.idle": "2025-04-04T13:36:06.097096Z",
          "shell.execute_reply.started": "2025-04-04T13:33:28.72186Z",
          "shell.execute_reply": "2025-04-04T13:36:06.096048Z"
        },
        "id": "Xix0amh2Zceq"
      },
      "outputs": [],
      "execution_count": 12
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install --upgrade --force-reinstall numpy pandas\n"
      ],
      "metadata": {
        "id": "3UqYbyvaWqog",
        "outputId": "35861d23-e9b8-4fed-f4f2-4ded8fd0b02d",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting numpy\n",
            "  Downloading numpy-2.2.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (62 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m62.0/62.0 kB\u001b[0m \u001b[31m5.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting pandas\n",
            "  Downloading pandas-2.2.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (89 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m89.9/89.9 kB\u001b[0m \u001b[31m8.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting python-dateutil>=2.8.2 (from pandas)\n",
            "  Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)\n",
            "Collecting pytz>=2020.1 (from pandas)\n",
            "  Downloading pytz-2025.2-py2.py3-none-any.whl.metadata (22 kB)\n",
            "Collecting tzdata>=2022.7 (from pandas)\n",
            "  Downloading tzdata-2025.2-py2.py3-none-any.whl.metadata (1.4 kB)\n",
            "Collecting six>=1.5 (from python-dateutil>=2.8.2->pandas)\n",
            "  Downloading six-1.17.0-py2.py3-none-any.whl.metadata (1.7 kB)\n",
            "Downloading numpy-2.2.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.4 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m16.4/16.4 MB\u001b[0m \u001b[31m90.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pandas-2.2.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m13.1/13.1 MB\u001b[0m \u001b[31m122.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m229.9/229.9 kB\u001b[0m \u001b[31m22.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pytz-2025.2-py2.py3-none-any.whl (509 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m509.2/509.2 kB\u001b[0m \u001b[31m43.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading tzdata-2025.2-py2.py3-none-any.whl (347 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m347.8/347.8 kB\u001b[0m \u001b[31m33.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading six-1.17.0-py2.py3-none-any.whl (11 kB)\n",
            "Installing collected packages: pytz, tzdata, six, numpy, python-dateutil, pandas\n",
            "  Attempting uninstall: pytz\n",
            "    Found existing installation: pytz 2025.2\n",
            "    Uninstalling pytz-2025.2:\n",
            "      Successfully uninstalled pytz-2025.2\n",
            "  Attempting uninstall: tzdata\n",
            "    Found existing installation: tzdata 2025.2\n",
            "    Uninstalling tzdata-2025.2:\n",
            "      Successfully uninstalled tzdata-2025.2\n",
            "  Attempting uninstall: six\n",
            "    Found existing installation: six 1.17.0\n",
            "    Uninstalling six-1.17.0:\n",
            "      Successfully uninstalled six-1.17.0\n",
            "  Attempting uninstall: numpy\n",
            "    Found existing installation: numpy 1.26.3\n",
            "    Uninstalling numpy-1.26.3:\n",
            "      Successfully uninstalled numpy-1.26.3\n",
            "  Attempting uninstall: python-dateutil\n",
            "    Found existing installation: python-dateutil 2.8.2\n",
            "    Uninstalling python-dateutil-2.8.2:\n",
            "      Successfully uninstalled python-dateutil-2.8.2\n",
            "  Attempting uninstall: pandas\n",
            "    Found existing installation: pandas 2.2.2\n",
            "    Uninstalling pandas-2.2.2:\n",
            "      Successfully uninstalled pandas-2.2.2\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "protenix 0.4.6 requires numpy==1.26.3, but you have numpy 2.2.4 which is incompatible.\n",
            "google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 2.2.3 which is incompatible.\n",
            "tensorflow 2.18.0 requires numpy<2.1.0,>=1.26.0, but you have numpy 2.2.4 which is incompatible.\n",
            "tensorflow 2.18.0 requires protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<6.0.0dev,>=3.20.3, but you have protobuf 3.20.2 which is incompatible.\n",
            "ydf 0.11.0 requires protobuf<6.0.0,>=5.29.1, but you have protobuf 3.20.2 which is incompatible.\n",
            "numba 0.60.0 requires numpy<2.1,>=1.22, but you have numpy 2.2.4 which is incompatible.\u001b[0m\u001b[31m\n",
            "\u001b[0mSuccessfully installed numpy-2.2.4 pandas-2.2.3 python-dateutil-2.9.0.post0 pytz-2025.2 six-1.17.0 tzdata-2025.2\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "dateutil",
                  "pytz",
                  "six"
                ]
              },
              "id": "5b69a83984e3482499814fb678afc51b"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "validation_sequence = pd.read_csv('/content/stanford_rna_3d_folding/validation_sequences.csv')\n",
        "validation_labels = pd.read_csv('/content/stanford_rna_3d_folding/validation_labels.csv')"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T13:36:06.098228Z",
          "iopub.execute_input": "2025-04-04T13:36:06.098484Z",
          "iopub.status.idle": "2025-04-04T13:36:06.753676Z",
          "shell.execute_reply.started": "2025-04-04T13:36:06.09846Z",
          "shell.execute_reply": "2025-04-04T13:36:06.752771Z"
        },
        "id": "RaCFHeYdZceq"
      },
      "outputs": [],
      "execution_count": 1
    },
    {
      "cell_type": "code",
      "source": [
        "test_sequences = pd.read_csv('/content/stanford_rna_3d_folding/train_sequences.csv')\n",
        "test_sequences.head()"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T13:36:06.755385Z",
          "iopub.execute_input": "2025-04-04T13:36:06.755709Z",
          "iopub.status.idle": "2025-04-04T13:36:06.830671Z",
          "shell.execute_reply.started": "2025-04-04T13:36:06.755678Z",
          "shell.execute_reply": "2025-04-04T13:36:06.829989Z"
        },
        "id": "8sgU5jUlZceq",
        "outputId": "a4fbcaf0-33fd-46e5-bf7e-475c91ff6cc6",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 598
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  target_id                            sequence temporal_cutoff  \\\n",
              "0    1SCL_A       GGGUGCUCAGUACGAGAGGAACCGCACCC      1995-01-26   \n",
              "1    1RNK_A  GGCGCAGUGGGCUAGCGCCACUCAAAAGGCCCAU      1995-02-27   \n",
              "2    1RHT_A            GGGACUGACGAUCACGCAGUCUAU      1995-06-03   \n",
              "3    1HLX_A                GGGAUAACUUCGGUUGUCCC      1995-09-15   \n",
              "4    1HMH_E  GGCGACCCUGAUGAGGCCGAAAGGCCGAAACCGU      1995-12-07   \n",
              "\n",
              "                                         description  \\\n",
              "0               THE SARCIN-RICIN LOOP, A MODULAR RNA   \n",
              "1  THE STRUCTURE OF AN RNA PSEUDOKNOT THAT CAUSES...   \n",
              "2  24-MER RNA HAIRPIN COAT PROTEIN BINDING SITE F...   \n",
              "3  P1 HELIX NUCLEIC ACIDS (DNA/RNA) RIBONUCLEIC ACID   \n",
              "4  THREE-DIMENSIONAL STRUCTURE OF A HAMMERHEAD RI...   \n",
              "\n",
              "                                       all_sequences  \n",
              "0  >1SCL_1|Chain A|RNA SARCIN-RICIN LOOP|Rattus n...  \n",
              "1  >1RNK_1|Chain A|RNA PSEUDOKNOT|null\\nGGCGCAGUG...  \n",
              "2  >1RHT_1|Chain A|RNA (5'-R(P*GP*GP*GP*AP*CP*UP*...  \n",
              "3  >1HLX_1|Chain A|RNA (5'-R(*GP*GP*GP*AP*UP*AP*A...  \n",
              "4  >1HMH_1|Chains A, C, E|HAMMERHEAD RIBOZYME-RNA...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-80ae6b31-15cd-483b-aad9-a716e0196e9c\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>target_id</th>\n",
              "      <th>sequence</th>\n",
              "      <th>temporal_cutoff</th>\n",
              "      <th>description</th>\n",
              "      <th>all_sequences</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1SCL_A</td>\n",
              "      <td>GGGUGCUCAGUACGAGAGGAACCGCACCC</td>\n",
              "      <td>1995-01-26</td>\n",
              "      <td>THE SARCIN-RICIN LOOP, A MODULAR RNA</td>\n",
              "      <td>&gt;1SCL_1|Chain A|RNA SARCIN-RICIN LOOP|Rattus n...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1RNK_A</td>\n",
              "      <td>GGCGCAGUGGGCUAGCGCCACUCAAAAGGCCCAU</td>\n",
              "      <td>1995-02-27</td>\n",
              "      <td>THE STRUCTURE OF AN RNA PSEUDOKNOT THAT CAUSES...</td>\n",
              "      <td>&gt;1RNK_1|Chain A|RNA PSEUDOKNOT|null\\nGGCGCAGUG...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1RHT_A</td>\n",
              "      <td>GGGACUGACGAUCACGCAGUCUAU</td>\n",
              "      <td>1995-06-03</td>\n",
              "      <td>24-MER RNA HAIRPIN COAT PROTEIN BINDING SITE F...</td>\n",
              "      <td>&gt;1RHT_1|Chain A|RNA (5'-R(P*GP*GP*GP*AP*CP*UP*...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1HLX_A</td>\n",
              "      <td>GGGAUAACUUCGGUUGUCCC</td>\n",
              "      <td>1995-09-15</td>\n",
              "      <td>P1 HELIX NUCLEIC ACIDS (DNA/RNA) RIBONUCLEIC ACID</td>\n",
              "      <td>&gt;1HLX_1|Chain A|RNA (5'-R(*GP*GP*GP*AP*UP*AP*A...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>1HMH_E</td>\n",
              "      <td>GGCGACCCUGAUGAGGCCGAAAGGCCGAAACCGU</td>\n",
              "      <td>1995-12-07</td>\n",
              "      <td>THREE-DIMENSIONAL STRUCTURE OF A HAMMERHEAD RI...</td>\n",
              "      <td>&gt;1HMH_1|Chains A, C, E|HAMMERHEAD RIBOZYME-RNA...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-80ae6b31-15cd-483b-aad9-a716e0196e9c')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-80ae6b31-15cd-483b-aad9-a716e0196e9c button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-80ae6b31-15cd-483b-aad9-a716e0196e9c');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-14fdad71-b9b0-46de-86d4-b7b7c3d52400\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-14fdad71-b9b0-46de-86d4-b7b7c3d52400')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-14fdad71-b9b0-46de-86d4-b7b7c3d52400 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "test_sequences",
              "summary": "{\n  \"name\": \"test_sequences\",\n  \"rows\": 844,\n  \"fields\": [\n    {\n      \"column\": \"target_id\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 844,\n        \"samples\": [\n          \"4YVJ_C\",\n          \"2AGN_A\",\n          \"6WLS_A\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"sequence\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 784,\n        \"samples\": [\n          \"AGUAGUGUGC\",\n          \"GGUCCGUUGAAGACUGCAGGAGAGUGGUUGUUAACCAGAUUUUAACAUCUGAGCCAAAUAACCCGCCGAAGAAGUAAAUCUUUCAGGUGCAUUAUUCUUAGCCAUAUAUUGGCAACGAAUAAGCGAGGACUGUAGUUGGAGGAACCUCUGGAGAGAACCGUUUAAUCGGUCGCCGAAGGAGCAAGCUCUGCGCAUAUGCAGAGUGAAACUCUCAGGCAAAAGGACAGAGGA\",\n          \"GGGCCAUACCUCUUGGGCCUGGUUAGUACCUCUUCGGUGGGAAUACCAGGUGCCC\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"temporal_cutoff\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"num_unique_values\": 476,\n        \"samples\": [\n          \"2024-09-11\",\n          \"1999-01-13\",\n          \"2005-06-07\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"description\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 716,\n        \"samples\": [\n          \"SOLUTION STRUCTURE OF THE COMPLEMENTARY RNA PROMOTER OF INFLUENZA A VIRUS\",\n          \"Novel re-arrangement of an RsmA/cSRa family protein to create a structurally distinct new RNA-binding family member\",\n          \"MS2 PROTEIN CAPSID/RNA COMPLEX\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"all_sequences\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 732,\n        \"samples\": [\n          \">7PTL_1|Chain A[auth B]|Chains: B|synthetic construct (32630)\\nGGGAGAGUACUAUUCAGAUGCAGACCGCAAGUUCAGAGCGGUUUGCAUCUAGGGUACGUUUUCGAACGUAUCCUCCGACUAAGUGUAUUCGUAUACUUAGUGCCUUGUGCCUGCUUCGGCAGGCAUGACCCAAAUGUGCCUUUCGGGGCACAUUUCCGGUCAUCCAAGUUCGCUUGGGUGAUGCGGGCGUAUAGGUUCGUCUAUACGUCCGCGUUUUCCGAGAAGAGGUAACUCGGGAAACCGGUCCACGUGACAAAGGUAGAGUUACGUGGAGGGAGCAGCUGCAAAGGGAUAAUGCAGUUGCUGGCUGGAUGCCAGAACUCACGACUGGCAUCUACGGGGAUGGUGCUCUCCCAAUUCUCCAUUUACCGCCGAAUCGACCCCAACGUGAGAGGGGUCGGUUCCCCGAGCAUAGACCAAUAUCCCAGGUUUAUGCUCCCCAACGCUGGACGAACUACCUACGUCUAGCGUUCCGGCAAAUGAGUCAAUACCUCAGACUUAUUUGCGGUGCCUGAGCCUAAACUGAACAUGGGUUCAGGCAUCUUGGCUCCAGUUCGCUGGAGCCGACGGUAGCGCUGCGUUCGCGCAGUGCUAGGGAGCAUCCGUUUUCGAGCGGAUGCUGGGCGGUUGCCUGUUCGCAGGCAAUCGGGCCUACUCAUGAUUCGUCAUGAGUGGUGACAGCGUGAUGUUCGCAUUACGCUGUCGGGUAGAUGGAGAAUU\\n\",\n          \">1A60_1|Chain A|TYMV PSEUDOKNOT|Turnip yellow mosaic virus (12154)\\nGGGAGCUCAACUCUCCCCCCCUUUUCCGAGGGUCAUCGGAACCA\\n\",\n          \">2L6I_1|Chain A|RNA (5'-R(*GP*AP*UP*CP*UP*CP*UP*UP*GP*UP*AP*GP*AP*UP*CP*A)-3')|\\nGAUCUCUUGUAGAUCA\\n\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 2
        }
      ],
      "execution_count": 2
    },
    {
      "cell_type": "code",
      "source": [
        "submission = pd.read_csv('/content/stanford_rna_3d_folding/sample_submission.csv')"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T13:09:42.991203Z",
          "iopub.execute_input": "2025-04-04T13:09:42.991506Z",
          "iopub.status.idle": "2025-04-04T13:09:43.009622Z",
          "shell.execute_reply.started": "2025-04-04T13:09:42.991479Z",
          "shell.execute_reply": "2025-04-04T13:09:43.009034Z"
        },
        "id": "EgsQfbT8Zceq"
      },
      "outputs": [],
      "execution_count": 3
    },
    {
      "cell_type": "code",
      "source": [
        "submission.head()"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T13:09:45.401912Z",
          "iopub.execute_input": "2025-04-04T13:09:45.402189Z",
          "iopub.status.idle": "2025-04-04T13:09:45.420398Z",
          "shell.execute_reply.started": "2025-04-04T13:09:45.402169Z",
          "shell.execute_reply": "2025-04-04T13:09:45.419538Z"
        },
        "id": "I8pPhpJBZceq",
        "outputId": "40b3240e-759d-45b4-9ec3-b6e297e790a2",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 204
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "        ID resname  resid  x_1  y_1  z_1  x_2  y_2  z_2  x_3  y_3  z_3  x_4  \\\n",
              "0  R1107_1       G      1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "1  R1107_2       G      2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "2  R1107_3       G      3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "3  R1107_4       G      4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "4  R1107_5       G      5  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "\n",
              "   y_4  z_4  x_5  y_5  z_5  \n",
              "0  0.0  0.0  0.0  0.0  0.0  \n",
              "1  0.0  0.0  0.0  0.0  0.0  \n",
              "2  0.0  0.0  0.0  0.0  0.0  \n",
              "3  0.0  0.0  0.0  0.0  0.0  \n",
              "4  0.0  0.0  0.0  0.0  0.0  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-574eb6f1-887d-41a6-958d-3392faaa8437\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>resname</th>\n",
              "      <th>resid</th>\n",
              "      <th>x_1</th>\n",
              "      <th>y_1</th>\n",
              "      <th>z_1</th>\n",
              "      <th>x_2</th>\n",
              "      <th>y_2</th>\n",
              "      <th>z_2</th>\n",
              "      <th>x_3</th>\n",
              "      <th>y_3</th>\n",
              "      <th>z_3</th>\n",
              "      <th>x_4</th>\n",
              "      <th>y_4</th>\n",
              "      <th>z_4</th>\n",
              "      <th>x_5</th>\n",
              "      <th>y_5</th>\n",
              "      <th>z_5</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>R1107_1</td>\n",
              "      <td>G</td>\n",
              "      <td>1</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>R1107_2</td>\n",
              "      <td>G</td>\n",
              "      <td>2</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>R1107_3</td>\n",
              "      <td>G</td>\n",
              "      <td>3</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>R1107_4</td>\n",
              "      <td>G</td>\n",
              "      <td>4</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>R1107_5</td>\n",
              "      <td>G</td>\n",
              "      <td>5</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-574eb6f1-887d-41a6-958d-3392faaa8437')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-574eb6f1-887d-41a6-958d-3392faaa8437 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-574eb6f1-887d-41a6-958d-3392faaa8437');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-d529251e-fa4b-4b34-93ac-60ab15a408b1\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-d529251e-fa4b-4b34-93ac-60ab15a408b1')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-d529251e-fa4b-4b34-93ac-60ab15a408b1 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "submission",
              "summary": "{\n  \"name\": \"submission\",\n  \"rows\": 2515,\n  \"fields\": [\n    {\n      \"column\": \"ID\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 2515,\n        \"samples\": [\n          \"R1136_345\",\n          \"R1126_281\",\n          \"R1126_296\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"resname\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 4,\n        \"samples\": [\n          \"C\",\n          \"U\",\n          \"G\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"resid\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 176,\n        \"min\": 1,\n        \"max\": 720,\n        \"num_unique_values\": 720,\n        \"samples\": [\n          341,\n          291,\n          55\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"x_1\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"y_1\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"z_1\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"x_2\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"y_2\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"z_2\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"x_3\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"y_3\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"z_3\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"x_4\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"y_4\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"z_4\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"x_5\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"y_5\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"z_5\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 4
        }
      ],
      "execution_count": 4
    },
    {
      "cell_type": "code",
      "source": [
        "len(submission)"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T13:10:05.712892Z",
          "iopub.execute_input": "2025-04-04T13:10:05.713183Z",
          "iopub.status.idle": "2025-04-04T13:10:05.71803Z",
          "shell.execute_reply.started": "2025-04-04T13:10:05.713161Z",
          "shell.execute_reply": "2025-04-04T13:10:05.717163Z"
        },
        "id": "s5scEV0oZceq",
        "outputId": "a35cfefb-6bc9-4682-e05d-3e058e8a6ec7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "2515"
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ],
      "execution_count": 5
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "print(os.listdir('/content/stanford_rna_3d_folding'))\n"
      ],
      "metadata": {
        "id": "_C1_-p9urK3t",
        "outputId": "2212d689-7d44-4e76-f5aa-a4fbab5f0c8c",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['sample_submission.csv', 'test_sequences.csv', 'train_labels.csv', 'train_sequences.csv', 'validation_labels.csv', 'MSA', 'validation_sequences.csv']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!ls -R /content/\n",
        "\n"
      ],
      "metadata": {
        "id": "m4iVsW6Z_ppb",
        "outputId": "3f05a51a-3da4-46ed-cfb0-cd289b739f24",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/:\n",
            "biotite\t\t  required-presets.zip\t   stanford-rna-3d-folding.zip\n",
            "kaggle.json\t  sample_data\n",
            "required_presets  stanford_rna_3d_folding\n",
            "\n",
            "/content/biotite:\n",
            "benchmarks  environment.yml  pyproject.toml  src\n",
            "doc\t    LICENSE.rst      README.rst      tests\n",
            "\n",
            "/content/biotite/benchmarks:\n",
            "conftest.py  __init__.py  sequence  structure\n",
            "\n",
            "/content/biotite/benchmarks/sequence:\n",
            "align  __init__.py\n",
            "\n",
            "/content/biotite/benchmarks/sequence/align:\n",
            "benchmark_kmers.py  __init__.py\n",
            "\n",
            "/content/biotite/benchmarks/structure:\n",
            "benchmark_alphabet.py  benchmark_compare.py  benchmark_superimpose.py\n",
            "benchmark_celllist.py  benchmark_pdbx.py     __init__.py\n",
            "\n",
            "/content/biotite/doc:\n",
            "404.rst\t\t conf.py\t index.rst    references.bib  tutorial\n",
            "apidoc.json\t contribution\t install.rst  scraper.py      viewcode.py\n",
            "apidoc.py\t examples\t logo.rst     static\n",
            "bibliography.py  extensions.rst  preamble.py  switcher.py\n",
            "\n",
            "/content/biotite/doc/contribution:\n",
            "deployment.rst\tdevelopment.rst  documentation.rst  index.rst  testing.rst\n",
            "\n",
            "/content/biotite/doc/examples:\n",
            "download  index.rst  scripts\n",
            "\n",
            "/content/biotite/doc/examples/download:\n",
            "Array_Seq.txt  glycosylase_anm_vectors.csv  NF54_10ug.csv\n",
            "dppc_n128.pdb  lysozyme_md.pdb\t\t    waterbox_md.pdb\n",
            "FCR3_10ug.csv  lysozyme_md.xtc\t\t    waterbox_md.xtc\n",
            "\n",
            "/content/biotite/doc/examples/scripts:\n",
            "sequence  structure\n",
            "\n",
            "/content/biotite/doc/examples/scripts/sequence:\n",
            "annotation  homology  misc  profile  README.rst  sequencing\n",
            "\n",
            "/content/biotite/doc/examples/scripts/sequence/annotation:\n",
            "operon_map.py\t       plasmid_map.py  region_visualization.py\n",
            "plasmid_map_custom.py  README.rst      sigma_domains.py\n",
            "\n",
            "/content/biotite/doc/examples/scripts/sequence/homology:\n",
            "avidin_alignment.py    hcn_hydropathy.py     pi3k_alignment.py\n",
            "bionigma_alignment.py  hcn_similarity.py     plotepiscan.py\n",
            "genome_comparison.py   homolog_msa.py\t     README.rst\n",
            "genome_search.py       lexa_conservation.py  residue_coevolution.py\n",
            "gpcr_evolution.py      luxa_comparison.py    thca_synthase_polymorphism.py\n",
            "\n",
            "/content/biotite/doc/examples/scripts/sequence/misc:\n",
            "blosum_dendrogram.py\t  color_schemes.py\t\t README.rst\n",
            "codon_usage.py\t\t  local_alignment_statistics.py\n",
            "color_schemes_protein.py  orf_identification.py\n",
            "\n",
            "/content/biotite/doc/examples/scripts/sequence/profile:\n",
            "anderson_logo.py  rbs_identification.py  README.rst\n",
            "\n",
            "/content/biotite/doc/examples/scripts/sequence/sequencing:\n",
            "gene_counts.py\t    quality_control.py\tread_quality.py\n",
            "genome_assembly.py  README.rst\n",
            "\n",
            "/content/biotite/doc/examples/scripts/structure:\n",
            "alphabet  contacts  misc  modeling  molecule  nucleotide  protein  README.rst\n",
            "\n",
            "/content/biotite/doc/examples/scripts/structure/alphabet:\n",
            "msa.py\tREADME.rst  structure_search.py\n",
            "\n",
            "/content/biotite/doc/examples/scripts/structure/contacts:\n",
            "adjacency_matrix.py  disulfide_bonds.py  leaflet.py\n",
            "contact_sites.py     domain_hbonds.py\t README.rst\n",
            "\n",
            "/content/biotite/doc/examples/scripts/structure/misc:\n",
            "biological_assembly.py\tglycan_visualization.license  pdb_statistics.py\n",
            "diameter.py\t\tglycan_visualization.py       README.rst\n",
            "gap_bars.py\t\thomolog_superimposition.py\n",
            "\n",
            "/content/biotite/doc/examples/scripts/structure/modeling:\n",
            "docking.py\t    normal_modes.py\t trajectory_sse.py\n",
            "md_analysis.py\t    README.rst\t\t water_exchange_noexec.py\n",
            "mmtf_trajectory.py  rotamer_library.py\t water_exchange.png\n",
            "model_lddt.py\t    solvation_shells.py\n",
            "\n",
            "/content/biotite/doc/examples/scripts/structure/molecule:\n",
            "alkane_isomers.py\t    peoe_visualization.py\n",
            "molecular_visualization.py  README.rst\n",
            "\n",
            "/content/biotite/doc/examples/scripts/structure/nucleotide:\n",
            "base_pairs.py\t    README.rst\t      watson_crick.py\n",
            "leontis_westhof.py  transfer_rnas.py\n",
            "\n",
            "/content/biotite/doc/examples/scripts/structure/protein:\n",
            "peptide_assembly.py  README.rst\t\t   sheet_arrangement.py\n",
            "ramachandran.py      residue_chirality.py  transketolase_sse.py\n",
            "\n",
            "/content/biotite/doc/static:\n",
            "assets\tbiotite.css  fonts  fonts.css\n",
            "\n",
            "/content/biotite/doc/static/assets:\n",
            "figures  general  sponsors\n",
            "\n",
            "/content/biotite/doc/static/assets/figures:\n",
            "app_lifecycle.png  dihedral.svg  rmsf.svg  sasa.svg  symbol_encoding.png\n",
            "\n",
            "/content/biotite/doc/static/assets/general:\n",
            "biotite_icon_32p.png  biotite_icon_thumb.png  biotite_logo.svg\n",
            "biotite_icon_404.svg  biotite_logo_l.png      button_pypi.svg\n",
            "biotite_icon_64p.png  biotite_logo_m.png\n",
            "biotite_icon.svg      biotite_logo_s.png\n",
            "\n",
            "/content/biotite/doc/static/assets/sponsors:\n",
            "LICENSE.rst  vantai_logo.jpg\n",
            "\n",
            "/content/biotite/doc/static/fonts:\n",
            "crete-round-v13-latin-italic.woff    fira-sans-v15-latin-italic.woff2\n",
            "crete-round-v13-latin-italic.woff2   fira-sans-v15-latin-regular.woff\n",
            "crete-round-v13-latin-regular.woff   fira-sans-v15-latin-regular.woff2\n",
            "crete-round-v13-latin-regular.woff2  LICENSE\n",
            "fira-sans-v15-latin-italic.woff\n",
            "\n",
            "/content/biotite/doc/tutorial:\n",
            "application  database  index.rst  interface  preamble.rst  sequence  structure\n",
            "\n",
            "/content/biotite/doc/tutorial/application:\n",
            "blast.rst  dssp.rst  index.rst\tmsa.rst\n",
            "\n",
            "/content/biotite/doc/tutorial/database:\n",
            "entrez.rst  index.rst  rcsb.rst\n",
            "\n",
            "/content/biotite/doc/tutorial/interface:\n",
            "index.rst  openmm.rst  pymol.rst  rdkit.rst\n",
            "\n",
            "/content/biotite/doc/tutorial/sequence:\n",
            "align_heuristic.rst  align_optimal.rst\tencoding.rst  io.rst\t    types.rst\n",
            "align_multiple.rst   annotations.rst\tindex.rst     profiles.rst\n",
            "\n",
            "/content/biotite/doc/tutorial/structure:\n",
            "alphabet.rst  bonds.rst    filter.rst  io.rst\t\ttrajectories.rst\n",
            "atoms.rst     editing.rst  index.rst   measurement.rst\n",
            "\n",
            "/content/biotite/src:\n",
            "biotite\n",
            "\n",
            "/content/biotite/src/biotite:\n",
            "application  database  __init__.py  sequence\t  structure\n",
            "copyable.py  file.py   interface    setup_ccd.py  visualize.py\n",
            "\n",
            "/content/biotite/src/biotite/application:\n",
            "application.py\tclustalo     localapp.py  muscle  util.py\n",
            "autodock\tdssp\t     mafft\t  sra\t  viennarna\n",
            "blast\t\t__init__.py  msaapp.py\t  tantan  webapp.py\n",
            "\n",
            "/content/biotite/src/biotite/application/autodock:\n",
            "app.py\t__init__.py\n",
            "\n",
            "/content/biotite/src/biotite/application/blast:\n",
            "alignment.py  __init__.py  webapp.py\n",
            "\n",
            "/content/biotite/src/biotite/application/clustalo:\n",
            "app.py\t__init__.py\n",
            "\n",
            "/content/biotite/src/biotite/application/dssp:\n",
            "app.py\t__init__.py\n",
            "\n",
            "/content/biotite/src/biotite/application/mafft:\n",
            "app.py\t__init__.py\n",
            "\n",
            "/content/biotite/src/biotite/application/muscle:\n",
            "app3.py  app5.py  __init__.py\n",
            "\n",
            "/content/biotite/src/biotite/application/sra:\n",
            "app.py\t__init__.py\n",
            "\n",
            "/content/biotite/src/biotite/application/tantan:\n",
            "app.py\t__init__.py\n",
            "\n",
            "/content/biotite/src/biotite/application/viennarna:\n",
            "__init__.py  rnaalifold.py  rnafold.py\trnaplot.py  util.py\n",
            "\n",
            "/content/biotite/src/biotite/database:\n",
            "afdb  entrez  error.py\t__init__.py  pubchem  rcsb  uniprot\n",
            "\n",
            "/content/biotite/src/biotite/database/afdb:\n",
            "download.py  __init__.py\n",
            "\n",
            "/content/biotite/src/biotite/database/entrez:\n",
            "check.py  dbnames.py  download.py  __init__.py\tkey.py\tquery.py\n",
            "\n",
            "/content/biotite/src/biotite/database/pubchem:\n",
            "download.py  error.py  __init__.py  query.py  throttle.py\n",
            "\n",
            "/content/biotite/src/biotite/database/rcsb:\n",
            "download.py  __init__.py  query.py\n",
            "\n",
            "/content/biotite/src/biotite/database/uniprot:\n",
            "check.py  download.py  __init__.py  query.py\n",
            "\n",
            "/content/biotite/src/biotite/interface:\n",
            "__init__.py  openmm  pymol  rdkit  version.py  warning.py\n",
            "\n",
            "/content/biotite/src/biotite/interface/openmm:\n",
            "__init__.py  state.py  system.py\n",
            "\n",
            "/content/biotite/src/biotite/interface/pymol:\n",
            "cgo.py\tconvert.py  display.py\t__init__.py  object.py\tshapes.py  startup.py\n",
            "\n",
            "/content/biotite/src/biotite/interface/rdkit:\n",
            "__init__.py  mol.py\n",
            "\n",
            "/content/biotite/src/biotite/sequence:\n",
            "align\t       codec.pyx\t graphics     phylo\t  seqtypes.py\n",
            "alphabet.py    codon.py\t\t __init__.py  profile.py  sequence.py\n",
            "annotation.py  codon_tables.txt  io\t      search.py\n",
            "\n",
            "/content/biotite/src/biotite/sequence/align:\n",
            "alignment.py  kmeralphabet.pyx\t  matrix_data\t   primes.txt\n",
            "banded.pyx    kmersimilarity.pyx  matrix.py\t   selector.pyx\n",
            "buckets.py    kmertable.pyx\t  multiple.pyx\t   statistics.py\n",
            "cigar.py      localgapped.pyx\t  pairwise.pyx\t   tracetable.pxd\n",
            "__init__.py   localungapped.pyx   permutation.pyx  tracetable.pyx\n",
            "\n",
            "/content/biotite/src/biotite/sequence/align/matrix_data:\n",
            "3Di.mat\t\t   CorBLOSUM49_5.0.mat\t PAM20.mat   PAM410.mat\n",
            "BLOSUM100.mat\t   CorBLOSUM57_13p.mat\t PAM210.mat  PAM420.mat\n",
            "BLOSUM30.mat\t   CorBLOSUM57_14.3.mat  PAM220.mat  PAM430.mat\n",
            "BLOSUM35.mat\t   CorBLOSUM61_5.0.mat\t PAM230.mat  PAM440.mat\n",
            "BLOSUM40.mat\t   CorBLOSUM66_13p.mat\t PAM240.mat  PAM450.mat\n",
            "BLOSUM45.mat\t   CorBLOSUM67_14.3.mat  PAM250.mat  PAM460.mat\n",
            "BLOSUM50_13p.mat   DAYHOFF.mat\t\t PAM260.mat  PAM470.mat\n",
            "BLOSUM50_14.3.mat  GONNET.mat\t\t PAM270.mat  PAM480.mat\n",
            "BLOSUM50_5.0.mat   IDENTITY.mat\t\t PAM280.mat  PAM490.mat\n",
            "BLOSUM50.mat\t   MATCH.mat\t\t PAM290.mat  PAM500.mat\n",
            "BLOSUM55.mat\t   NUC.mat\t\t PAM300.mat  PAM50.mat\n",
            "BLOSUM60.mat\t   PAM100.mat\t\t PAM30.mat   PAM60.mat\n",
            "BLOSUM62_13p.mat   PAM10.mat\t\t PAM310.mat  PAM70.mat\n",
            "BLOSUM62_14.3.mat  PAM110.mat\t\t PAM320.mat  PAM80.mat\n",
            "BLOSUM62_5.0.mat   PAM120.mat\t\t PAM330.mat  PAM90.mat\n",
            "BLOSUM62.mat\t   PAM130.mat\t\t PAM340.mat  PB.license\n",
            "BLOSUM65.mat\t   PAM140.mat\t\t PAM350.mat  PB.mat\n",
            "BLOSUM70.mat\t   PAM150.mat\t\t PAM360.mat  RBLOSUM52_5.0.mat\n",
            "BLOSUM75.mat\t   PAM160.mat\t\t PAM370.mat  RBLOSUM59_13p.mat\n",
            "BLOSUM80.mat\t   PAM170.mat\t\t PAM380.mat  RBLOSUM59_14.3.mat\n",
            "BLOSUM85.mat\t   PAM180.mat\t\t PAM390.mat  RBLOSUM64_5.0.mat\n",
            "BLOSUM90.mat\t   PAM190.mat\t\t PAM400.mat  RBLOSUM69_13p.mat\n",
            "BLOSUMN.mat\t   PAM200.mat\t\t PAM40.mat   RBLOSUM69_14.3.mat\n",
            "\n",
            "/content/biotite/src/biotite/sequence/graphics:\n",
            "alignment.py   colorschemes.py\tfeatures.py  logo.py\n",
            "color_schemes  dendrogram.py\t__init__.py  plasmid.py\n",
            "\n",
            "/content/biotite/src/biotite/sequence/graphics/color_schemes:\n",
            "3di_flower.json        jalview_hydrophobicity.json  pb_flower.json\n",
            "autumn.json\t       jalview_prop_helix.json\t    rainbow_dna.json\n",
            "blossom.json\t       jalview_prop_strand.json     rainbow_protein.json\n",
            "clustalx_dna.json      jalview_prop_turn.json\t    spring.json\n",
            "clustalx_protein.json  jalview_taylor.json\t    sunset.json\n",
            "flower.json\t       jalview_zappo.json\t    wither.json\n",
            "jalview_buried.json    ocean.json\n",
            "\n",
            "/content/biotite/src/biotite/sequence/io:\n",
            "fasta  fastq  genbank  general.py  gff\t__init__.py\n",
            "\n",
            "/content/biotite/src/biotite/sequence/io/fasta:\n",
            "convert.py  file.py  __init__.py\n",
            "\n",
            "/content/biotite/src/biotite/sequence/io/fastq:\n",
            "convert.py  file.py  __init__.py\n",
            "\n",
            "/content/biotite/src/biotite/sequence/io/genbank:\n",
            "annotation.py  file.py\t__init__.py  metadata.py  sequence.py\n",
            "\n",
            "/content/biotite/src/biotite/sequence/io/gff:\n",
            "convert.py  file.py  __init__.py\n",
            "\n",
            "/content/biotite/src/biotite/sequence/phylo:\n",
            "__init__.py  nj.pyx  tree.pyx  upgma.pyx\n",
            "\n",
            "/content/biotite/src/biotite/structure:\n",
            "alphabet      dotbracket.py  mechanics.py    spacegroups.json\n",
            "atoms.py      error.py\t     molecules.py    spacegroups.license\n",
            "basepairs.py  filter.py      pseudoknots.py  sse.py\n",
            "bonds.pyx     geometry.py    rdf.py\t     superimpose.py\n",
            "box.py\t      graphics\t     repair.py\t     tm.py\n",
            "celllist.pyx  hbond.py\t     residues.py     transform.py\n",
            "chains.py     info\t     rings.py\t     util.py\n",
            "charges.pyx   __init__.py    sasa.pyx\n",
            "compare.py    integrity.py   segments.py\n",
            "density.py    io\t     sequence.py\n",
            "\n",
            "/content/biotite/src/biotite/structure/alphabet:\n",
            "encoder.py\t\t      i3d.py\t   layers.py   pb.py\n",
            "encoder_weights_3di.kerasify  __init__.py  pb.license  unkerasify.py\n",
            "\n",
            "/content/biotite/src/biotite/structure/graphics:\n",
            "atoms.py  __init__.py  rna.py\n",
            "\n",
            "/content/biotite/src/biotite/structure/info:\n",
            "atom_masses.json  bonds.py  groups.py\t masses.py  radii.py\n",
            "atoms.py\t  ccd.py    __init__.py  misc.py    standardize.py\n",
            "\n",
            "/content/biotite/src/biotite/structure/io:\n",
            "dcd\t    gro\t\t mol\t pdb\tpdbx\t     trr      xtc\n",
            "general.py  __init__.py  netcdf  pdbqt\ttrajfile.py  util.py\n",
            "\n",
            "/content/biotite/src/biotite/structure/io/dcd:\n",
            "file.py  __init__.py\n",
            "\n",
            "/content/biotite/src/biotite/structure/io/gro:\n",
            "file.py  __init__.py\n",
            "\n",
            "/content/biotite/src/biotite/structure/io/mol:\n",
            "convert.py  ctab.py  header.py\t__init__.py  mol.py  sdf.py\n",
            "\n",
            "/content/biotite/src/biotite/structure/io/netcdf:\n",
            "file.py  __init__.py\n",
            "\n",
            "/content/biotite/src/biotite/structure/io/pdb:\n",
            "convert.py  file.py  hybrid36.pyx  __init__.py\n",
            "\n",
            "/content/biotite/src/biotite/structure/io/pdbqt:\n",
            "convert.py  file.py  __init__.py\n",
            "\n",
            "/content/biotite/src/biotite/structure/io/pdbx:\n",
            "bcif.py  component.py  convert.py    __init__.py\n",
            "cif.py\t compress.py   encoding.pyx\n",
            "\n",
            "/content/biotite/src/biotite/structure/io/trr:\n",
            "file.py  __init__.py\n",
            "\n",
            "/content/biotite/src/biotite/structure/io/xtc:\n",
            "file.py  __init__.py\n",
            "\n",
            "/content/biotite/tests:\n",
            "application  __init__.py  structure\t   test_repr.py\n",
            "conftest.py  interface\t  test_doctest.py  test_version.py\n",
            "database     sequence\t  test_modname.py  util.py\n",
            "\n",
            "/content/biotite/tests/application:\n",
            "data\t\t  test_blast.py  test_rnaalifold.py  test_sra.py\n",
            "__init__.py\t  test_dssp.py\t test_rnafold.py     test_tantan.py\n",
            "test_autodock.py  test_msa.py\t test_rnaplot.py\n",
            "\n",
            "/content/biotite/tests/application/data:\n",
            "2rtg.bcif  2rtg.mmtf\n",
            "\n",
            "/content/biotite/tests/database:\n",
            "__init__.py   test_entrez.py   test_rcsb.py\n",
            "test_afdb.py  test_pubchem.py  test_uniprot.py\n",
            "\n",
            "/content/biotite/tests/interface:\n",
            "data  __init__.py  pymol  test_openmm.py  test_rdkit.py  test_version.py\n",
            "\n",
            "/content/biotite/tests/interface/data:\n",
            "README.rst  smiles.txt\n",
            "\n",
            "/content/biotite/tests/interface/pymol:\n",
            "conftest.py  test_command.py\t test_selection.py\n",
            "__init__.py  test_conversion.py  test_shapes.py\n",
            "test_cgo.py  test_display.py\t test_startup.py\n",
            "\n",
            "/content/biotite/tests/sequence:\n",
            "align\t\t    test_codon.py      test_gff.py\t test_seqtypes.py\n",
            "data\t\t    test_fasta.py      test_graphics.py  test_sequence.py\n",
            "__init__.py\t    test_fastq.py      test_phylo.py\n",
            "test_alphabet.py    test_genbank.py    test_profile.py\n",
            "test_annotation.py  test_generalio.py  test_search.py\n",
            "\n",
            "/content/biotite/tests/sequence/align:\n",
            "conftest.py\t      test_kmersimilarity.py  test_pairwise.py\n",
            "__init__.py\t      test_kmertable.py       test_permutation.py\n",
            "test_alignment.py     test_localgapped.py     test_selector.py\n",
            "test_banded.py\t      test_localungapped.py   test_statistics.py\n",
            "test_cigar.py\t      test_matrix.py\n",
            "test_kmeralphabet.py  test_multiple.py\n",
            "\n",
            "/content/biotite/tests/sequence/data:\n",
            "alignment.fasta   ec_bl21.gb\t      invalid.fasta\t random.fasta\n",
            "bt_lysozyme.gff3  ec_bl21.gff3\t      multifile.gp\t random.fastq\n",
            "bt_lysozyme.gp\t  gg_avidin.gb\t      newick_upgma.txt\t sc_chrom1.gb\n",
            "cas9.fasta\t  gg_avidin.gff3      nuc.fasta\t\t sc_chrom1.gff3\n",
            "distances.txt\t  indexing_test.gff3  percent_test.gff3  selenocysteine.fasta\n",
            "ec_bl21.fasta\t  influenca.gb\t      prot.fasta\n",
            "\n",
            "/content/biotite/tests/structure:\n",
            "alphabet\t   test_chains.py      test_info.py\t    test_sasa.py\n",
            "data\t\t   test_charges.py     test_integrity.py    test_sequence.py\n",
            "__init__.py\t   test_compare.py     test_mechanics.py    test_sse.py\n",
            "io\t\t   test_density.py     test_molecules.py    test_superimpose.py\n",
            "test_atoms.py\t   test_dotbracket.py  test_pseudoknots.py  test_tm.py\n",
            "test_basepairs.py  test_filter.py      test_rdf.py\t    test_transform.py\n",
            "test_bonds.py\t   test_generalio.py   test_repair.py\n",
            "test_box.py\t   test_geometry.py    test_residues.py\n",
            "test_celllist.py   test_hbond.py       test_rings.py\n",
            "\n",
            "/content/biotite/tests/structure/alphabet:\n",
            "__init__.py  test_i3d.py  test_pb.py\n",
            "\n",
            "/content/biotite/tests/structure/data:\n",
            "1aki.bcif  1gya.cif\t1o1z.gro   4p5j.cif   7gsa.bcif\n",
            "1aki.cif   1gya.gro\t1o1z.pdb   4p5j.gro   7gsa.cif\n",
            "1aki.gro   1gya.pdb\t2axd.bcif  4p5j.pdb   alphabet\n",
            "1aki.pdb   1igy.bcif\t2axd.cif   5eil.bcif  base_pairs\n",
            "1crr.bcif  1igy.cif\t2axd.pdb   5eil.cif   create_test_structures.py\n",
            "1crr.cif   1igy.gro\t2d0f.bcif  5eil.pdb   homologs\n",
            "1crr.gro   1igy.pdb\t2d0f.cif   5h73.bcif  ids.txt\n",
            "1crr.pdb   1l2y.bcif\t2d0f.pdb   5h73.cif   misc\n",
            "1dix.bcif  1l2y.cif\t3o5r.bcif  5h73.gro   molecules\n",
            "1dix.cif   1l2y.dcd\t3o5r.cif   5h73.pdb   nextgen\n",
            "1dix.gro   1l2y.gro\t3o5r.gro   5ugo.bcif  psea.fasta\n",
            "1dix.pdb   1l2y.netcdf\t3o5r.pdb   5ugo.cif   pseudoknots\n",
            "1f2n.bcif  1l2y.pdb\t4gxy.bcif  5ugo.gro   README.rst\n",
            "1f2n.cif   1l2y.trr\t4gxy.cif   5ugo.pdb   tm\n",
            "1f2n.gro   1l2y.xtc\t4gxy.gro   5zng.bcif  waterbox.gro\n",
            "1f2n.pdb   1o1z.bcif\t4gxy.pdb   5zng.cif\n",
            "1gya.bcif  1o1z.cif\t4p5j.bcif  5zng.pdb\n",
            "\n",
            "/content/biotite/tests/structure/data/alphabet:\n",
            "1ay7.bcif  i3d.fasta  pb.fasta\tREADME.rst\n",
            "\n",
            "/content/biotite/tests/structure/data/base_pairs:\n",
            "1bna.cif\t 1nkw.cif\t  1xnr_edges.json\n",
            "1bna.pdb\t 1nkw_edges.json  1xnr_sugar.json\n",
            "1gid.cif\t 1nkw_sugar.json  create_bond_orientation_test_data.py\n",
            "1gid_edges.json  1qxb.cif\t  create_interacting_edge_test_data.py\n",
            "1gid_sugar.json  1xnr.cif\n",
            "\n",
            "/content/biotite/tests/structure/data/homologs:\n",
            "1gl4.bcif  1p4k.bcif  2nwd.bcif  3lsj.bcif  3rd3.bcif  6oa8.bcif\n",
            "1hml.bcif  1qgi.bcif  3kcs.bcif  3lzm.bcif  4osx.bcif  README.rst\n",
            "\n",
            "/content/biotite/tests/structure/data/misc:\n",
            "hbond.json  hbond.py  rdf.json\trdf.py\tREADME.rst  sasa.json  sasa.py\n",
            "\n",
            "/content/biotite/tests/structure/data/molecules:\n",
            "03F.cif        13136.v3000.sdf\t    CYN.v3000.sdf  README.rst\n",
            "03F.sdf        create_v3000_sdf.py  HWB.cif\t   TYR.cif\n",
            "03F.v3000.sdf  CYN.cif\t\t    HWB.sdf\t   TYR.sdf\n",
            "13136.sdf      CYN.sdf\t\t    HWB.v3000.sdf  TYR.v3000.sdf\n",
            "\n",
            "/content/biotite/tests/structure/data/nextgen:\n",
            "pdb_00001l2y_xyz-enrich.cif  README.rst\n",
            "\n",
            "/content/biotite/tests/structure/data/pseudoknots:\n",
            "test0_knotted.json     test16_knotted.json    test3_knotted.json\n",
            "test0_unknotted.json   test16_unknotted.json  test3_unknotted.json\n",
            "test10_knotted.json    test17_knotted.json    test4_knotted.json\n",
            "test10_unknotted.json  test17_unknotted.json  test4_unknotted.json\n",
            "test11_knotted.json    test18_knotted.json    test5_knotted.json\n",
            "test11_unknotted.json  test18_unknotted.json  test5_unknotted.json\n",
            "test12_knotted.json    test19_knotted.json    test6_knotted.json\n",
            "test12_unknotted.json  test19_unknotted.json  test6_unknotted.json\n",
            "test13_knotted.json    test1_knotted.json     test7_knotted.json\n",
            "test13_unknotted.json  test1_unknotted.json   test7_unknotted.json\n",
            "test14_knotted.json    test20_knotted.json    test8_knotted.json\n",
            "test14_unknotted.json  test20_unknotted.json  test8_unknotted.json\n",
            "test15_knotted.json    test2_knotted.json     test9_knotted.json\n",
            "test15_unknotted.json  test2_unknotted.json   test9_unknotted.json\n",
            "\n",
            "/content/biotite/tests/structure/data/tm:\n",
            "README.rst  tm.py  tm_scores.json\n",
            "\n",
            "/content/biotite/tests/structure/io:\n",
            "__init__.py  test_mol.py  test_pdbqt.py  test_trajectory.py\n",
            "test_gro.py  test_pdb.py  test_pdbx.py\n",
            "\n",
            "/content/required_presets:\n",
            "Protenix  wheels\n",
            "\n",
            "/content/required_presets/Protenix:\n",
            "assets\t\t    finetune_demo.sh\t\t   requirements.txt\n",
            "build\t\t    inference_demo.sh\t\t   runner\n",
            "CODE_OF_CONDUCT.md  LICENSE\t\t\t   scripts\n",
            "configs\t\t    notebooks\t\t\t   setup.py\n",
            "CONTRIBUTING.md     protenix\t\t\t   tests\n",
            "Dockerfile\t    protenix.egg-info\t\t   train_demo.sh\n",
            "docs\t\t    Protenix_Technical_Report.pdf\n",
            "examples\t    README.md\n",
            "\n",
            "/content/required_presets/Protenix/assets:\n",
            "constraint_metrics.png\toverall_metrics.png\n",
            "license_header.txt\tprotenix_predictions.gif\n",
            "\n",
            "/content/required_presets/Protenix/build:\n",
            "lib\n",
            "\n",
            "/content/required_presets/Protenix/build/lib:\n",
            "configs  protenix  runner  scripts  tests\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/configs:\n",
            "configs_base.py  configs_data.py  configs_inference.py\t__init__.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix:\n",
            "config\tdata  __init__.py  metrics  model  openfold_local  utils  web_service\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/config:\n",
            "config.py  extend_types.py  __init__.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/data:\n",
            "ccd.py\t\t  filter.py\t\t  msa_featurizer.py\n",
            "constants.py\t  infer_data_pipeline.py  msa_utils.py\n",
            "dataloader.py\t  __init__.py\t\t  parser.py\n",
            "data_pipeline.py  json_maker.py\t\t  substructure_perms.py\n",
            "dataset.py\t  json_parser.py\t  tokenizer.py\n",
            "featurizer.py\t  json_to_feature.py\t  utils.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/metrics:\n",
            "clash.py  __init__.py  lddt_metrics.py\trmsd.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/model:\n",
            "generator.py  layer_norm  modules      sample_confidence.py\n",
            "__init__.py   loss.py\t  protenix.py  utils.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/model/layer_norm:\n",
            "__init__.py  kernel  layer_norm.py  torch_ext_compile.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/model/layer_norm/kernel:\n",
            "compat.h  layer_norm_cuda.cpp  layer_norm_cuda_kernel.cu  type_shim.h\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/model/modules:\n",
            "confidence.py  embedders.py  head.py\t  pairformer.py  transformer.py\n",
            "diffusion.py   frames.py     __init__.py  primitives.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/openfold_local:\n",
            "data  __init__.py  model  np  README.md  utils\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/openfold_local/data:\n",
            "data_transforms.py  __init__.py       msa_identifiers.py  parsers.py\ttools\n",
            "errors.py\t    mmcif_parsing.py  msa_pairing.py\t  templates.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/openfold_local/data/tools:\n",
            "__init__.py  jackhmmer.py  utils.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/openfold_local/model:\n",
            "dropout.py   outer_product_mean.py  triangular_attention.py\n",
            "__init__.py  primitives.py\t    triangular_multiplicative_update.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/openfold_local/np:\n",
            "__init__.py  residue_constants.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/openfold_local/utils:\n",
            "all_atom_multimer.py  feats.py\t   kernel\t       tensor_utils.py\n",
            "checkpointing.py      geometry\t   precision_utils.py\n",
            "chunk_utils.py\t      __init__.py  rigid_utils.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/openfold_local/utils/geometry:\n",
            "__init__.py    rigid_matrix_vector.py  test_utils.py  vector.py\n",
            "quat_rigid.py  rotation_matrix.py      utils.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/openfold_local/utils/kernel:\n",
            "attention_core.py  csrc  __init__.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/openfold_local/utils/kernel/csrc:\n",
            "compat.h  softmax_cuda.cpp  softmax_cuda_kernel.cu  softmax_cuda_stub.cpp\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/utils:\n",
            "cropping.py\tgeometry.py  lr_scheduler.py  scatter_utils.py\ttraining.py\n",
            "distributed.py\t__init__.py  metrics.py       seed.py\n",
            "file_io.py\tlogger.py    permutation      torch_utils.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/utils/permutation:\n",
            "atom_permutation.py  chain_permutation\t__init__.py  permutation.py  utils.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/utils/permutation/chain_permutation:\n",
            "heuristic.py  __init__.py  pocket_based_permutation.py\tutils.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/protenix/web_service:\n",
            "colab_request_parser.py  dependency_url.py  prediction_visualization.py\n",
            "colab_request_utils.py\t __init__.py\t    viewer.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/runner:\n",
            "batch_inference.py  ema.py\t  __init__.py\t train.py\n",
            "dumper.py\t    inference.py  msa_search.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/scripts:\n",
            "colabfold_msa.py  gen_ccd_cache.py  __init__.py  msa  prepare_training_data.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/scripts/msa:\n",
            "data\t\t       step3-uniref_add_taxid.py\n",
            "step1-get_prot_seq.py  step4-split_msa_to_uniref_and_others.py\n",
            "step2-get_msa.ipynb    utils.py\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/scripts/msa/data:\n",
            "mmcif  mmcif_msa  mmcif_msa_initial  mmcif_msa_with_taxid  pdb_seqs\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/scripts/msa/data/mmcif:\n",
            "102m.cif  1k1a.cif  5zyh.cif  7zzx.cif\t8f9h.cif\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa:\n",
            "0  1  2  3\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa/0:\n",
            "mmseqs_other_hits.a3m  uniref100_hits.a3m\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa/1:\n",
            "mmseqs_other_hits.a3m  uniref100_hits.a3m\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa/2:\n",
            "mmseqs_other_hits.a3m  uniref100_hits.a3m\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa/3:\n",
            "mmseqs_other_hits.a3m  uniref100_hits.a3m\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa_initial:\n",
            "0.a3m  1.a3m  2.a3m  3.a3m  pdb70_220313_db.m8\tuniref_tax.m8\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/scripts/msa/data/mmcif_msa_with_taxid:\n",
            "0.a3m  1.a3m  2.a3m  3.a3m\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/scripts/msa/data/pdb_seqs:\n",
            "pdb_index_to_seq.json  pdb_seq.fasta\t\t     seq_to_pdb_index.json\n",
            "pdb_seq.csv\t       seq_to_pdb_id_entity_id.json\n",
            "\n",
            "/content/required_presets/Protenix/build/lib/tests:\n",
            "__init__.py\t\t       test_local_rearrange.py\n",
            "test_attention_pair_bias.py    test_lr_schedule.py\n",
            "test_condition_transition.py   test_utils.py\n",
            "test_diffusion_transformer.py  test_weighted_rigid_align.py\n",
            "test_frame.py\n",
            "\n",
            "/content/required_presets/Protenix/configs:\n",
            "configs_base.py  configs_data.py  configs_inference.py\t__init__.py\n",
            "\n",
            "/content/required_presets/Protenix/docs:\n",
            "colabfold_compatiable_msa.md  model_train_inference_cost.md\n",
            "docker_installation.md\t      msa_pipeline.md\n",
            "infer_json_format.md\t      prepare_training_data.md\n",
            "kernels.md\t\t      training.md\n",
            "\n",
            "/content/required_presets/Protenix/examples:\n",
            "7pzb\t  7r6r\tdimer.fasta   example_without_msa.json\tligands\n",
            "7pzb.pdb  7wux\texample.json  finetune_subset.txt\tprot.fasta\n",
            "\n",
            "/content/required_presets/Protenix/examples/7pzb:\n",
            "msa\n",
            "\n",
            "/content/required_presets/Protenix/examples/7pzb/msa:\n",
            "1\n",
            "\n",
            "/content/required_presets/Protenix/examples/7pzb/msa/1:\n",
            "non_pairing.a3m  pairing.a3m\n",
            "\n",
            "/content/required_presets/Protenix/examples/7r6r:\n",
            "msa\n",
            "\n",
            "/content/required_presets/Protenix/examples/7r6r/msa:\n",
            "1\n",
            "\n",
            "/content/required_presets/Protenix/examples/7r6r/msa/1:\n",
            "non_pairing.a3m  pairing.a3m\n",
            "\n",
            "/content/required_presets/Protenix/examples/7wux:\n",
            "msa\n",
            "\n",
            "/content/required_presets/Protenix/examples/7wux/msa:\n",
            "1  2\n",
            "\n",
            "/content/required_presets/Protenix/examples/7wux/msa/1:\n",
            "non_pairing.a3m  pairing.a3m\n",
            "\n",
            "/content/required_presets/Protenix/examples/7wux/msa/2:\n",
            "non_pairing.a3m  pairing.a3m\n",
            "\n",
            "/content/required_presets/Protenix/examples/ligands:\n",
            "7wux_smiles.smi  compounds-3d-R.sdf  compounds-3d-RS.sdf\n",
            "\n",
            "/content/required_presets/Protenix/notebooks:\n",
            "protenix_inference.ipynb\n",
            "\n",
            "/content/required_presets/Protenix/protenix:\n",
            "config\tdata  __init__.py  metrics  model  openfold_local  utils  web_service\n",
            "\n",
            "/content/required_presets/Protenix/protenix/config:\n",
            "config.py  extend_types.py  __init__.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/data:\n",
            "ccd.py\t\t  filter.py\t\t  msa_featurizer.py\n",
            "constants.py\t  infer_data_pipeline.py  msa_utils.py\n",
            "dataloader.py\t  __init__.py\t\t  parser.py\n",
            "data_pipeline.py  json_maker.py\t\t  substructure_perms.py\n",
            "dataset.py\t  json_parser.py\t  tokenizer.py\n",
            "featurizer.py\t  json_to_feature.py\t  utils.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/metrics:\n",
            "clash.py  __init__.py  lddt_metrics.py\trmsd.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/model:\n",
            "generator.py  layer_norm  modules      sample_confidence.py\n",
            "__init__.py   loss.py\t  protenix.py  utils.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/model/layer_norm:\n",
            "__init__.py  kernel  layer_norm.py  torch_ext_compile.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/model/layer_norm/kernel:\n",
            "compat.h  layer_norm_cuda.cpp  layer_norm_cuda_kernel.cu  type_shim.h\n",
            "\n",
            "/content/required_presets/Protenix/protenix/model/modules:\n",
            "confidence.py  embedders.py  head.py\t  pairformer.py  transformer.py\n",
            "diffusion.py   frames.py     __init__.py  primitives.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/openfold_local:\n",
            "data  __init__.py  model  np  README.md  utils\n",
            "\n",
            "/content/required_presets/Protenix/protenix/openfold_local/data:\n",
            "data_transforms.py  __init__.py       msa_identifiers.py  parsers.py\ttools\n",
            "errors.py\t    mmcif_parsing.py  msa_pairing.py\t  templates.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/openfold_local/data/tools:\n",
            "__init__.py  jackhmmer.py  utils.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/openfold_local/model:\n",
            "dropout.py   outer_product_mean.py  triangular_attention.py\n",
            "__init__.py  primitives.py\t    triangular_multiplicative_update.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/openfold_local/np:\n",
            "__init__.py  residue_constants.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/openfold_local/utils:\n",
            "all_atom_multimer.py  feats.py\t   kernel\t       tensor_utils.py\n",
            "checkpointing.py      geometry\t   precision_utils.py\n",
            "chunk_utils.py\t      __init__.py  rigid_utils.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/openfold_local/utils/geometry:\n",
            "__init__.py    rigid_matrix_vector.py  test_utils.py  vector.py\n",
            "quat_rigid.py  rotation_matrix.py      utils.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/openfold_local/utils/kernel:\n",
            "attention_core.py  csrc  __init__.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/openfold_local/utils/kernel/csrc:\n",
            "compat.h  softmax_cuda.cpp  softmax_cuda_kernel.cu  softmax_cuda_stub.cpp\n",
            "\n",
            "/content/required_presets/Protenix/protenix/utils:\n",
            "cropping.py\tgeometry.py  lr_scheduler.py  scatter_utils.py\ttraining.py\n",
            "distributed.py\t__init__.py  metrics.py       seed.py\n",
            "file_io.py\tlogger.py    permutation      torch_utils.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/utils/permutation:\n",
            "atom_permutation.py  chain_permutation\t__init__.py  permutation.py  utils.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/utils/permutation/chain_permutation:\n",
            "heuristic.py  __init__.py  pocket_based_permutation.py\tutils.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix/web_service:\n",
            "colab_request_parser.py  dependency_url.py  prediction_visualization.py\n",
            "colab_request_utils.py\t __init__.py\t    viewer.py\n",
            "\n",
            "/content/required_presets/Protenix/protenix.egg-info:\n",
            "dependency_links.txt  PKG-INFO\t    SOURCES.txt\n",
            "entry_points.txt      requires.txt  top_level.txt\n",
            "\n",
            "/content/required_presets/Protenix/runner:\n",
            "batch_inference.py  ema.py\t  __init__.py\t train.py\n",
            "dumper.py\t    inference.py  msa_search.py\n",
            "\n",
            "/content/required_presets/Protenix/scripts:\n",
            "colabfold_msa.py  gen_ccd_cache.py  __init__.py  msa  prepare_training_data.py\n",
            "\n",
            "/content/required_presets/Protenix/scripts/msa:\n",
            "data\t\t       step3-uniref_add_taxid.py\n",
            "step1-get_prot_seq.py  step4-split_msa_to_uniref_and_others.py\n",
            "step2-get_msa.ipynb    utils.py\n",
            "\n",
            "/content/required_presets/Protenix/scripts/msa/data:\n",
            "mmcif  mmcif_msa  mmcif_msa_initial  mmcif_msa_with_taxid  pdb_seqs\n",
            "\n",
            "/content/required_presets/Protenix/scripts/msa/data/mmcif:\n",
            "102m.cif  1k1a.cif  5zyh.cif  7zzx.cif\t8f9h.cif\n",
            "\n",
            "/content/required_presets/Protenix/scripts/msa/data/mmcif_msa:\n",
            "0  1  2  3\n",
            "\n",
            "/content/required_presets/Protenix/scripts/msa/data/mmcif_msa/0:\n",
            "mmseqs_other_hits.a3m  uniref100_hits.a3m\n",
            "\n",
            "/content/required_presets/Protenix/scripts/msa/data/mmcif_msa/1:\n",
            "mmseqs_other_hits.a3m  uniref100_hits.a3m\n",
            "\n",
            "/content/required_presets/Protenix/scripts/msa/data/mmcif_msa/2:\n",
            "mmseqs_other_hits.a3m  uniref100_hits.a3m\n",
            "\n",
            "/content/required_presets/Protenix/scripts/msa/data/mmcif_msa/3:\n",
            "mmseqs_other_hits.a3m  uniref100_hits.a3m\n",
            "\n",
            "/content/required_presets/Protenix/scripts/msa/data/mmcif_msa_initial:\n",
            "0.a3m  1.a3m  2.a3m  3.a3m  pdb70_220313_db.m8\tuniref_tax.m8\n",
            "\n",
            "/content/required_presets/Protenix/scripts/msa/data/mmcif_msa_with_taxid:\n",
            "0.a3m  1.a3m  2.a3m  3.a3m\n",
            "\n",
            "/content/required_presets/Protenix/scripts/msa/data/pdb_seqs:\n",
            "pdb_index_to_seq.json  pdb_seq.fasta\t\t     seq_to_pdb_index.json\n",
            "pdb_seq.csv\t       seq_to_pdb_id_entity_id.json\n",
            "\n",
            "/content/required_presets/Protenix/tests:\n",
            "__init__.py\t\t       test_local_rearrange.py\n",
            "test_attention_pair_bias.py    test_lr_schedule.py\n",
            "test_condition_transition.py   test_utils.py\n",
            "test_diffusion_transformer.py  test_weighted_rigid_align.py\n",
            "test_frame.py\n",
            "\n",
            "/content/required_presets/wheels:\n",
            "absl_py-2.2.2-py3-none-any.whl\n",
            "annotated_types-0.7.0-py3-none-any.whl\n",
            "asttokens-3.0.0-py3-none-any.whl\n",
            "biopython-1.83-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "biotite-1.0.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "biotraj-1.2.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "certifi-2025.1.31-py3-none-any.whl\n",
            "charset_normalizer-3.4.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "click-8.1.8-py3-none-any.whl\n",
            "colorama-0.4.6-py2.py3-none-any.whl\n",
            "comm-0.2.2-py3-none-any.whl\n",
            "contourpy-1.3.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "cycler-0.12.1-py3-none-any.whl\n",
            "decorator-5.2.1-py3-none-any.whl\n",
            "deepspeed-0.16.5-py3-none-any.whl\n",
            "docker_pycreds-0.4.0-py2.py3-none-any.whl\n",
            "einops-0.8.1-py3-none-any.whl\n",
            "exceptiongroup-1.2.2-py3-none-any.whl\n",
            "executing-2.2.0-py2.py3-none-any.whl\n",
            "filelock-3.18.0-py3-none-any.whl\n",
            "fonttools-4.57.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "fsspec-2025.3.2-py3-none-any.whl\n",
            "gitdb-4.0.12-py3-none-any.whl\n",
            "GitPython-3.1.44-py3-none-any.whl\n",
            "hjson-3.1.0-py3-none-any.whl\n",
            "icecream-2.1.4-py3-none-any.whl\n",
            "idna-3.10-py3-none-any.whl\n",
            "ihm-2.4-cp310-cp310-linux_x86_64.whl\n",
            "ipdb-0.13.13-py3-none-any.whl\n",
            "ipython-8.35.0-py3-none-any.whl\n",
            "ipywidgets-8.1.6-py3-none-any.whl\n",
            "jedi-0.19.2-py2.py3-none-any.whl\n",
            "jinja2-3.1.6-py3-none-any.whl\n",
            "joblib-1.4.2-py3-none-any.whl\n",
            "jupyterlab_widgets-3.0.14-py3-none-any.whl\n",
            "kiwisolver-1.4.8-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl\n",
            "MarkupSafe-3.0.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "matplotlib-3.9.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "matplotlib_inline-0.1.7-py3-none-any.whl\n",
            "ml_collections-1.0.0-py3-none-any.whl\n",
            "modelcif-0.7-py3-none-any.whl\n",
            "mpmath-1.3.0-py3-none-any.whl\n",
            "msgpack-1.1.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "networkx-3.4.2-py3-none-any.whl\n",
            "ninja-1.11.1.4-py3-none-manylinux_2_12_x86_64.manylinux2010_x86_64.whl\n",
            "numpy-1.26.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl\n",
            "nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl\n",
            "nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl\n",
            "nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl\n",
            "nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl\n",
            "nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl\n",
            "nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl\n",
            "nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl\n",
            "nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl\n",
            "nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl\n",
            "nvidia_nvjitlink_cu12-12.8.93-py3-none-manylinux2010_x86_64.manylinux_2_12_x86_64.whl\n",
            "nvidia_nvtx_cu12-12.1.105-py3-none-manylinux1_x86_64.whl\n",
            "optree-0.15.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "packaging-24.2-py3-none-any.whl\n",
            "pandas-2.2.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "parso-0.8.4-py2.py3-none-any.whl\n",
            "pexpect-4.9.0-py2.py3-none-any.whl\n",
            "pillow-11.2.1-cp310-cp310-manylinux_2_28_x86_64.whl\n",
            "platformdirs-4.3.7-py3-none-any.whl\n",
            "prompt_toolkit-3.0.50-py3-none-any.whl\n",
            "protenix-0.4.6-py3-none-any.whl\n",
            "protobuf-3.20.2-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl\n",
            "psutil-7.0.0-cp36-abi3-manylinux_2_12_x86_64.manylinux2010_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "ptyprocess-0.7.0-py2.py3-none-any.whl\n",
            "pure_eval-0.2.3-py3-none-any.whl\n",
            "py3Dmol-2.4.2-py2.py3-none-any.whl\n",
            "py_cpuinfo-9.0.0-py3-none-any.whl\n",
            "pydantic-2.11.3-py3-none-any.whl\n",
            "pydantic_core-2.33.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "pygments-2.19.1-py3-none-any.whl\n",
            "pyparsing-3.2.3-py3-none-any.whl\n",
            "python_dateutil-2.9.0.post0-py2.py3-none-any.whl\n",
            "pytz-2025.2-py2.py3-none-any.whl\n",
            "PyYAML-6.0.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "rdkit-2024.9.6-cp310-cp310-manylinux_2_28_x86_64.whl\n",
            "requests-2.32.3-py3-none-any.whl\n",
            "scikit_learn-1.6.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "scikit_learn_extra-0.3.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "scipy-1.15.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "sentry_sdk-2.26.0-py2.py3-none-any.whl\n",
            "setproctitle-1.3.5-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "setuptools-78.1.0-py3-none-any.whl\n",
            "six-1.17.0-py2.py3-none-any.whl\n",
            "smmap-5.0.2-py3-none-any.whl\n",
            "stack_data-0.6.3-py3-none-any.whl\n",
            "sympy-1.13.3-py3-none-any.whl\n",
            "threadpoolctl-3.6.0-py3-none-any.whl\n",
            "tomli-2.2.1-py3-none-any.whl\n",
            "torch-2.3.1-cp310-cp310-manylinux1_x86_64.whl\n",
            "tqdm-4.67.1-py3-none-any.whl\n",
            "traitlets-5.14.3-py3-none-any.whl\n",
            "triton-2.3.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "typing_extensions-4.13.2-py3-none-any.whl\n",
            "typing_inspection-0.4.0-py3-none-any.whl\n",
            "tzdata-2025.2-py2.py3-none-any.whl\n",
            "urllib3-2.4.0-py3-none-any.whl\n",
            "wandb-0.19.9-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "wcwidth-0.2.13-py2.py3-none-any.whl\n",
            "widgetsnbextension-4.0.14-py3-none-any.whl\n",
            "\n",
            "/content/sample_data:\n",
            "anscombe.json\t\t      mnist_test.csv\n",
            "california_housing_test.csv   mnist_train_small.csv\n",
            "california_housing_train.csv  README.md\n",
            "\n",
            "/content/stanford_rna_3d_folding:\n",
            "MSA\t\t       train_labels.csv       validation_sequences.csv\n",
            "sample_submission.csv  train_sequences.csv\n",
            "test_sequences.csv     validation_labels.csv\n",
            "\n",
            "/content/stanford_rna_3d_folding/MSA:\n",
            "17RA_A.MSA.fasta  2FEY_A.MSA.fasta  3WC2_P.MSA.fasta\t6Z6B_UUU.MSA.fasta\n",
            "1A1T_B.MSA.fasta  2FY1_B.MSA.fasta  3WC2_Q.MSA.fasta\t6ZDU_C.MSA.fasta\n",
            "1A4T_A.MSA.fasta  2G1W_A.MSA.fasta  3WFQ_C.MSA.fasta\t6ZVK_d2.MSA.fasta\n",
            "1A51_A.MSA.fasta  2GIO_A.MSA.fasta  3WFR_B.MSA.fasta\t6ZVK_E1.MSA.fasta\n",
            "1A60_A.MSA.fasta  2GIP_A.MSA.fasta  3WFR_D.MSA.fasta\t6ZVK_e2.MSA.fasta\n",
            "1A9N_Q.MSA.fasta  2GO5_9.MSA.fasta  3WQZ_C.MSA.fasta\t6ZVK_h2.MSA.fasta\n",
            "1AFX_A.MSA.fasta  2GO5_A.MSA.fasta  484D_B.MSA.fasta\t6ZVK_K3.MSA.fasta\n",
            "1ANR_A.MSA.fasta  2GRW_A.MSA.fasta  4A4S_A.MSA.fasta\t7A01_e2.MSA.fasta\n",
            "1AQO_A.MSA.fasta  2GV3_A.MSA.fasta  4A4T_A.MSA.fasta\t7A01_K3.MSA.fasta\n",
            "1ATO_A.MSA.fasta  2GV4_A.MSA.fasta  4A4U_A.MSA.fasta\t7B9V_2.MSA.fasta\n",
            "1ATV_A.MSA.fasta  2GVO_A.MSA.fasta  4BY9_A.MSA.fasta\t7B9V_5.MSA.fasta\n",
            "1ATW_A.MSA.fasta  2HEM_A.MSA.fasta  4C4Q_N.MSA.fasta\t7D1A_A.MSA.fasta\n",
            "1AUD_B.MSA.fasta  2HGH_B.MSA.fasta  4D5N_X.MSA.fasta\t7D8C_B.MSA.fasta\n",
            "1B36_A.MSA.fasta  2HNS_A.MSA.fasta  4DR5_W.MSA.fasta\t7DD4_A.MSA.fasta\n",
            "1BAU_B.MSA.fasta  2HUA_A.MSA.fasta  4ILL_C.MSA.fasta\t7DLZ_Y.MSA.fasta\n",
            "1BGZ_A.MSA.fasta  2IHX_B.MSA.fasta  4ILM_M.MSA.fasta\t7ELE_G.MSA.fasta\n",
            "1BIV_A.MSA.fasta  2IL9_A.MSA.fasta  4KJI_D.MSA.fasta\t7EXY_A.MSA.fasta\n",
            "1BVJ_A.MSA.fasta  2IL9_M.MSA.fasta  4KR6_D.MSA.fasta\t7JRS_B.MSA.fasta\n",
            "1BZ2_A.MSA.fasta  2IXY_A.MSA.fasta  4OQ9_3.MSA.fasta\t7JRT_B.MSA.fasta\n",
            "1CQ5_A.MSA.fasta  2IXZ_A.MSA.fasta  4P8Z_A.MSA.fasta\t7K4L_A.MSA.fasta\n",
            "1D0U_A.MSA.fasta  2IY3_B.MSA.fasta  4QIL_C.MSA.fasta\t7KUB_A.MSA.fasta\n",
            "1D6K_B.MSA.fasta  2IZ8_R.MSA.fasta  4TUA_XY.MSA.fasta\t7KUC_A.MSA.fasta\n",
            "1E4P_A.MSA.fasta  2IZ8_S.MSA.fasta  4TUC_QY.MSA.fasta\t7KUD_A.MSA.fasta\n",
            "1E7K_D.MSA.fasta  2IZM_S.MSA.fasta  4TZP_C.MSA.fasta\t7LHD_A.MSA.fasta\n",
            "1E95_A.MSA.fasta  2IZN_S.MSA.fasta  4TZV_B.MSA.fasta\t7LJY_B.MSA.fasta\n",
            "1EBQ_A.MSA.fasta  2J28_8.MSA.fasta  4V4G_CA.MSA.fasta\t7LMA_B.MSA.fasta\n",
            "1EBR_A.MSA.fasta  2J28_A.MSA.fasta  4V4N_A1.MSA.fasta\t7LMB_B.MSA.fasta\n",
            "1EIY_C.MSA.fasta  2J37_A.MSA.fasta  4V4N_A3.MSA.fasta\t7LVA_A.MSA.fasta\n",
            "1EKZ_B.MSA.fasta  2JPP_C.MSA.fasta  4V4N_B1.MSA.fasta\t7LYG_A.MSA.fasta\n",
            "1ESH_A.MSA.fasta  2JR4_A.MSA.fasta  4V4N_B2.MSA.fasta\t7LYS_B.MSA.fasta\n",
            "1ESY_A.MSA.fasta  2JSE_A.MSA.fasta  4V4V_AA.MSA.fasta\t7LYT_B.MSA.fasta\n",
            "1ETF_A.MSA.fasta  2JTP_A.MSA.fasta  4V4V_B0.MSA.fasta\t7M2T_ss.MSA.fasta\n",
            "1EUQ_B.MSA.fasta  2JWV_A.MSA.fasta  4V4W_B9.MSA.fasta\t7M57_cc.MSA.fasta\n",
            "1EXY_A.MSA.fasta  2JXV_A.MSA.fasta  4V5F_CA.MSA.fasta\t7M57_ii.MSA.fasta\n",
            "1F5U_B.MSA.fasta  2JYM_A.MSA.fasta  4V5G_BA.MSA.fasta\t7M5O_B.MSA.fasta\n",
            "1F6X_A.MSA.fasta  2K4C_A.MSA.fasta  4V5G_BB.MSA.fasta\t7MSF_S.MSA.fasta\n",
            "1F6Z_A.MSA.fasta  2K5Z_A.MSA.fasta  4V5Z_AF.MSA.fasta\t7NK4_D.MSA.fasta\n",
            "1F84_A.MSA.fasta  2K66_A.MSA.fasta  4V5Z_AH.MSA.fasta\t7NQ4_C.MSA.fasta\n",
            "1F85_A.MSA.fasta  2K95_A.MSA.fasta  4V5Z_BA.MSA.fasta\t7OBQ_1.MSA.fasta\n",
            "1F9L_A.MSA.fasta  2KD8_A.MSA.fasta  4V5Z_BC.MSA.fasta\t7ORI_H.MSA.fasta\n",
            "1FFK_9.MSA.fasta  2KE6_A.MSA.fasta  4V5Z_BH.MSA.fasta\t7ORJ_H.MSA.fasta\n",
            "1FHK_A.MSA.fasta  2KHY_A.MSA.fasta  4V5Z_BK.MSA.fasta\t7ORM_H.MSA.fasta\n",
            "1FJE_A.MSA.fasta  2KMJ_A.MSA.fasta  4V5Z_BL.MSA.fasta\t7OZS_3.MSA.fasta\n",
            "1FOQ_A.MSA.fasta  2KOC_A.MSA.fasta  4V5Z_BM.MSA.fasta\t7PTL_B.MSA.fasta\n",
            "1FQZ_A.MSA.fasta  2KPC_A.MSA.fasta  4V5Z_BP.MSA.fasta\t7PTQ_C.MSA.fasta\n",
            "1FYO_A.MSA.fasta  2KPD_A.MSA.fasta  4V5Z_BQ.MSA.fasta\t7PTS_A.MSA.fasta\n",
            "1G70_A.MSA.fasta  2KPV_A.MSA.fasta  4V5Z_BU.MSA.fasta\t7QDU_Q.MSA.fasta\n",
            "1HLX_A.MSA.fasta  2KRL_A.MSA.fasta  4V65_A1.MSA.fasta\t7QGG_v.MSA.fasta\n",
            "1HMH_E.MSA.fasta  2KRP_A.MSA.fasta  4V65_BB.MSA.fasta\t7RGU_H.MSA.fasta\n",
            "1HS1_A.MSA.fasta  2KUR_A.MSA.fasta  4V6D_AV.MSA.fasta\t7RQ5_A.MSA.fasta\n",
            "1HS2_A.MSA.fasta  2KUU_A.MSA.fasta  4V6U_A0.MSA.fasta\t7RYG_B.MSA.fasta\n",
            "1HS3_A.MSA.fasta  2KUV_A.MSA.fasta  4V6U_B1.MSA.fasta\t7S36_R.MSA.fasta\n",
            "1HS4_A.MSA.fasta  2KUW_A.MSA.fasta  4V6W_A5.MSA.fasta\t7S38_R.MSA.fasta\n",
            "1HS8_A.MSA.fasta  2KVN_A.MSA.fasta  4V6W_A7.MSA.fasta\t7S3H_R.MSA.fasta\n",
            "1HVU_I.MSA.fasta  2KZL_A.MSA.fasta  4V6W_A8.MSA.fasta\t7S4U_B.MSA.fasta\n",
            "1HWQ_A.MSA.fasta  2L1F_A.MSA.fasta  4V6W_B2.MSA.fasta\t7S4V_B.MSA.fasta\n",
            "1I3X_A.MSA.fasta  2L1F_B.MSA.fasta  4V6W_BC.MSA.fasta\t7SAM_A.MSA.fasta\n",
            "1I46_A.MSA.fasta  2L2J_A.MSA.fasta  4V6X_A5.MSA.fasta\t7SHX_A.MSA.fasta\n",
            "1I4C_A.MSA.fasta  2L3C_B.MSA.fasta  4V6X_A7.MSA.fasta\t7SLP_R.MSA.fasta\n",
            "1IBM_Y.MSA.fasta  2L3E_A.MSA.fasta  4V6X_A8.MSA.fasta\t7U2A_C.MSA.fasta\n",
            "1IDV_A.MSA.fasta  2L3J_B.MSA.fasta  4V6X_B2.MSA.fasta\t7UGA_A.MSA.fasta\n",
            "1IK1_A.MSA.fasta  2L5Z_A.MSA.fasta  4V6X_BC.MSA.fasta\t7UMC_A.MSA.fasta\n",
            "1IKD_A.MSA.fasta  2L6I_A.MSA.fasta  4V7E_Aa.MSA.fasta\t7UMD_A.MSA.fasta\n",
            "1JO7_A.MSA.fasta  2LA5_A.MSA.fasta  4V7E_Ab.MSA.fasta\t7UME_A.MSA.fasta\n",
            "1JOX_A.MSA.fasta  2LAC_A.MSA.fasta  4V7E_Ac.MSA.fasta\t7UO0_C.MSA.fasta\n",
            "1JTJ_A.MSA.fasta  2LBJ_A.MSA.fasta  4V7E_Ad.MSA.fasta\t7UO1_C.MSA.fasta\n",
            "1JTW_A.MSA.fasta  2LBK_A.MSA.fasta  4V7E_Ae.MSA.fasta\t7UR5_y.MSA.fasta\n",
            "1JUR_A.MSA.fasta  2LBL_A.MSA.fasta  4V8J_AV.MSA.fasta\t7URM_v.MSA.fasta\n",
            "1JWC_A.MSA.fasta  2LBS_A.MSA.fasta  4V8J_CW.MSA.fasta\t7UVT_A.MSA.fasta\n",
            "1JZC_A.MSA.fasta  2LC8_A.MSA.fasta  4V99_FX.MSA.fasta\t7V06_A.MSA.fasta\n",
            "1K2G_A.MSA.fasta  2LDL_A.MSA.fasta  4V9B_CD.MSA.fasta\t7V59_C.MSA.fasta\n",
            "1K4A_A.MSA.fasta  2LDT_A.MSA.fasta  4WC3_B.MSA.fasta\t7VTN_B.MSA.fasta\n",
            "1K4B_A.MSA.fasta  2LHP_A.MSA.fasta  4WJ3_Q.MSA.fasta\t7VW3_B.MSA.fasta\n",
            "1K5I_A.MSA.fasta  2LI4_A.MSA.fasta  4WSA_V.MSA.fasta\t7WB1_D.MSA.fasta\n",
            "1K6G_A.MSA.fasta  2LJJ_A.MSA.fasta  4WZJ_VV.MSA.fasta\t7X34_3.MSA.fasta\n",
            "1K6H_A.MSA.fasta  2LK3_A.MSA.fasta  4X0A_B.MSA.fasta\t7XSN_N.MSA.fasta\n",
            "1K9W_A.MSA.fasta  2LKR_A.MSA.fasta  4X0B_B.MSA.fasta\t7XW2_C.MSA.fasta\n",
            "1KAJ_A.MSA.fasta  2LP9_A.MSA.fasta  4X4P_B.MSA.fasta\t7YOJ_B.MSA.fasta\n",
            "1KIS_A.MSA.fasta  2LPA_A.MSA.fasta  4X4S_B.MSA.fasta\t7YR7_A.MSA.fasta\n",
            "1KIS_B.MSA.fasta  2LQZ_A.MSA.fasta  4YVI_C.MSA.fasta\t7ZAP_B.MSA.fasta\n",
            "1KKA_A.MSA.fasta  2LU0_A.MSA.fasta  4YVJ_C.MSA.fasta\t7ZPQ_6.MSA.fasta\n",
            "1KKS_A.MSA.fasta  2LUB_A.MSA.fasta  4YVK_C.MSA.fasta\t7ZRS_6.MSA.fasta\n",
            "1KP7_A.MSA.fasta  2LUN_A.MSA.fasta  4Z7L_F.MSA.fasta\t7ZRZ_ZN1.MSA.fasta\n",
            "1KPD_A.MSA.fasta  2LV0_A.MSA.fasta  4ZT9_D.MSA.fasta\t8AW3_1.MSA.fasta\n",
            "1L1C_C.MSA.fasta  2M12_A.MSA.fasta  5A18_A.MSA.fasta\t8BD5_B.MSA.fasta\n",
            "1L1W_A.MSA.fasta  2M21_A.MSA.fasta  5AMQ_C.MSA.fasta\t8BTZ_A.MSA.fasta\n",
            "1LC6_A.MSA.fasta  2M22_A.MSA.fasta  5ELS_I.MSA.fasta\t8BU8_A.MSA.fasta\n",
            "1LDZ_A.MSA.fasta  2M23_A.MSA.fasta  5EPI_H.MSA.fasta\t8BWT_A.MSA.fasta\n",
            "1LS2_B.MSA.fasta  2M4W_A.MSA.fasta  5FJ1_H.MSA.fasta\t8C8X_B.MSA.fasta\n",
            "1M5L_A.MSA.fasta  2M57_A.MSA.fasta  5FMZ_V.MSA.fasta\t8C98_A.MSA.fasta\n",
            "1M82_A.MSA.fasta  2M5U_A.MSA.fasta  5GAP_U.MSA.fasta\t8C9C_A.MSA.fasta\n",
            "1MFJ_A.MSA.fasta  2M8K_A.MSA.fasta  5GAP_V.MSA.fasta\t8CI5_C.MSA.fasta\n",
            "1MFK_A.MSA.fasta  2MEQ_A.MSA.fasta  5IEM_A.MSA.fasta\t8CLR_A.MSA.fasta\n",
            "1MFY_A.MSA.fasta  2MF0_G.MSA.fasta  5KH8_A.MSA.fasta\t8CQ1_A.MSA.fasta\n",
            "1MME_D.MSA.fasta  2MFC_B.MSA.fasta  5KK5_B.MSA.fasta\t8CTI_C.MSA.fasta\n",
            "1MNX_A.MSA.fasta  2MFD_A.MSA.fasta  5KMZ_A.MSA.fasta\t8D49_B.MSA.fasta\n",
            "1MT4_A.MSA.fasta  2MFE_D.MSA.fasta  5KQE_A.MSA.fasta\t8D4A_B.MSA.fasta\n",
            "1N34_A.MSA.fasta  2MFF_B.MSA.fasta  5LSN_A.MSA.fasta\t8DC2_B.MSA.fasta\n",
            "1N66_A.MSA.fasta  2MFG_D.MSA.fasta  5MC6_BR.MSA.fasta\t8DK7_C.MSA.fasta\n",
            "1N8X_A.MSA.fasta  2MHI_A.MSA.fasta  5MC6_BS.MSA.fasta\t8DVS_B.MSA.fasta\n",
            "1NA2_A.MSA.fasta  2MI0_A.MSA.fasta  5MC6_m.MSA.fasta\t8E29_B.MSA.fasta\n",
            "1NC0_A.MSA.fasta  2MI0_B.MSA.fasta  5MC6_n.MSA.fasta\t8E79_R.MSA.fasta\n",
            "1NYB_B.MSA.fasta  2MNC_A.MSA.fasta  5MS0_R.MSA.fasta\t8EVQ_EC.MSA.fasta\n",
            "1OQ0_A.MSA.fasta  2MQT_A.MSA.fasta  5MSF_S.MSA.fasta\t8EVR_EC.MSA.fasta\n",
            "1OSW_A.MSA.fasta  2MS1_B.MSA.fasta  5N5C_A.MSA.fasta\t8EWG_B.MSA.fasta\n",
            "1OW9_A.MSA.fasta  2MTJ_A.MSA.fasta  5O7H_A.MSA.fasta\t8EXY_R.MSA.fasta\n",
            "1P5M_A.MSA.fasta  2MXJ_A.MSA.fasta  5OA3_1.MSA.fasta\t8FCS_A.MSA.fasta\n",
            "1P5N_A.MSA.fasta  2MXL_A.MSA.fasta  5UF3_A.MSA.fasta\t8FZT_B.MSA.fasta\n",
            "1P5P_A.MSA.fasta  2N0R_A.MSA.fasta  5UZT_A.MSA.fasta\t8G1I_B.MSA.fasta\n",
            "1P6V_B.MSA.fasta  2N1Q_A.MSA.fasta  5V16_A.MSA.fasta\t8G1S_R.MSA.fasta\n",
            "1P6V_D.MSA.fasta  2N2O_A.MSA.fasta  5V17_A.MSA.fasta\t8GAP_B.MSA.fasta\n",
            "1PJY_A.MSA.fasta  2N2P_A.MSA.fasta  5V6X_C.MSA.fasta\t8H69_5.MSA.fasta\n",
            "1Q75_A.MSA.fasta  2N3O_B.MSA.fasta  5V6X_D.MSA.fasta\t8H6E_4A.MSA.fasta\n",
            "1QC8_A.MSA.fasta  2N3Q_A.MSA.fasta  5V93_a.MSA.fasta\t8H6E_5A.MSA.fasta\n",
            "1QFQ_A.MSA.fasta  2N4L_A.MSA.fasta  5V93_B.MSA.fasta\t8HUD_B.MSA.fasta\n",
            "1QWA_A.MSA.fasta  2N6S_A.MSA.fasta  5WQ1_A.MSA.fasta\t8I0P_F.MSA.fasta\n",
            "1QWB_A.MSA.fasta  2N6T_A.MSA.fasta  5XBL_B.MSA.fasta\t8IDF_B.MSA.fasta\n",
            "1QZA_B.MSA.fasta  2N6W_A.MSA.fasta  5XJ2_G.MSA.fasta\t8IEW_B.MSA.fasta\n",
            "1QZB_B.MSA.fasta  2N6X_A.MSA.fasta  5Y36_B.MSA.fasta\t8J72_C.MSA.fasta\n",
            "1QZC_B.MSA.fasta  2N7M_X.MSA.fasta  5Z3G_B.MSA.fasta\t8JHP_A.MSA.fasta\n",
            "1QZC_C.MSA.fasta  2N7X_A.MSA.fasta  5ZAL_C.MSA.fasta\t8JTJ_B.MSA.fasta\n",
            "1R2P_A.MSA.fasta  2N8V_X.MSA.fasta  5ZAM_C.MSA.fasta\t8JTR_B.MSA.fasta\n",
            "1R2W_C.MSA.fasta  2NBX_A.MSA.fasta  5ZEY_B.MSA.fasta\t8K8B_A.MSA.fasta\n",
            "1R4H_A.MSA.fasta  2NBY_A.MSA.fasta  6AAX_D.MSA.fasta\t8KAG_A.MSA.fasta\n",
            "1R7W_A.MSA.fasta  2NBZ_A.MSA.fasta  6AEB_E.MSA.fasta\t8KAI_I.MSA.fasta\n",
            "1R7Z_A.MSA.fasta  2NC0_A.MSA.fasta  6AGB_A.MSA.fasta\t8KAM_A.MSA.fasta\n",
            "1RFR_A.MSA.fasta  2NC1_A.MSA.fasta  6AH3_T.MSA.fasta\t8KDA_O.MSA.fasta\n",
            "1RHT_A.MSA.fasta  2NCI_A.MSA.fasta  6AHU_A.MSA.fasta\t8KI7_E.MSA.fasta\n",
            "1RNG_A.MSA.fasta  2NOQ_A.MSA.fasta  6AHU_T.MSA.fasta\t8OKD_X.MSA.fasta\n",
            "1RNK_A.MSA.fasta  2NOQ_B.MSA.fasta  6AWC_A.MSA.fasta\t8OMR_C.MSA.fasta\n",
            "1ROQ_A.MSA.fasta  2NOQ_E.MSA.fasta  6AWD_A.MSA.fasta\t8OPP_D.MSA.fasta\n",
            "1RY1_E.MSA.fasta  2NR0_E.MSA.fasta  6BZ7_QW.MSA.fasta\t8OPS_C.MSA.fasta\n",
            "1S34_A.MSA.fasta  2NR0_F.MSA.fasta  6BZ7_XY.MSA.fasta\t8OPT_D.MSA.fasta\n",
            "1S9S_A.MSA.fasta  2NR0_G.MSA.fasta  6BZ8_XY.MSA.fasta\t8OST_C.MSA.fasta\n",
            "1SCL_A.MSA.fasta  2NR0_H.MSA.fasta  6C4H_x.MSA.fasta\t8P0B_V.MSA.fasta\n",
            "1SLO_A.MSA.fasta  2O33_A.MSA.fasta  6C66_J.MSA.fasta\t8P0G_V.MSA.fasta\n",
            "1SZY_A.MSA.fasta  2OB7_D.MSA.fasta  6CYT_N.MSA.fasta\t8PM4_B.MSA.fasta\n",
            "1T28_A.MSA.fasta  2OJ8_A.MSA.fasta  6DTI_X.MSA.fasta\t8QO2_A.MSA.fasta\n",
            "1T4L_A.MSA.fasta  2OM3_R.MSA.fasta  6DU5_B.MSA.fasta\t8QO3_A.MSA.fasta\n",
            "1TBK_A.MSA.fasta  2OM7_A.MSA.fasta  6EVJ_V.MSA.fasta\t8QO4_A.MSA.fasta\n",
            "1TJZ_A.MSA.fasta  2OM7_C.MSA.fasta  6FHI_V.MSA.fasta\t8QO5_A.MSA.fasta\n",
            "1TLR_A.MSA.fasta  2OM7_F.MSA.fasta  6FLQ_R.MSA.fasta\t8SA6_A.MSA.fasta\n",
            "1TXS_A.MSA.fasta  2OM7_G.MSA.fasta  6FT6_2.MSA.fasta\t8SCF_A.MSA.fasta\n",
            "1U63_D.MSA.fasta  2OM7_H.MSA.fasta  6GBM_A.MSA.fasta\t8SCH_A.MSA.fasta\n",
            "1ULL_A.MSA.fasta  2OM7_I.MSA.fasta  6HYK_A.MSA.fasta\t8SFH_B.MSA.fasta\n",
            "1UUU_A.MSA.fasta  2OM7_J.MSA.fasta  6HYU_D.MSA.fasta\t8SFI_B.MSA.fasta\n",
            "1VOP_A.MSA.fasta  2PCW_A.MSA.fasta  6IFO_D.MSA.fasta\t8SFJ_B.MSA.fasta\n",
            "1WKS_A.MSA.fasta  2PN9_A.MSA.fasta  6IV6_G.MSA.fasta\t8SFL_B.MSA.fasta\n",
            "1WZ2_D.MSA.fasta  2QH2_A.MSA.fasta  6K0A_X.MSA.fasta\t8SFN_B.MSA.fasta\n",
            "1X18_A.MSA.fasta  2QH3_A.MSA.fasta  6K0B_U.MSA.fasta\t8SFO_B.MSA.fasta\n",
            "1X18_B.MSA.fasta  2QH4_A.MSA.fasta  6K3Z_A.MSA.fasta\t8T29_R.MSA.fasta\n",
            "1X18_D.MSA.fasta  2R1G_A.MSA.fasta  6K4S_A.MSA.fasta\t8T2A_R.MSA.fasta\n",
            "1XHP_A.MSA.fasta  2R1G_B.MSA.fasta  6KUT_V.MSA.fasta\t8T2B_R.MSA.fasta\n",
            "1XSG_A.MSA.fasta  2R1G_C.MSA.fasta  6KUV_V.MSA.fasta\t8T2O_R.MSA.fasta\n",
            "1XSH_A.MSA.fasta  2R1G_E.MSA.fasta  6LTP_H.MSA.fasta\t8T2P_A.MSA.fasta\n",
            "1XWP_A.MSA.fasta  2R1G_F.MSA.fasta  6MCE_A.MSA.fasta\t8T2P_B.MSA.fasta\n",
            "1XWU_A.MSA.fasta  2R1G_X.MSA.fasta  6MJ0_B.MSA.fasta\t8T2Y_tA.MSA.fasta\n",
            "1Y1Y_P.MSA.fasta  2R93_R.MSA.fasta  6MXQ_A.MSA.fasta\t8T2Y_tP.MSA.fasta\n",
            "1YLG_A.MSA.fasta  2RLU_A.MSA.fasta  6N7R_R.MSA.fasta\t8T3E_EC.MSA.fasta\n",
            "1YMO_A.MSA.fasta  2RN1_B.MSA.fasta  6NDK_XY.MSA.fasta\t8T3F_EC.MSA.fasta\n",
            "1YNE_A.MSA.fasta  2RO2_A.MSA.fasta  6NM9_G.MSA.fasta\t8TOC_R.MSA.fasta\n",
            "1YSH_A.MSA.fasta  2RPK_A.MSA.fasta  6NOA_A.MSA.fasta\t8TVZ_C.MSA.fasta\n",
            "1YSH_B.MSA.fasta  2RPT_A.MSA.fasta  6O0Y_B.MSA.fasta\t8U3M_A.MSA.fasta\n",
            "1YSH_F.MSA.fasta  2RVO_A.MSA.fasta  6O0Z_B.MSA.fasta\t8UAU_R.MSA.fasta\n",
            "1YSV_A.MSA.fasta  2TPK_A.MSA.fasta  6O3M_XV.MSA.fasta\t8UBE_I.MSA.fasta\n",
            "1Z2J_A.MSA.fasta  2U2A_A.MSA.fasta  6OF6_XW.MSA.fasta\t8UO6_B.MSA.fasta\n",
            "1Z30_A.MSA.fasta  2W2H_S.MSA.fasta  6OJ2_QW.MSA.fasta\t8UPT_A.MSA.fasta\n",
            "1Z31_A.MSA.fasta  2X7N_A.MSA.fasta  6OJ2_QY.MSA.fasta\t8URY_A.MSA.fasta\n",
            "1ZBH_E.MSA.fasta  2XXA_F.MSA.fasta  6OPE_XY.MSA.fasta\t8URY_d.MSA.fasta\n",
            "1ZC5_A.MSA.fasta  2Y95_A.MSA.fasta  6ORD_QY.MSA.fasta\t8UYJ_A.MSA.fasta\n",
            "1ZC8_F.MSA.fasta  2ZJQ_X.MSA.fasta  6OSI_RB.MSA.fasta\t8UYL_A.MSA.fasta\n",
            "1ZC8_G.MSA.fasta  2ZJQ_Y.MSA.fasta  6PK9_A.MSA.fasta\t8UYP_A.MSA.fasta\n",
            "1ZC8_H.MSA.fasta  3A2K_C.MSA.fasta  6POM_B.MSA.fasta\t8UYS_A.MSA.fasta\n",
            "1ZC8_I.MSA.fasta  3A3A_A.MSA.fasta  6QW6_5.MSA.fasta\t8VCI_A.MSA.fasta\n",
            "1ZC8_J.MSA.fasta  3AKZ_H.MSA.fasta  6QWL_W.MSA.fasta\t8WT6_E.MSA.fasta\n",
            "1ZC8_Z.MSA.fasta  3AMU_B.MSA.fasta  6QX3_D.MSA.fasta\t8WT6_F.MSA.fasta\n",
            "1ZDI_S.MSA.fasta  3CW1_v.MSA.fasta  6QX9_1.MSA.fasta\t8WT7_E.MSA.fasta\n",
            "1ZDK_S.MSA.fasta  3DEG_E.MSA.fasta  6RFL_U.MSA.fasta\t8WT7_F.MSA.fasta\n",
            "1ZIF_A.MSA.fasta  3DEG_G.MSA.fasta  6RR7_D.MSA.fasta\t8WT8_E.MSA.fasta\n",
            "1ZIG_A.MSA.fasta  3DEG_I.MSA.fasta  6S0X_a.MSA.fasta\t8WT8_F.MSA.fasta\n",
            "1ZIH_A.MSA.fasta  3DEG_J.MSA.fasta  6S0X_B.MSA.fasta\t8WUS_B.MSA.fasta\n",
            "1ZN1_B.MSA.fasta  3DEG_K.MSA.fasta  6S0Z_B.MSA.fasta\t8XCA_B.MSA.fasta\n",
            "28SP_A.MSA.fasta  3DKN_D.MSA.fasta  6SDW_B.MSA.fasta\t8XCC_B.MSA.fasta\n",
            "2A64_A.MSA.fasta  3DKN_E.MSA.fasta  6SY6_D.MSA.fasta\t8XPP_B.MSA.fasta\n",
            "2A9L_A.MSA.fasta  3DKN_F.MSA.fasta  6U79_A.MSA.fasta\t8YE6_C.MSA.fasta\n",
            "2ADT_A.MSA.fasta  3EP2_B.MSA.fasta  6UES_A.MSA.fasta\t8YGJ_B.MSA.fasta\n",
            "2AGN_A.MSA.fasta  3EP2_C.MSA.fasta  6UFM_A.MSA.fasta\t8Z1F_T.MSA.fasta\n",
            "2AGN_C.MSA.fasta  3EP2_D.MSA.fasta  6V5B_D.MSA.fasta\t8Z1G_T.MSA.fasta\n",
            "2AHT_A.MSA.fasta  3EQ3_E.MSA.fasta  6VAR_A.MSA.fasta\t8Z85_D.MSA.fasta\n",
            "2AKE_B.MSA.fasta  3EQ3_Y.MSA.fasta  6VZC_A.MSA.fasta\t8Z8N_D.MSA.fasta\n",
            "2B2E_R.MSA.fasta  3EQ4_A.MSA.fasta  6W3M_A.MSA.fasta\t8Z90_D.MSA.fasta\n",
            "2B2E_S.MSA.fasta  3HAY_E.MSA.fasta  6W5C_B.MSA.fasta\t9CF1_W.MSA.fasta\n",
            "2B6G_B.MSA.fasta  3IYR_A.MSA.fasta  6W62_B.MSA.fasta\t9CF3_W.MSA.fasta\n",
            "2BJ2_A.MSA.fasta  3IZ4_A.MSA.fasta  6W64_B.MSA.fasta\t9DTT_B.MSA.fasta\n",
            "2BJ2_B.MSA.fasta  3IZD_A.MSA.fasta  6WB0_C.MSA.fasta\t9ENC_D.MSA.fasta\n",
            "2BQ5_R.MSA.fasta  3IZZ_E.MSA.fasta  6WB1_C.MSA.fasta\t9ENE_D.MSA.fasta\n",
            "2BQ5_S.MSA.fasta  3J3V_A.MSA.fasta  6WD2_1.MSA.fasta\t9ENF_C.MSA.fasta\n",
            "2BS0_R.MSA.fasta  3J3V_B.MSA.fasta  6WD9_2.MSA.fasta\t9EOW_A.MSA.fasta\n",
            "2BS0_S.MSA.fasta  3J3W_A.MSA.fasta  6WLK_A.MSA.fasta\t9FO8_A.MSA.fasta\n",
            "2D19_A.MSA.fasta  3J6B_e.MSA.fasta  6WLM_A.MSA.fasta\t9FO9_A.MSA.fasta\n",
            "2D1B_B.MSA.fasta  3J8G_A.MSA.fasta  6WLN_A.MSA.fasta\t9G7C_A.MSA.fasta\n",
            "2DER_D.MSA.fasta  3J8G_B.MSA.fasta  6WLO_A.MSA.fasta\t9G8O_S2.MSA.fasta\n",
            "2DET_C.MSA.fasta  3JB9_C.MSA.fasta  6WLS_A.MSA.fasta\tR1107.MSA.fasta\n",
            "2DR2_B.MSA.fasta  3JD5_A.MSA.fasta  6WLT_A.MSA.fasta\tR1108.MSA.fasta\n",
            "2DU4_C.MSA.fasta  3JQ4_B.MSA.fasta  6WPI_C.MSA.fasta\tR1116.MSA.fasta\n",
            "2DU5_D.MSA.fasta  3LWQ_D.MSA.fasta  6WW6_C.MSA.fasta\tR1117v2.MSA.fasta\n",
            "2DU6_D.MSA.fasta  3NDB_M.MSA.fasta  6WW6_E.MSA.fasta\tR1126.MSA.fasta\n",
            "2ESE_B.MSA.fasta  3PGW_N.MSA.fasta  6WW6_F.MSA.fasta\tR1128.MSA.fasta\n",
            "2EUY_A.MSA.fasta  3PHP_A.MSA.fasta  6XRZ_A.MSA.fasta\tR1136.MSA.fasta\n",
            "2EVY_A.MSA.fasta  3PIP_Y.MSA.fasta  6XWJ_A.MSA.fasta\tR1138.MSA.fasta\n",
            "2F4X_B.MSA.fasta  3TUP_T.MSA.fasta  6XWW_A.MSA.fasta\tR1149.MSA.fasta\n",
            "2F87_A.MSA.fasta  3UZS_C.MSA.fasta  6XZD_IN1.MSA.fasta\tR1156.MSA.fasta\n",
            "2F88_A.MSA.fasta  3W1K_F.MSA.fasta  6XZR_IN1.MSA.fasta\tR1189.MSA.fasta\n",
            "2FDT_A.MSA.fasta  3WC1_Q.MSA.fasta  6Y0C_IN1.MSA.fasta\tR1190.MSA.fasta\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip uninstall -y torch torchvision torchaudio\n",
        "!pip install torch==2.1.1+cu118 torchvision==0.16.1+cu118 torchaudio==2.1.1 --index-url https://download.pytorch.org/whl/cu118\n",
        "\n",
        "import torch\n",
        "print(torch.cuda.get_device_name(0))\n",
        "print(\"CUDA available:\", torch.cuda.is_available())\n"
      ],
      "metadata": {
        "id": "VJSyNc2TMJ_P",
        "outputId": "10ec0ec0-2aff-44ba-d56b-3e3d90058240",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Found existing installation: torch 2.3.1+cpu\n",
            "Uninstalling torch-2.3.1+cpu:\n",
            "  Successfully uninstalled torch-2.3.1+cpu\n",
            "Found existing installation: torchvision 0.18.1+cpu\n",
            "Uninstalling torchvision-0.18.1+cpu:\n",
            "  Successfully uninstalled torchvision-0.18.1+cpu\n",
            "Found existing installation: torchaudio 2.3.1+cpu\n",
            "Uninstalling torchaudio-2.3.1+cpu:\n",
            "  Successfully uninstalled torchaudio-2.3.1+cpu\n",
            "Looking in indexes: https://download.pytorch.org/whl/cu118\n",
            "Collecting torch==2.1.1+cu118\n",
            "  Using cached https://download.pytorch.org/whl/cu118/torch-2.1.1%2Bcu118-cp311-cp311-linux_x86_64.whl (2326.0 MB)\n",
            "Collecting torchvision==0.16.1+cu118\n",
            "  Using cached https://download.pytorch.org/whl/cu118/torchvision-0.16.1%2Bcu118-cp311-cp311-linux_x86_64.whl (6.1 MB)\n",
            "Collecting torchaudio==2.1.1\n",
            "  Using cached https://download.pytorch.org/whl/cu118/torchaudio-2.1.1%2Bcu118-cp311-cp311-linux_x86_64.whl (3.2 MB)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.11/dist-packages (from torch==2.1.1+cu118) (3.18.0)\n",
            "Requirement already satisfied: typing-extensions in /usr/local/lib/python3.11/dist-packages (from torch==2.1.1+cu118) (4.13.2)\n",
            "Requirement already satisfied: sympy in /usr/local/lib/python3.11/dist-packages (from torch==2.1.1+cu118) (1.13.1)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.11/dist-packages (from torch==2.1.1+cu118) (3.4.2)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.11/dist-packages (from torch==2.1.1+cu118) (3.1.6)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.11/dist-packages (from torch==2.1.1+cu118) (2025.3.2)\n",
            "Collecting triton==2.1.0 (from torch==2.1.1+cu118)\n",
            "  Using cached https://download.pytorch.org/whl/triton-2.1.0-0-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (89.2 MB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from torchvision==0.16.1+cu118) (2.2.4)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.11/dist-packages (from torchvision==0.16.1+cu118) (2.32.3)\n",
            "Requirement already satisfied: pillow!=8.3.*,>=5.3.0 in /usr/local/lib/python3.11/dist-packages (from torchvision==0.16.1+cu118) (11.1.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2->torch==2.1.1+cu118) (3.0.2)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests->torchvision==0.16.1+cu118) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests->torchvision==0.16.1+cu118) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests->torchvision==0.16.1+cu118) (2.3.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests->torchvision==0.16.1+cu118) (2025.1.31)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy->torch==2.1.1+cu118) (1.3.0)\n",
            "Installing collected packages: triton, torch, torchvision, torchaudio\n",
            "  Attempting uninstall: triton\n",
            "    Found existing installation: triton 2.3.1\n",
            "    Uninstalling triton-2.3.1:\n",
            "      Successfully uninstalled triton-2.3.1\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "protenix 0.4.6 requires numpy==1.26.3, but you have numpy 2.2.4 which is incompatible.\n",
            "protenix 0.4.6 requires torch==2.3.1, but you have torch 2.1.1+cu118 which is incompatible.\u001b[0m\u001b[31m\n",
            "\u001b[0mSuccessfully installed torch-2.1.1+cu118 torchaudio-2.1.1+cu118 torchvision-0.16.1+cu118 triton-2.1.0\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "A module that was compiled using NumPy 1.x cannot be run in\n",
            "NumPy 2.2.4 as it may crash. To support both 1.x and 2.x\n",
            "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
            "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
            "\n",
            "If you are a user of the module, the easiest solution will be to\n",
            "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
            "We expect that some modules will need time to support NumPy 2.\n",
            "\n",
            "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
            "  File \"<frozen runpy>\", line 88, in _run_code\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/colab_kernel_launcher.py\", line 37, in <module>\n",
            "    ColabKernelApp.launch_instance()\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/traitlets/config/application.py\", line 992, in launch_instance\n",
            "    app.start()\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/ipykernel/kernelapp.py\", line 712, in start\n",
            "    self.io_loop.start()\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/tornado/platform/asyncio.py\", line 205, in start\n",
            "    self.asyncio_loop.run_forever()\n",
            "  File \"/usr/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
            "    self._run_once()\n",
            "  File \"/usr/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
            "    handle._run()\n",
            "  File \"/usr/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
            "    self._context.run(self._callback, *self._args)\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/ipykernel/kernelbase.py\", line 510, in dispatch_queue\n",
            "    await self.process_one()\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/ipykernel/kernelbase.py\", line 499, in process_one\n",
            "    await dispatch(*args)\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/ipykernel/kernelbase.py\", line 406, in dispatch_shell\n",
            "    await result\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/ipykernel/kernelbase.py\", line 730, in execute_request\n",
            "    reply_content = await reply_content\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/ipykernel/ipkernel.py\", line 383, in do_execute\n",
            "    res = shell.run_cell(\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/ipykernel/zmqshell.py\", line 528, in run_cell\n",
            "    return super().run_cell(*args, **kwargs)\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py\", line 2975, in run_cell\n",
            "    result = self._run_cell(\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py\", line 3030, in _run_cell\n",
            "    return runner(coro)\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/IPython/core/async_helpers.py\", line 78, in _pseudo_sync_runner\n",
            "    coro.send(None)\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py\", line 3257, in run_cell_async\n",
            "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py\", line 3473, in run_ast_nodes\n",
            "    if (await self.run_code(code, result,  async_=asy)):\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py\", line 3553, in run_code\n",
            "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
            "  File \"<ipython-input-8-41af59469d93>\", line 4, in <cell line: 0>\n",
            "    import torch\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/torch/__init__.py\", line 1382, in <module>\n",
            "    from .functional import *  # noqa: F403\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/torch/functional.py\", line 7, in <module>\n",
            "    import torch.nn.functional as F\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/torch/nn/__init__.py\", line 1, in <module>\n",
            "    from .modules import *  # noqa: F403\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/torch/nn/modules/__init__.py\", line 35, in <module>\n",
            "    from .transformer import TransformerEncoder, TransformerDecoder, \\\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/torch/nn/modules/transformer.py\", line 20, in <module>\n",
            "    device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),\n",
            "/usr/local/lib/python3.11/dist-packages/torch/nn/modules/transformer.py:20: UserWarning: Failed to initialize NumPy: _ARRAY_API not found (Triggered internally at ../torch/csrc/utils/tensor_numpy.cpp:84.)\n",
            "  device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Tesla T4\n",
            "CUDA available: True\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "RNA 3D Structure Prediction and Submission Generator for Kaggle Environment\n",
        "\n",
        "This script:\n",
        "1. Reads RNA sequences from test_sequences.csv\n",
        "2. Creates input JSONs for each sequence\n",
        "3. Runs the Protenix model to predict 3D structures\n",
        "4. Extracts C1' atom coordinates from the output CIF files\n",
        "5. Creates a submission.csv file in the format required\n",
        "\"\"\"\n",
        "\n",
        "import os\n",
        "import json\n",
        "import subprocess\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import glob\n",
        "from tqdm import tqdm\n",
        "from biotite.structure.io import pdbx\n",
        "\n",
        "def create_input_json(sequence, target_id):\n",
        "    \"\"\"\n",
        "    Create the input JSON for a single RNA sequence\n",
        "    \"\"\"\n",
        "    input_json = [{\n",
        "        \"sequences\": [\n",
        "            {\n",
        "                \"rnaSequence\": {\n",
        "                    \"sequence\": sequence,\n",
        "                    \"count\": 1,\n",
        "                    \"modifications\": []\n",
        "                }\n",
        "            }\n",
        "        ],\n",
        "        \"name\": target_id,\n",
        "        \"covalent_bonds\": []\n",
        "    }]\n",
        "    return input_json\n",
        "\n",
        "def run_inference(input_json_path, output_dir, target_id):\n",
        "    \"\"\"\n",
        "    Run inference using the Protenix model with kaggle-specific paths\n",
        "    \"\"\"\n",
        "    # Ensure output directory exists\n",
        "    os.makedirs(output_dir, exist_ok=True)\n",
        "\n",
        "    # Define command with kaggle-specific paths\n",
        "    cmd = [\n",
        "        \"python\", \"/content/required_presets/Protenix/runner/inference.py\",\n",
        "        \"--seeds\", \"42\",\n",
        "        \"--dump_dir\", output_dir,\n",
        "        \"--input_json_path\", input_json_path,\n",
        "        \"--model.N_cycle\", \"10\",\n",
        "        \"--sample_diffusion.N_sample\", \"5\",\n",
        "        \"--sample_diffusion.N_step\", \"200\",\n",
        "        \"--load_checkpoint_path\", \"/content/required_presets/Protenix/release_data/checkpoint/model_v0.2.0.pt\",\n",
        "        \"--use_deepspeed_evo_attention\", \"false\"\n",
        "    ]\n",
        "\n",
        "    print(f\"Running inference for {target_id}\")\n",
        "    try:\n",
        "        result = subprocess.run(cmd, capture_output=True, text=True)\n",
        "        if result.returncode != 0:\n",
        "            print(f\"Error running inference for {target_id}: {result.stderr}\")\n",
        "            return False\n",
        "        return True\n",
        "    except Exception as e:\n",
        "        print(f\"Exception running inference for {target_id}: {e}\")\n",
        "        return False\n",
        "\n",
        "def extract_c1_coordinates(cif_file_path):\n",
        "    \"\"\"\n",
        "    Extract C1' atom coordinates from a CIF file using biotite\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # Read the CIF file using the correct biotite method\n",
        "        with open(cif_file_path, 'r') as f:\n",
        "            cif_data = pdbx.CIFFile.read(f)\n",
        "\n",
        "        # Get structure from CIF data\n",
        "        atom_array = pdbx.get_structure(cif_data, model=1)\n",
        "\n",
        "        # Clean atom names and find C1' atoms\n",
        "        atom_names_clean = np.char.strip(atom_array.atom_name.astype(str))\n",
        "        mask_c1 = atom_names_clean == \"C1'\"\n",
        "        c1_atoms = atom_array[mask_c1]\n",
        "\n",
        "        if len(c1_atoms) == 0:\n",
        "            print(f\"Warning: No C1' atoms found in {cif_file_path}\")\n",
        "            return None\n",
        "\n",
        "        # Sort by residue ID and return coordinates\n",
        "        sort_indices = np.argsort(c1_atoms.res_id)\n",
        "        c1_atoms_sorted = c1_atoms[sort_indices]\n",
        "        c1_coords = c1_atoms_sorted.coord\n",
        "\n",
        "        return c1_coords\n",
        "    except Exception as e:\n",
        "        print(f\"Error extracting C1' coordinates from {cif_file_path}: {e}\")\n",
        "        return None\n",
        "\n",
        "\n",
        "def process_sequence(sequence, target_id, temp_dir, output_dir):\n",
        "    \"\"\"\n",
        "    Process a single RNA sequence and return C1' coordinates\n",
        "    \"\"\"\n",
        "    print(f\"Processing {target_id}: {sequence}\")\n",
        "\n",
        "    # Create input JSON\n",
        "    input_json = create_input_json(sequence, target_id)\n",
        "\n",
        "    # Save JSON to temporary file\n",
        "    os.makedirs(temp_dir, exist_ok=True)\n",
        "    input_json_path = os.path.join(temp_dir, f\"{target_id}_input.json\")\n",
        "    with open(input_json_path, \"w\") as f:\n",
        "        json.dump(input_json, f, indent=4)\n",
        "\n",
        "    # Run inference\n",
        "    success = run_inference(input_json_path, output_dir, target_id)\n",
        "\n",
        "    if not success:\n",
        "        print(f\"Inference failed for {target_id}\")\n",
        "        return None\n",
        "\n",
        "    # Find the CIF files for this target\n",
        "    target_prediction_dir = os.path.join(output_dir, target_id, \"seed_42\", \"predictions\")\n",
        "    if not os.path.exists(target_prediction_dir):\n",
        "        print(f\"Prediction directory not found for {target_id}\")\n",
        "        return None\n",
        "\n",
        "    # Look for CIF files with the pattern {target_id}_seed_42_sample_*.cif\n",
        "    cif_files = sorted(glob.glob(os.path.join(target_prediction_dir, f\"{target_id}_seed_42_sample_*.cif\")))\n",
        "\n",
        "    # If no CIF files found, return None\n",
        "    if not cif_files:\n",
        "        print(f\"No CIF files found for {target_id}\")\n",
        "        return None\n",
        "\n",
        "    print(f\"Found {len(cif_files)} CIF files for {target_id}\")\n",
        "\n",
        "    # Extract C1' coordinates from each CIF file\n",
        "    all_coords = []\n",
        "    for cif_file in cif_files:\n",
        "        coords = extract_c1_coordinates(cif_file)\n",
        "        if coords is not None:\n",
        "            all_coords.append(coords)\n",
        "\n",
        "    if not all_coords:\n",
        "        print(f\"No valid C1' coordinates found for {target_id}\")\n",
        "        return None\n",
        "\n",
        "    # Ensure we have 5 models (if we have fewer, duplicate the last one)\n",
        "    while len(all_coords) < 5:\n",
        "        print(f\"Only {len(all_coords)} models found for {target_id}, duplicating last model\")\n",
        "        all_coords.append(all_coords[-1])\n",
        "\n",
        "    return all_coords[:5]  # Ensure we only have 5 models\n",
        "\n",
        "def create_submission(test_sequences_df, c1_coords_dict, output_file):\n",
        "    \"\"\"\n",
        "    Create the submission CSV file with C1' coordinates\n",
        "    \"\"\"\n",
        "    rows = []\n",
        "\n",
        "    # Process each sequence\n",
        "    for _, row in test_sequences_df.iterrows():\n",
        "        target_id = row['target_id']\n",
        "        sequence = row['sequence']\n",
        "\n",
        "        if target_id not in c1_coords_dict or c1_coords_dict[target_id] is None:\n",
        "            print(f\"No prediction found for {target_id}, using zeros\")\n",
        "            # Create empty predictions (all zeros)\n",
        "            for i, residue in enumerate(sequence):\n",
        "                row_data = {\n",
        "                    'ID': f\"{target_id}_{i+1}\",\n",
        "                    'resname': residue,\n",
        "                    'resid': i+1\n",
        "                }\n",
        "                for model in range(1, 6):\n",
        "                    row_data[f'x_{model}'] = 0.0\n",
        "                    row_data[f'y_{model}'] = 0.0\n",
        "                    row_data[f'z_{model}'] = 0.0\n",
        "                rows.append(row_data)\n",
        "        else:\n",
        "            # Get the 5 models for this target\n",
        "            models = c1_coords_dict[target_id]\n",
        "\n",
        "            # Create a row for each residue\n",
        "            for i, residue in enumerate(sequence):\n",
        "                row_data = {\n",
        "                    'ID': f\"{target_id}_{i+1}\",\n",
        "                    'resname': residue,\n",
        "                    'resid': i+1\n",
        "                }\n",
        "\n",
        "                # Add coordinates for each model\n",
        "                for model_idx in range(5):\n",
        "                    if model_idx < len(models) and i < len(models[model_idx]):\n",
        "                        row_data[f'x_{model_idx+1}'] = models[model_idx][i][0]\n",
        "                        row_data[f'y_{model_idx+1}'] = models[model_idx][i][1]\n",
        "                        row_data[f'z_{model_idx+1}'] = models[model_idx][i][2]\n",
        "                    else:\n",
        "                        # If coordinates are not available, use zeros\n",
        "                        row_data[f'x_{model_idx+1}'] = 0.0\n",
        "                        row_data[f'y_{model_idx+1}'] = 0.0\n",
        "                        row_data[f'z_{model_idx+1}'] = 0.0\n",
        "\n",
        "                rows.append(row_data)\n",
        "\n",
        "    # Create DataFrame and save to CSV\n",
        "    df = pd.DataFrame(rows)\n",
        "    df.to_csv(output_file, index=False)\n",
        "    print(f\"Created submission file: {output_file}\")\n",
        "\n",
        "def main():\n",
        "    \"\"\"\n",
        "    Main function\n",
        "    \"\"\"\n",
        "    # Set up required symlinks for CCD cache as in kaggle_inference.py\n",
        "    os.makedirs(\"/usr/local/lib/python3.10/dist-packages/release_data/ccd_cache\", exist_ok=True)\n",
        "\n",
        "    source_ccd_file = \"/content/required_presets/Protenix/release_data/ccd_cache/components.v20240608.cif\"\n",
        "    target_ccd_file = \"/usr/local/lib/python3.10/dist-packages/release_data/ccd_cache/components.v20240608.cif\"\n",
        "\n",
        "    source_rdkit_file = \"/content/required_presets/Protenix/release_data/ccd_cache/components.v20240608.cif.rdkit_mol.pkl\"\n",
        "    target_rdkit_file = \"/usr/local/lib/python3.10/dist-packages/release_data/ccd_cache/components.v20240608.cif.rdkit_mol.pkl\"\n",
        "\n",
        "    # Create the symlinks if the source files exist\n",
        "    if os.path.exists(source_ccd_file) and not os.path.exists(target_ccd_file):\n",
        "        try:\n",
        "            os.symlink(source_ccd_file, target_ccd_file)\n",
        "            print(f\"Created symlink for CCD file\")\n",
        "        except Exception as e:\n",
        "            print(f\"Error creating symlink for CCD file: {e}\")\n",
        "\n",
        "    if os.path.exists(source_rdkit_file) and not os.path.exists(target_rdkit_file):\n",
        "        try:\n",
        "            os.symlink(source_rdkit_file, target_rdkit_file)\n",
        "            print(f\"Created symlink for RDKIT file\")\n",
        "        except Exception as e:\n",
        "            print(f\"Error creating symlink for RDKIT file: {e}\")\n",
        "\n",
        "    # Create directories\n",
        "    temp_dir = \"./input\"  # Same as in kaggle_inference.py\n",
        "    output_dir = \"./output\"  # Same as in kaggle_inference.py\n",
        "    os.makedirs(temp_dir, exist_ok=True)\n",
        "    os.makedirs(output_dir, exist_ok=True)\n",
        "\n",
        "    # Read test sequences\n",
        "\n",
        "    test_sequences_df = pd.read_csv(\"/content/stanford_rna_3d_folding/test_sequences.csv\")\n",
        "    print(f\"Loaded {len(test_sequences_df)} test sequences\")\n",
        "\n",
        "    # Process each sequence\n",
        "    c1_coords_dict = {}\n",
        "    for _, row in tqdm(test_sequences_df.iterrows(), total=len(test_sequences_df)):\n",
        "        target_id = row['target_id']\n",
        "        sequence = row['sequence']\n",
        "\n",
        "        # Check if we already have predictions for this target\n",
        "        target_prediction_dir = os.path.join(output_dir, target_id, \"seed_42\", \"predictions\")\n",
        "        if os.path.exists(target_prediction_dir):\n",
        "            print(f\"Found existing prediction for {target_id}, loading coordinates\")\n",
        "            # Extract coordinates from existing predictions\n",
        "            cif_files = sorted(glob.glob(os.path.join(target_prediction_dir, f\"{target_id}_seed_42_sample_*.cif\")))\n",
        "\n",
        "            all_coords = []\n",
        "            for cif_file in cif_files:\n",
        "                coords = extract_c1_coordinates(cif_file)\n",
        "                if coords is not None:\n",
        "                    all_coords.append(coords)\n",
        "\n",
        "            if all_coords:\n",
        "                # Ensure we have 5 models\n",
        "                while len(all_coords) < 5:\n",
        "                    all_coords.append(all_coords[-1])\n",
        "                c1_coords_dict[target_id] = all_coords[:5]\n",
        "                continue\n",
        "\n",
        "        # Process the sequence if no existing prediction was found or was invalid\n",
        "        c1_coords = process_sequence(sequence, target_id, temp_dir, output_dir)\n",
        "        c1_coords_dict[target_id] = c1_coords\n",
        "\n",
        "    # Create submission file\n",
        "    create_submission(test_sequences_df, c1_coords_dict, \"submission.csv\")\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    main()"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T13:43:34.995568Z",
          "iopub.execute_input": "2025-04-04T13:43:34.99588Z",
          "iopub.status.idle": "2025-04-04T14:21:40.421062Z",
          "shell.execute_reply.started": "2025-04-04T13:43:34.995853Z",
          "shell.execute_reply": "2025-04-04T14:21:40.420066Z"
        },
        "id": "LFlU5cKHZceq",
        "outputId": "6f95b2fe-adad-45b4-8cad-9bea6e18f0ee",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Loaded 12 test sequences\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\r  0%|          | 0/12 [00:00<?, ?it/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Processing R1107: GGGGGCCACAGCAGAAGCGUUCACGUCGCAGCCCCUGUCAGCCAUUGCACUCCGGCUGCGAAUUCUGCU\n",
            "Running inference for R1107\n"
          ]
        }
      ],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [
        "pd.read_csv(\"submission.csv\")"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T14:25:33.391803Z",
          "iopub.execute_input": "2025-04-04T14:25:33.392158Z",
          "iopub.status.idle": "2025-04-04T14:25:33.424164Z",
          "shell.execute_reply.started": "2025-04-04T14:25:33.392131Z",
          "shell.execute_reply": "2025-04-04T14:25:33.423395Z"
        },
        "id": "hBIXET76Zcer",
        "outputId": "0657575b-e6bc-4694-cb69-046b93bd579e",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 419
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "             ID resname  resid  x_1  y_1  z_1  x_2  y_2  z_2  x_3  y_3  z_3  \\\n",
              "0       R1107_1       G      1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "1       R1107_2       G      2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "2       R1107_3       G      3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "3       R1107_4       G      4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "4       R1107_5       G      5  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "...         ...     ...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
              "2510  R1190_114       U    114  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "2511  R1190_115       U    115  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "2512  R1190_116       U    116  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "2513  R1190_117       U    117  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "2514  R1190_118       U    118  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
              "\n",
              "      x_4  y_4  z_4  x_5  y_5  z_5  \n",
              "0     0.0  0.0  0.0  0.0  0.0  0.0  \n",
              "1     0.0  0.0  0.0  0.0  0.0  0.0  \n",
              "2     0.0  0.0  0.0  0.0  0.0  0.0  \n",
              "3     0.0  0.0  0.0  0.0  0.0  0.0  \n",
              "4     0.0  0.0  0.0  0.0  0.0  0.0  \n",
              "...   ...  ...  ...  ...  ...  ...  \n",
              "2510  0.0  0.0  0.0  0.0  0.0  0.0  \n",
              "2511  0.0  0.0  0.0  0.0  0.0  0.0  \n",
              "2512  0.0  0.0  0.0  0.0  0.0  0.0  \n",
              "2513  0.0  0.0  0.0  0.0  0.0  0.0  \n",
              "2514  0.0  0.0  0.0  0.0  0.0  0.0  \n",
              "\n",
              "[2515 rows x 18 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-b0d0f547-c530-42c6-a52b-6ea5151b38f5\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>resname</th>\n",
              "      <th>resid</th>\n",
              "      <th>x_1</th>\n",
              "      <th>y_1</th>\n",
              "      <th>z_1</th>\n",
              "      <th>x_2</th>\n",
              "      <th>y_2</th>\n",
              "      <th>z_2</th>\n",
              "      <th>x_3</th>\n",
              "      <th>y_3</th>\n",
              "      <th>z_3</th>\n",
              "      <th>x_4</th>\n",
              "      <th>y_4</th>\n",
              "      <th>z_4</th>\n",
              "      <th>x_5</th>\n",
              "      <th>y_5</th>\n",
              "      <th>z_5</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>R1107_1</td>\n",
              "      <td>G</td>\n",
              "      <td>1</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>R1107_2</td>\n",
              "      <td>G</td>\n",
              "      <td>2</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>R1107_3</td>\n",
              "      <td>G</td>\n",
              "      <td>3</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>R1107_4</td>\n",
              "      <td>G</td>\n",
              "      <td>4</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>R1107_5</td>\n",
              "      <td>G</td>\n",
              "      <td>5</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2510</th>\n",
              "      <td>R1190_114</td>\n",
              "      <td>U</td>\n",
              "      <td>114</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2511</th>\n",
              "      <td>R1190_115</td>\n",
              "      <td>U</td>\n",
              "      <td>115</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2512</th>\n",
              "      <td>R1190_116</td>\n",
              "      <td>U</td>\n",
              "      <td>116</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2513</th>\n",
              "      <td>R1190_117</td>\n",
              "      <td>U</td>\n",
              "      <td>117</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2514</th>\n",
              "      <td>R1190_118</td>\n",
              "      <td>U</td>\n",
              "      <td>118</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>2515 rows × 18 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-b0d0f547-c530-42c6-a52b-6ea5151b38f5')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-b0d0f547-c530-42c6-a52b-6ea5151b38f5 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-b0d0f547-c530-42c6-a52b-6ea5151b38f5');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-01b81d06-467c-441c-bbd9-4367fcf24fd6\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-01b81d06-467c-441c-bbd9-4367fcf24fd6')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-01b81d06-467c-441c-bbd9-4367fcf24fd6 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"pd\",\n  \"rows\": 2515,\n  \"fields\": [\n    {\n      \"column\": \"ID\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 2515,\n        \"samples\": [\n          \"R1126_293\",\n          \"R1136_2\",\n          \"R1136_17\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"resname\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 4,\n        \"samples\": [\n          \"C\",\n          \"U\",\n          \"G\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"resid\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 176,\n        \"min\": 1,\n        \"max\": 720,\n        \"num_unique_values\": 720,\n        \"samples\": [\n          341,\n          291,\n          55\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"x_1\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"y_1\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"z_1\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"x_2\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"y_2\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"z_2\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"x_3\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"y_3\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"z_3\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"x_4\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"y_4\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"z_4\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"x_5\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"y_5\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"z_5\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.0,\n        \"min\": 0.0,\n        \"max\": 0.0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 24
        }
      ],
      "execution_count": 24
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "print(os.listdir())  # List files in the current directory\n"
      ],
      "metadata": {
        "id": "3DZsQBOlr3yS",
        "outputId": "b99407a2-2f5d-4a48-d682-452e247478a2",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['.config', 'input', 'required_presets', 'stanford-rna-3d-folding.zip', 'biotite', 'required-presets.zip', 'stanford_rna_3d_folding', 'output', 'submission.csv', 'sample_data']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import files\n",
        "files.download('/content/submission.csv')\n"
      ],
      "metadata": {
        "id": "WZfJIaajryDH",
        "outputId": "8ecbbd2a-3e1e-4ca6-84a7-ac339a1746ad",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        }
      },
      "execution_count": 46,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_17477453-3f00-4d3a-b72f-503e43d41459\", \"submission.csv\", 188943)"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# import os\n",
        "\n",
        "# # Check if the components.cif file exists\n",
        "# ccd_file = \"/kaggle/input/required-presets/Protenix/release_data/ccd_cache/components.v20240608.cif\"\n",
        "# rdkit_file = \"/kaggle/input/required-presets/Protenix/release_data/ccd_cache/components.v20240608.cif.rdkit_mol.pkl\"\n",
        "\n",
        "# print(f\"CCD file exists: {os.path.exists(ccd_file)}\")\n",
        "# print(f\"RDKIT file exists: {os.path.exists(rdkit_file)}\")\n",
        "\n",
        "# # If they don't exist, let's look for them\n",
        "# if not (os.path.exists(ccd_file) and os.path.exists(rdkit_file)):\n",
        "#     print(\"Searching for CCD files in /kaggle/input/required-presets/Protenix/release_data/...\")\n",
        "#     for root, dirs, files in os.walk(\"/kaggle/input/required-presets/Protenix/release_data/\"):\n",
        "#         for file in files:\n",
        "#             if \"components\" in file and (\"cif\" in file or \"rdkit\" in file):\n",
        "#                 print(f\"Found: {os.path.join(root, file)}\")"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T13:39:03.501642Z",
          "iopub.status.idle": "2025-04-04T13:39:03.501895Z",
          "shell.execute_reply": "2025-04-04T13:39:03.501788Z"
        },
        "id": "gruhH1wxZcer"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "trusted": true,
        "id": "RQP9AzxRZcer"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [
        "# import os\n",
        "# import json\n",
        "# import subprocess\n",
        "\n",
        "# # Create the directory structure needed\n",
        "# os.makedirs(\"/usr/local/lib/python3.10/dist-packages/release_data/ccd_cache\", exist_ok=True)\n",
        "\n",
        "# # Create symlinks to the actual files\n",
        "# source_ccd_file = \"/kaggle/input/required-presets/Protenix/release_data/ccd_cache/components.v20240608.cif\"\n",
        "# target_ccd_file = \"/usr/local/lib/python3.10/dist-packages/release_data/ccd_cache/components.v20240608.cif\"\n",
        "\n",
        "# source_rdkit_file = \"/kaggle/input/required-presets/Protenix/release_data/ccd_cache/components.v20240608.cif.rdkit_mol.pkl\"\n",
        "# target_rdkit_file = \"/usr/local/lib/python3.10/dist-packages/release_data/ccd_cache/components.v20240608.cif.rdkit_mol.pkl\"\n",
        "\n",
        "# # Check if the source files exist\n",
        "# print(f\"Source CCD file exists: {os.path.exists(source_ccd_file)}\")\n",
        "# print(f\"Source RDKIT file exists: {os.path.exists(source_rdkit_file)}\")\n",
        "\n",
        "# # Create the symlinks if the source files exist\n",
        "# if os.path.exists(source_ccd_file):\n",
        "#     try:\n",
        "#         os.symlink(source_ccd_file, target_ccd_file)\n",
        "#         print(f\"Created symlink for CCD file\")\n",
        "#     except FileExistsError:\n",
        "#         print(f\"Symlink for CCD file already exists\")\n",
        "#     except Exception as e:\n",
        "#         print(f\"Error creating symlink for CCD file: {e}\")\n",
        "# else:\n",
        "#     print(f\"Cannot create symlink, source CCD file doesn't exist\")\n",
        "\n",
        "# if os.path.exists(source_rdkit_file):\n",
        "#     try:\n",
        "#         os.symlink(source_rdkit_file, target_rdkit_file)\n",
        "#         print(f\"Created symlink for RDKIT file\")\n",
        "#     except FileExistsError:\n",
        "#         print(f\"Symlink for RDKIT file already exists\")\n",
        "#     except Exception as e:\n",
        "#         print(f\"Error creating symlink for RDKIT file: {e}\")\n",
        "# else:\n",
        "#     print(f\"Cannot create symlink, source RDKIT file doesn't exist\")\n",
        "\n",
        "# # Create RNA input JSON\n",
        "# input_json = [{\n",
        "#     \"sequences\": [\n",
        "#         {\n",
        "#             \"rnaSequence\": {\n",
        "#                 \"sequence\": \"GGGUGCUCAGUACGAGAGGAACCGCACCC\",\n",
        "#                 \"count\": 1,\n",
        "#                 \"modifications\": []\n",
        "#             }\n",
        "#         }\n",
        "#     ],\n",
        "#     \"name\": \"rna_prediction\",\n",
        "#     \"covalent_bonds\": []\n",
        "# }]\n",
        "\n",
        "# # Save input JSON\n",
        "# os.makedirs(\"./input\", exist_ok=True)\n",
        "# with open(\"./input/rna_input.json\", \"w\") as f:\n",
        "#     json.dump(input_json, f, indent=4)\n",
        "\n",
        "# # Run inference using subprocess\n",
        "# cmd = [\n",
        "#     \"python\", \"/kaggle/input/required-presets/Protenix/runner/inference.py\",\n",
        "#     \"--seeds\", \"42\",\n",
        "#     \"--dump_dir\", \"./output\",\n",
        "#     \"--input_json_path\", \"./input/rna_input.json\",\n",
        "#     \"--model.N_cycle\", \"10\",\n",
        "#     \"--sample_diffusion.N_sample\", \"5\",\n",
        "#     \"--sample_diffusion.N_step\", \"200\",\n",
        "#     \"--load_checkpoint_path\", \"/kaggle/input/required-presets/Protenix/release_data/checkpoint/model_v0.2.0.pt\",\n",
        "#     \"--use_deepspeed_evo_attention\", \"false\"\n",
        "# ]\n",
        "\n",
        "# # Run the command\n",
        "# result = subprocess.run(cmd, capture_output=True, text=True)\n",
        "# print(\"STDOUT:\", result.stdout)\n",
        "# print(\"STDERR:\", result.stderr)"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T13:39:03.50271Z",
          "iopub.status.idle": "2025-04-04T13:39:03.50311Z",
          "shell.execute_reply": "2025-04-04T13:39:03.50295Z"
        },
        "id": "iVFPRnEEZcer"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [
        "# # Check for error files\n",
        "# if os.path.exists(\"./output/ERR\"):\n",
        "#     print(\"Error directory exists!\")\n",
        "#     print(\"Contents:\")\n",
        "#     for item in os.listdir(\"./output/ERR\"):\n",
        "#         print(f\" - {item}\")\n",
        "\n",
        "#     # If there are error files, show their contents\n",
        "#     error_files = os.listdir(\"./output/ERR\")\n",
        "#     if error_files:\n",
        "#         with open(os.path.join(\"./output/ERR\", error_files[0]), \"r\") as f:\n",
        "#             print(f\"Contents of {error_files[0]}:\")\n",
        "#             print(f.read())\n",
        ""
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T10:43:02.500381Z",
          "iopub.execute_input": "2025-04-04T10:43:02.500601Z",
          "iopub.status.idle": "2025-04-04T10:43:02.506377Z",
          "shell.execute_reply.started": "2025-04-04T10:43:02.500582Z",
          "shell.execute_reply": "2025-04-04T10:43:02.505427Z"
        },
        "id": "JqMJRpmiZces"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [
        "# import os\n",
        "# import json\n",
        "# import pandas as pd\n",
        "# import biotite.structure.io.pdbx as pdbx\n",
        "# import matplotlib.pyplot as plt\n",
        "# from mpl_toolkits.mplot3d import Axes3D\n",
        "# import numpy as np\n",
        "\n",
        "# # List all the prediction files\n",
        "# output_dir = \"/kaggle/working/output/rna_prediction/seed_42/predictions/\"\n",
        "# cif_files = [f for f in os.listdir(output_dir) if f.endswith(\".cif\")]\n",
        "# print(f\"Found {len(cif_files)} prediction files:\")\n",
        "# for file in cif_files:\n",
        "#     print(f\" - {file}\")\n",
        "\n",
        "# # Read and analyze the first prediction file\n",
        "# if cif_files:\n",
        "#     # Read CIF file\n",
        "#     cif_file = os.path.join(output_dir, cif_files[0])\n",
        "#     with open(cif_file, 'r') as f:\n",
        "#         cif_data = pdbx.CIFFile.read(f)\n",
        "\n",
        "#     atom_array = pdbx.get_structure(cif_data, model=1)\n",
        "\n",
        "#     print(f\"\\nStructure information for {cif_files[0]}:\")\n",
        "#     print(f\"Number of atoms: {len(atom_array)}\")\n",
        "#     print(f\"Residue count: {len(np.unique(atom_array.res_id))}\")\n",
        "\n",
        "#     # Clean and extract C1' atoms\n",
        "#     atom_names_clean = np.char.strip(atom_array.atom_name.astype(str))\n",
        "#     mask_c1 = atom_names_clean == \"C1'\"\n",
        "#     c1_atoms = atom_array[mask_c1]\n",
        "#     c1_coords = c1_atoms.coord\n",
        "\n",
        "#     print(f\"\\nFound {len(c1_atoms)} C1' atoms\")\n",
        "\n",
        "#     # Create DataFrame\n",
        "#     df = pd.DataFrame({\n",
        "#         \"res_name\": c1_atoms.res_name,\n",
        "#         \"res_id\": c1_atoms.res_id,\n",
        "#         \"chain_id\": c1_atoms.chain_id,\n",
        "#         \"x\": c1_coords[:, 0],\n",
        "#         \"y\": c1_coords[:, 1],\n",
        "#         \"z\": c1_coords[:, 2]\n",
        "#     })\n",
        "\n",
        "#     print(\"\\nFirst few C1' atoms:\")\n",
        "#     print(df.head())\n",
        "\n",
        "#     # Save to CSV and JSON\n",
        "#     df.to_csv(\"c1_prime_coordinates.csv\", index=False)\n",
        "#     df.to_json(\"c1_prime_coordinates.json\", orient=\"records\", indent=2)\n",
        "#     print(\"Saved C1' coordinates to CSV and JSON.\")\n",
        "\n",
        "#     # Plot C1' atoms with backbone\n",
        "#     fig = plt.figure(figsize=(10, 8))\n",
        "#     ax = fig.add_subplot(111, projection='3d')\n",
        "\n",
        "#     chain_ids = np.unique(c1_atoms.chain_id)\n",
        "#     colors = plt.cm.rainbow(np.linspace(0, 1, len(chain_ids)))\n",
        "\n",
        "#     for i, chain_id in enumerate(chain_ids):\n",
        "#         chain_mask = c1_atoms.chain_id == chain_id\n",
        "#         chain_df = df[df[\"chain_id\"] == chain_id]\n",
        "#         chain_df_sorted = chain_df.sort_values(\"res_id\")\n",
        "\n",
        "#         ax.scatter(\n",
        "#             chain_df_sorted[\"x\"],\n",
        "#             chain_df_sorted[\"y\"],\n",
        "#             chain_df_sorted[\"z\"],\n",
        "#             c=[colors[i]],\n",
        "#             label=f\"Chain {chain_id}\",\n",
        "#             alpha=0.9,\n",
        "#             s=30\n",
        "#         )\n",
        "\n",
        "#         ax.plot(\n",
        "#             chain_df_sorted[\"x\"],\n",
        "#             chain_df_sorted[\"y\"],\n",
        "#             chain_df_sorted[\"z\"],\n",
        "#             color=colors[i],\n",
        "#             alpha=0.6,\n",
        "#             linewidth=2\n",
        "#         )\n",
        "\n",
        "#     ax.set_xlabel(\"X (Å)\")\n",
        "#     ax.set_ylabel(\"Y (Å)\")\n",
        "#     ax.set_zlabel(\"Z (Å)\")\n",
        "#     ax.set_title(f\"C1' atom backbone - {cif_files[0]}\")\n",
        "#     ax.legend()\n",
        "#     plt.tight_layout()\n",
        "#     plt.show()\n",
        "\n",
        "#     # --- NEW: Plot all atoms in another image ---\n",
        "#     print(\"\\nGenerating full atom visualization...\")\n",
        "\n",
        "#     all_coords = atom_array.coord\n",
        "#     all_chain_ids = np.unique(atom_array.chain_id)\n",
        "#     colors_all = plt.cm.rainbow(np.linspace(0, 1, len(all_chain_ids)))\n",
        "\n",
        "#     fig_all = plt.figure(figsize=(10, 8))\n",
        "#     ax_all = fig_all.add_subplot(111, projection='3d')\n",
        "\n",
        "#     for i, chain_id in enumerate(all_chain_ids):\n",
        "#         chain_mask = atom_array.chain_id == chain_id\n",
        "#         ax_all.scatter(\n",
        "#             all_coords[chain_mask, 0],\n",
        "#             all_coords[chain_mask, 1],\n",
        "#             all_coords[chain_mask, 2],\n",
        "#             c=[colors_all[i]],\n",
        "#             label=f\"Chain {chain_id}\",\n",
        "#             alpha=0.6,\n",
        "#             s=5  # smaller for all atoms\n",
        "#         )\n",
        "\n",
        "#     ax_all.set_xlabel(\"X (Å)\")\n",
        "#     ax_all.set_ylabel(\"Y (Å)\")\n",
        "#     ax_all.set_zlabel(\"Z (Å)\")\n",
        "#     ax_all.set_title(f\"All atoms - {cif_files[0]}\")\n",
        "#     ax_all.legend()\n",
        "#     plt.tight_layout()\n",
        "#     plt.show()\n"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T10:53:14.639027Z",
          "iopub.execute_input": "2025-04-04T10:53:14.639415Z",
          "iopub.status.idle": "2025-04-04T10:53:15.084099Z",
          "shell.execute_reply.started": "2025-04-04T10:53:14.639385Z",
          "shell.execute_reply": "2025-04-04T10:53:15.083226Z"
        },
        "id": "faSK1qgOZces"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [
        "# !pip install nglview"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T10:56:30.789313Z",
          "iopub.execute_input": "2025-04-04T10:56:30.789587Z",
          "iopub.status.idle": "2025-04-04T10:57:06.942843Z",
          "shell.execute_reply.started": "2025-04-04T10:56:30.789566Z",
          "shell.execute_reply": "2025-04-04T10:57:06.941693Z"
        },
        "id": "9JZDuuSjZces"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [
        "# import json\n",
        "# import os\n",
        "# import pandas as pd\n",
        "# import matplotlib.pyplot as plt\n",
        "# import numpy as np\n",
        "\n",
        "# # Path to the summary confidence JSON file\n",
        "# confidence_file = \"/kaggle/working/output/rna_prediction/seed_42/predictions/rna_prediction_seed_42_summary_confidence_sample_0.json\"\n",
        "\n",
        "# # Read and display the file contents\n",
        "# with open(confidence_file, 'r') as f:\n",
        "#     confidence_data = json.load(f)\n",
        "\n",
        "# # Show the keys in the JSON file\n",
        "# print(\"Keys in the confidence file:\")\n",
        "# for key in confidence_data.keys():\n",
        "#     print(f\" - {key}\")\n",
        "\n",
        "# # Function to display and visualize specific metrics\n",
        "# def analyze_metric(data, metric_name):\n",
        "#     if metric_name in data:\n",
        "#         print(f\"\\n{metric_name} data:\")\n",
        "\n",
        "#         # Handle different data types\n",
        "#         if isinstance(data[metric_name], (int, float)):\n",
        "#             print(f\"{metric_name}: {data[metric_name]}\")\n",
        "\n",
        "#         elif isinstance(data[metric_name], list):\n",
        "#             print(f\"{metric_name} (first 5 values): {data[metric_name][:5]}\")\n",
        "\n",
        "#             # Plot if it's a list of numbers\n",
        "#             if data[metric_name] and isinstance(data[metric_name][0], (int, float)):\n",
        "#                 plt.figure(figsize=(10, 6))\n",
        "#                 plt.plot(data[metric_name])\n",
        "#                 plt.title(f\"{metric_name} across residues\")\n",
        "#                 plt.xlabel(\"Residue index\")\n",
        "#                 plt.ylabel(metric_name)\n",
        "#                 plt.grid(True, alpha=0.3)\n",
        "#                 plt.savefig(f\"{metric_name}_plot.png\")\n",
        "#                 plt.close()\n",
        "#                 print(f\"Saved visualization to '{metric_name}_plot.png'\")\n",
        "\n",
        "#         elif isinstance(data[metric_name], dict):\n",
        "#             # For dictionaries, show keys and sample values\n",
        "#             print(f\"{metric_name} contains {len(data[metric_name])} keys:\")\n",
        "#             for k in list(data[metric_name].keys())[:5]:\n",
        "#                 print(f\"  - {k}: {data[metric_name][k]}\")\n",
        "#     else:\n",
        "#         print(f\"\\n{metric_name} not found in the data\")\n",
        "\n",
        "# # Analyze common confidence metrics\n",
        "# analyze_metric(confidence_data, \"plddt\")  # Per-residue confidence scores\n",
        "# analyze_metric(confidence_data, \"ranking_score\")  # Overall ranking score of the model\n",
        "# analyze_metric(confidence_data, \"gpde\")  # Global predicted distance error\n",
        "\n",
        "# # If there's PAE (Predicted Aligned Error) matrix, visualize it\n",
        "# if \"pae\" in confidence_data and isinstance(confidence_data[\"pae\"], list):\n",
        "#     pae_data = np.array(confidence_data[\"pae\"])\n",
        "\n",
        "#     if len(pae_data.shape) == 2:\n",
        "#         plt.figure(figsize=(10, 8))\n",
        "#         im = plt.imshow(pae_data, cmap='viridis_r')\n",
        "#         plt.colorbar(im, label=\"Predicted Aligned Error (Å)\")\n",
        "#         plt.title(\"PAE Matrix\")\n",
        "#         plt.xlabel(\"Residue\")\n",
        "#         plt.ylabel(\"Residue\")\n",
        "#         plt.savefig(\"pae_matrix.png\")\n",
        "#         plt.close()\n",
        "#         print(\"\\nSaved PAE matrix visualization to 'pae_matrix.png'\")\n",
        "\n",
        "# # Display the structure's overall quality assessment\n",
        "# print(\"\\nOverall structure quality assessment:\")\n",
        "# quality_metrics = [\"ranking_score\", \"gpde\", \"plddt_avg\", \"pae_avg\"]\n",
        "# for metric in quality_metrics:\n",
        "#     if metric in confidence_data:\n",
        "#         print(f\" - {metric}: {confidence_data[metric]}\")"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T10:43:03.559591Z",
          "iopub.execute_input": "2025-04-04T10:43:03.55982Z",
          "iopub.status.idle": "2025-04-04T10:43:03.574372Z",
          "shell.execute_reply.started": "2025-04-04T10:43:03.5598Z",
          "shell.execute_reply": "2025-04-04T10:43:03.573528Z"
        },
        "id": "sz3O5Z7CZces"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [
        "# import json\n",
        "# import pprint\n",
        "\n",
        "# # Path to the summary confidence JSON file\n",
        "# confidence_file = \"/kaggle/working/output/rna_prediction/seed_42/predictions/rna_prediction_seed_42_summary_confidence_sample_0.json\"\n",
        "\n",
        "# # Read and display the complete file contents\n",
        "# with open(confidence_file, 'r') as f:\n",
        "#     confidence_data = json.load(f)\n",
        "\n",
        "# # Use pretty print to display formatted JSON\n",
        "# print(\"Complete JSON content:\")\n",
        "# pp = pprint.PrettyPrinter(indent=2)\n",
        "# pp.pprint(confidence_data)\n",
        "\n",
        "# # Or alternatively, print it with json.dumps for more control over formatting\n",
        "# print(\"\\nAlternative JSON formatting:\")\n",
        "# print(json.dumps(confidence_data, indent=2))\n"
      ],
      "metadata": {
        "trusted": true,
        "execution": {
          "iopub.status.busy": "2025-04-04T10:43:03.575142Z",
          "iopub.execute_input": "2025-04-04T10:43:03.575459Z",
          "iopub.status.idle": "2025-04-04T10:43:03.598221Z",
          "shell.execute_reply.started": "2025-04-04T10:43:03.575428Z",
          "shell.execute_reply": "2025-04-04T10:43:03.597511Z"
        },
        "id": "rvXwYLiWZces"
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "trusted": true,
        "id": "qQyxnV-BZcet"
      },
      "outputs": [],
      "execution_count": null
    }
  ]
}